// @business-app/fabric/lib-cjs/components/FilterSearchBox/examples/FilterSearchBox.SingleFilters.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterSearchBoxSingleFiltersExample = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var Label_1 = require("@fluentui/react/lib/Label");
var __1 = require("../");
var filters = [{ key: 'only', text: 'Only Filter' }];
var FilterSearchBoxSingleFiltersExample = /** @class */ (function (_super) {
    tslib_1.__extends(FilterSearchBoxSingleFiltersExample, _super);
    function FilterSearchBoxSingleFiltersExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            searchText: '',
            selectedFilterKey: filters[0].key
        };
        _this._onSearchTextChange = function (event, text) {
            _this.setState({ searchText: text });
        };
        _this._onFilterChange = function (key) {
            _this.setState({ selectedFilterKey: key });
        };
        return _this;
    }
    FilterSearchBoxSingleFiltersExample.prototype.render = function () {
        var _a = this.state, searchText = _a.searchText, selectedFilterKey = _a.selectedFilterKey;
        return (React.createElement(React.Fragment, null,
            React.createElement(__1.FilterSearchBox, { strings: {
                    searchBoxPlaceholder: 'Search',
                    filterButtonTooltipText: 'Tooltip text'
                }, searchText: searchText, onSearchTextChange: this._onSearchTextChange, filters: filters, selectedFilterKey: selectedFilterKey, onFilterChange: this._onFilterChange, styles: {
                    root: {
                        marginBottom: 16
                    }
                } }),
            React.createElement(Label_1.Label, null, "Current search text: " + searchText),
            React.createElement(Label_1.Label, null, "Current filter key: " + selectedFilterKey)));
    };
    return FilterSearchBoxSingleFiltersExample;
}(React.Component));
exports.FilterSearchBoxSingleFiltersExample = FilterSearchBoxSingleFiltersExample;
