// @business-app/fabric/lib/components/Toolbox/ToolboxItem.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __assign, __extends } from "tslib";
import { mergeStyles } from '@fluentui/react/lib/Styling';
import { BaseComponent, divProperties, getNativeProps } from '@fluentui/react/lib/Utilities';
import * as React from 'react';
var ToolboxItem = /** @class */ (function (_super) {
    __extends(ToolboxItem, _super);
    function ToolboxItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ToolboxItem.prototype.render = function () {
        var className = mergeStyles({
            minHeight: 0,
            display: 'flex',
            flexGrow: 1,
            selectors: {
                // For IE 10+ overflow property with fieldheaderregion minheight
                '@media all and (-ms-high-contrast: none)': {
                    minHeight: 'calc(100% - 48px)'
                }
            }
        }, this.props.className);
        return (React.createElement("div", __assign({}, getNativeProps(this.props, divProperties), { className: className }), this.props.children));
    };
    return ToolboxItem;
}(BaseComponent));
export { ToolboxItem };
