// @business-app/fabric/lib/components/Toolbox/Toolbox.styles.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __spreadArrays } from "tslib";
import { AnimationClassNames, normalize, ZIndexes } from '@fluentui/react/lib/Styling';
var linkStyles = function (props) {
    var palette = props.theme.palette;
    return [
        {
            padding: 0,
            minHeight: '48px',
            width: '100%',
            border: '1px solid transparent',
            selectors: {
                ':hover': {
                    backgroundColor: palette.neutralLight
                },
                ':active': {
                    backgroundColor: palette.neutralQuaternary
                }
            }
        }
    ];
};
export var getStyles = function (props) {
    var className = props.className, theme = props.theme, linksAreExpanded = props.linksAreExpanded, drawerIsExpanded = props.drawerIsExpanded;
    var palette = theme.palette;
    return {
        root: [
            className,
            {
                display: 'flex',
                flexShrink: 0
            }
        ],
        focusZone: {
            // Needs to be higher than the drawer
            zIndex: ZIndexes.Nav + 1,
            borderRight: '1px solid ' + palette.neutralQuaternaryAlt,
            boxShadow: drawerIsExpanded ? 'none' : '0 0 18px rgba(0,0,0,.05)',
            transition: 'boxShadow .3s ease',
            backgroundColor: palette.neutralLighter
        },
        list: [
            normalize,
            {
                display: 'inline-flex',
                flexDirection: 'column',
                width: linksAreExpanded ? '220px' : '48px'
            }
        ],
        drawer: [
            {
                display: drawerIsExpanded ? 'flex' : 'none',
                flexDirection: 'column',
                backgroundColor: palette.neutralLighter,
                zIndex: ZIndexes.Nav,
                boxShadow: '0 0 18px rgba(0,0,0,.05)',
                borderRight: '1px solid ' + palette.neutralQuaternaryAlt
            },
            AnimationClassNames.slideRightIn40
        ],
        drawerHeaderRegion: {
            position: 'relative',
            height: '48px',
            selectors: {
                // For IE 10+ flex box
                '@media all and (-ms-high-contrast: none)': {
                    minHeight: '48px'
                }
            }
        },
        drawerHeader: {
            userSelect: 'none',
            fontWeight: '600',
            marginLeft: '16px',
            fontSize: '18px',
            lineHeight: '48px',
            display: 'inline-block',
            // Clip the header when it gets too big
            whiteSpace: 'nowrap',
            overflow: 'hidden',
            textOverflow: 'ellipsis',
            // marginLeft + drawerClose width = 64px
            maxWidth: 'calc(100% - 64px)',
            // Remove any default margin top and bottom styles that may be inherited from the H2
            marginTop: 0,
            marginBottom: 0
        },
        drawerClose: {
            position: 'absolute',
            right: 0,
            top: 0,
            height: '48px',
            padding: 0
        },
        drawerCloseIcon: {
            fontSize: '16px',
            margin: '16px',
            lineHeight: '16px',
            height: '16px',
            width: '16px'
        },
        item: linkStyles(props),
        itemIsSelected: __spreadArrays(linkStyles(props), [
            {
                backgroundColor: palette.white,
                selectors: {
                    '&:hover,&:active': {
                        backgroundColor: palette.white,
                        color: palette.neutralPrimary
                    }
                }
            }
        ]),
        itemContent: {
            display: 'flex',
            alignItems: 'center'
        },
        text: [
            {
                textOverflow: 'ellipsis',
                whiteSpace: 'nowrap',
                overflow: 'hidden',
                display: 'inline-block',
                maxWidth: '154px'
            }
        ],
        iconContainer: {
            display: 'flex',
            marginLeft: '15px',
            marginRight: '15px'
        },
        icon: {
            fontSize: '16px',
            lineHeight: '16px',
            height: '16px',
            width: '16px',
            selectors: {
                '@media screen and (-ms-high-contrast: active)': {
                    fill: palette.white
                },
                '@media screen and (-ms-high-contrast: white-on-black)': {
                    fill: palette.white
                },
                '@media screen and (-ms-high-contrast: black-on-white)': {
                    fill: palette.black
                }
            }
        },
        selectionIndicator: {
            position: 'absolute',
            left: '3px',
            width: '4px',
            height: '24px',
            background: palette.themePrimary,
            top: '11px',
            borderRadius: '2px'
        }
    };
};
