// @business-app/fabric/lib-cjs/components/Sidebar/SidebarAccordion.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.styles = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
var Sidebar_styles_1 = require("./Sidebar.styles");
var Sidebar_styles_2 = require("./Sidebar.styles");
var SidebarButton_styles_1 = require("./SidebarButton.styles");
var styles = function (props) {
    var currentSidebarColors = Sidebar_styles_1.getSidebarColors(props.theme);
    var buttonStyles = SidebarButton_styles_1.getSidebarButtonStyles(props.theme);
    var accordionStyles = {
        root: {
            backgroundColor: currentSidebarColors.background
        },
        rootHovered: {
            backgroundColor: currentSidebarColors.backgroundHovered
        },
        rootPressed: {
            backgroundColor: currentSidebarColors.backgroundPressed
        },
        rootChecked: {
            backgroundColor: props.isContentVisible
                ? currentSidebarColors.background
                : currentSidebarColors.backgroundChecked
        },
        rootCheckedHovered: {
            backgroundColor: props.isContentVisible
                ? currentSidebarColors.backgroundHovered
                : currentSidebarColors.backgroundChecked
        },
        rootCheckedPressed: {
            backgroundColor: props.isContentVisible
                ? currentSidebarColors.backgroundPressed
                : currentSidebarColors.backgroundChecked
        },
        labelChecked: {
            fontWeight: props.isContentVisible ? 'normal' : 'bold',
            fontFamily: props.isContentVisible ? Sidebar_styles_2.sidebarFonts.segoeUi : Sidebar_styles_2.sidebarFonts.segoeUiSemibold
        },
        flexContainer: {
            selectors: {
                '.is-checked > &:before': { content: props.isContentVisible ? 'none' : '""' }
            }
        },
        menuIcon: {
            fontSize: '12px'
        }
    };
    return Styling_1.concatStyleSets(buttonStyles, accordionStyles);
};
exports.styles = styles;
