// @business-app/fabric/lib-cjs/components/Sidebar/Sidebar.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCollapsedButtonStyles = exports.getButtonColoredStyles = exports.getSidebarStyles = exports.getSidebarColors = exports.sidebarFonts = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
var Utilities_1 = require("@fluentui/react/lib/Utilities");
var Sidebar_types_1 = require("./Sidebar.types");
exports.sidebarFonts = {
    segoeUiSemibold: 'wf_segoe-ui_semibold, "Segoe UI Semibold", "Segoe WP Semibold", "Segoe UI", "Segoe WP", Tahoma, Arial, sans-serif',
    segoeUiSemilight: 'wf_segoe-ui_semilight, "Segoe UI Light", "Segoe WP Light", "Segoe UI", "Segoe WP",Tahoma, Arial, sans-serif',
    segoeUi: 'wf_segoe-ui_normal, "Segoe UI", "Segoe WP", Tahoma, Arial, sans-serif'
};
var getSidebarColors = function (theme) { return ({
    background: theme.palette.neutralLighter,
    backgroundHovered: theme.palette.neutralQuaternaryAlt,
    backgroundChecked: theme.palette.white,
    backgroundPressed: theme.palette.neutralQuaternary,
    textColor: theme.palette.neutralPrimary
}); };
exports.getSidebarColors = getSidebarColors;
exports.getSidebarStyles = Utilities_1.memoizeFunction(function (theme, customStyles) {
    var currentSidebarColors = exports.getSidebarColors(theme);
    var sidebarStyles = {
        root: {
            height: '100%',
            backgroundColor: currentSidebarColors.background,
            width: Sidebar_types_1.SidebarStylingConstants.sidebarWidth,
            position: 'relative',
            borderRight: "solid 1px " + currentSidebarColors.backgroundPressed
        },
        rootCollapsed: {
            width: Sidebar_types_1.SidebarStylingConstants.sidebarCollapsedWidth,
            backgroundColor: currentSidebarColors.background,
            height: '100%',
            position: 'relative'
        },
        content: {
            height: 'auto',
            width: '100%',
            backgroundColor: currentSidebarColors.background,
            overflowX: 'hidden'
        },
        contentCollapsed: {
            backgroundColor: currentSidebarColors.background,
            overflow: 'hidden'
        },
        footer: {
            bottom: '0',
            position: 'absolute',
            width: '100%'
        }
    };
    return Styling_1.concatStyleSets(sidebarStyles, customStyles);
});
exports.getButtonColoredStyles = Utilities_1.memoizeFunction(function (theme, customStyles) {
    var currentSidebarColors = exports.getSidebarColors(theme);
    var buttonStyles = {
        root: {
            display: 'flex',
            borderColor: 'transparent',
            backgroundColor: currentSidebarColors.background
        },
        rootExpanded: {
            backgroundColor: currentSidebarColors.background
        },
        rootExpandedHovered: {
            backgroundColor: currentSidebarColors.backgroundHovered
        },
        rootDisabled: {
            backgroundColor: currentSidebarColors.background
        },
        rootHovered: {
            backgroundColor: currentSidebarColors.backgroundHovered
        },
        rootPressed: {
            backgroundColor: currentSidebarColors.backgroundPressed
        },
        rootChecked: {
            backgroundColor: currentSidebarColors.backgroundChecked
        },
        rootCheckedHovered: {
            backgroundColor: currentSidebarColors.backgroundChecked
        },
        rootCheckedPressed: {
            backgroundColor: currentSidebarColors.backgroundChecked
        },
        flexContainer: {
            width: '100%'
        },
        icon: {
            fill: currentSidebarColors.textColor,
            color: currentSidebarColors.textColor
        },
        iconDisabled: {
            fill: theme.semanticColors.buttonTextDisabled,
            color: theme.semanticColors.buttonTextDisabled
        },
        iconExpanded: {
            color: currentSidebarColors.textColor,
            fill: currentSidebarColors.textColor
        },
        labelDisabled: {
            color: theme.semanticColors.buttonTextDisabled
        },
        label: {
            color: currentSidebarColors.textColor
        },
        menuIcon: {
            color: currentSidebarColors.textColor,
            fontSize: '10px'
        },
        menuIconDisabled: {
            color: theme.semanticColors.buttonTextDisabled
        }
    };
    return Styling_1.concatStyleSets(buttonStyles, customStyles);
});
exports.getCollapsedButtonStyles = Utilities_1.memoizeFunction(function (theme) {
    var currentSidebarColors = exports.getSidebarColors(theme);
    return {
        rootCheckedHovered: {
            backgroundColor: currentSidebarColors.backgroundChecked,
            selectors: {
                '&.is-expanded:hover': {
                    backgroundColor: currentSidebarColors.backgroundHovered
                }
            }
        },
        flexContainer: {
            selectors: {
                '.is-expanded > &:before': {
                    content: 'none'
                }
            }
        }
    };
});
