// @business-app/fabric/lib-cjs/components/Sidebar/SidebarButton.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSidebarChildrenStyles = exports.getSidebarButtonStyles = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
var Utilities_1 = require("@fluentui/react/lib/Utilities");
var Sidebar_styles_1 = require("./Sidebar.styles");
var Sidebar_types_1 = require("./Sidebar.types");
exports.getSidebarButtonStyles = Utilities_1.memoizeFunction(function (theme, sidebarButtonStyles, customStyles) {
    var currentSidebarColors = Sidebar_styles_1.getSidebarColors(theme);
    var buttonStyles = {
        root: {
            width: '100%',
            minWidth: Sidebar_types_1.SidebarStylingConstants.sidebarCollapsedWidth,
            padding: '0',
            minHeight: '40px',
            height: 'auto',
            lineHeight: '20px',
            borderColor: 'transparent',
            backgroundColor: currentSidebarColors.background
        },
        rootHovered: {
            backgroundColor: currentSidebarColors.backgroundHovered
        },
        rootFocused: {
            '@media screen and (-ms-high-contrast: active)': {
                border: '3px solid'
            }
        },
        flexContainer: {
            justifyContent: 'flex-start',
            minHeight: '40px',
            height: 'auto',
            selectors: {
                '.is-checked > &:before': {
                    content: '""',
                    width: '4px',
                    height: '24px',
                    backgroundColor: theme.palette.themePrimary,
                    borderRadius: '4px',
                    position: 'absolute',
                    left: '4px',
                    top: '50%',
                    transform: 'translateY(-50%)'
                }
            }
        },
        icon: {
            marginLeft: '16px',
            marginRight: '20px',
            fontSize: Sidebar_types_1.SidebarStylingConstants.sidebarIconSize,
            height: Sidebar_types_1.SidebarStylingConstants.sidebarIconSize,
            width: Sidebar_types_1.SidebarStylingConstants.sidebarIconSize
        },
        labelChecked: {
            fontWeight: 'bold',
            fontFamily: Sidebar_styles_1.sidebarFonts.segoeUiSemibold
        },
        labelDisabled: {
            fontFamily: Sidebar_styles_1.sidebarFonts.segoeUi
        },
        label: {
            verticalAlign: 'middle',
            textAlign: 'left',
            fontSize: '14px',
            whiteSpace: 'normal',
            display: 'inline-block',
            fontFamily: Sidebar_styles_1.sidebarFonts.segoeUi,
            float: 'left',
            lineHeight: '20px',
            padding: '10px 0',
            fontWeight: 'normal'
        },
        menuIcon: {
            marginLeft: 'auto',
            marginRight: '10px',
            fontSize: Sidebar_types_1.SidebarStylingConstants.sidebarIconSize,
            height: Sidebar_types_1.SidebarStylingConstants.sidebarIconSize,
            width: Sidebar_types_1.SidebarStylingConstants.sidebarIconSize
        }
    };
    return Styling_1.concatStyleSets(buttonStyles, sidebarButtonStyles, customStyles);
});
exports.getSidebarChildrenStyles = Utilities_1.memoizeFunction(function (theme, customStyles) {
    var sidebarChildrenStyles = {
        flexContainer: {
            paddingLeft: '52px'
        },
        icon: {
            padding: '0',
            marginLeft: '0'
        }
    };
    return Styling_1.concatStyleSets(exports.getSidebarButtonStyles(theme), sidebarChildrenStyles, customStyles);
});
