// @business-app/fabric/lib-cjs/components/Error/Error.test.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Enzyme = require("enzyme");
require("jest");
var Styling_1 = require("@fluentui/react/lib/Styling");
var React = require("react");
var Error_base_1 = require("./Error.base");
var Error_styles_1 = require("./Error.styles");
describe('Error', function () {
    var wrapper;
    var props;
    beforeEach(function () {
        props = {
            styles: Error_styles_1.styles,
            theme: Styling_1.getTheme()
        };
    });
    afterEach(function () { return wrapper.unmount(); });
    it('should not unexpectedly change', function () {
        var seeDetailsShowLabel = 'See details';
        var seeDetailsHideLabel = 'Hide details';
        wrapper = Enzyme.mount(React.createElement(Error_base_1.ErrorBase, tslib_1.__assign({}, props, { title: 'Title', subtitle: 'Subtitle', seeDetailsLinkShowLabel: seeDetailsShowLabel, seeDetailsLinkHideLabel: seeDetailsHideLabel, errorDetails: {
                messageLabel: 'message label',
                message: 'message',
                customerSupportLink: {
                    label: 'Customer Support',
                    url: '/',
                    target: '_blank'
                }
            }, learnMoreLink: {
                label: 'Learn more',
                url: '/',
                target: '_blank'
            }, callToActionLink: {
                label: 'Call to Action',
                url: '/',
                target: '_blank'
            } })));
        expect(wrapper).toMatchSnapshot();
    });
    it('should expand See Details section', function () {
        var seeDetailsShowLabel = 'See details';
        var seeDetailsHideLabel = 'Hide details';
        wrapper = Enzyme.mount(React.createElement(Error_base_1.ErrorBase, tslib_1.__assign({}, props, { title: 'Title', seeDetailsLinkShowLabel: seeDetailsShowLabel, seeDetailsLinkHideLabel: seeDetailsHideLabel, errorDetails: {
                messageLabel: 'message label',
                message: 'message',
                customerSupportLink: {
                    label: 'Customer Support',
                    url: '/',
                    target: '_blank'
                }
            } })));
        expect(wrapper.find('.ba-Error-SeeDetailsButton').length).toBe(4);
        expect(wrapper.find('.ba-Error-SeeDetailsButton').first().getElement().props.text).toEqual(seeDetailsShowLabel);
        wrapper.find('.ba-Error-SeeDetailsButton').first().simulate('click');
        expect(wrapper.find('.ba-Error-SeeDetailsButton').first().getElement().props.text).toEqual(seeDetailsHideLabel);
        expect(wrapper.find('.ba-Error-SeeDetailsDiv').length).toBe(1);
        expect(wrapper.find('.ba-Error-CustomerSupportLink').length).toBe(3);
    });
    it('should render Learn More link when subtitle not provided', function () {
        wrapper = Enzyme.mount(React.createElement(Error_base_1.ErrorBase, tslib_1.__assign({ title: 'Title' }, props, { learnMoreLink: {
                label: 'Learn more',
                url: '/',
                target: '_blank'
            }, callToActionLink: {
                label: 'Call to Action',
                url: '/',
                target: '_blank'
            } })));
        expect(wrapper.find('.ba-Error-LearnMoreLink').length).toBe(3);
        expect(wrapper.find('.ba-Error-ImageLarge').length).toBe(1);
    });
    it('should render all provided controls optionally', function () {
        var seeDetailsShowLabel = 'See details';
        var seeDetailsHideLabel = 'Hide details';
        wrapper = Enzyme.mount(React.createElement(Error_base_1.ErrorBase, tslib_1.__assign({}, props, { title: 'Title', subtitle: 'Subtitle', seeDetailsLinkShowLabel: seeDetailsShowLabel, seeDetailsLinkHideLabel: seeDetailsHideLabel, errorDetails: {
                messageLabel: 'message label',
                message: 'message',
                customerSupportLink: {
                    label: 'Customer Support',
                    url: '/',
                    target: '_blank'
                }
            }, learnMoreLink: {
                label: 'Learn more',
                url: '/',
                target: '_blank'
            }, callToActionLink: {
                label: 'Call to Action',
                url: '/',
                target: '_blank'
            } })));
        expect(wrapper.find('.ba-Error-Title').length).toBe(1);
        expect(wrapper.find('.ba-Error-Subtitle').length).toBe(1);
        expect(wrapper.find('.ba-Error-LearnMoreLink').length).toBe(3);
        expect(wrapper.find('.ba-Error-CallToActionButton').length).toBeGreaterThanOrEqual(6);
        expect(wrapper.find('.ba-Error-ImageLarge').length).toBe(1);
        expect(wrapper.find('.ba-Error-SeeDetailsButton').length).toBe(4);
        wrapper.find('.ba-Error-SeeDetailsButton').first().simulate('click');
        expect(wrapper.find('.ba-Error-SeeDetailsDiv').length).toBe(1);
        expect(wrapper.find('.ba-Error-CustomerSupportLink').length).toBe(3);
    });
});
