// @business-app/fabric/lib-cjs/components/Error/Error.base.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorBase = void 0;
var tslib_1 = require("tslib");
var Button_1 = require("@fluentui/react/lib/Button");
var Link_1 = require("@fluentui/react/lib/Link");
var Utilities_1 = require("@fluentui/react/lib/Utilities");
var React = require("react");
var Icon_1 = require("@fluentui/react/lib/Icon");
var Styling_1 = require("@fluentui/react/lib/Styling");
var Error_styles_1 = require("./Error.styles");
var ErrorSplashSmall_1 = require("./assets/ErrorSplashSmall");
var ErrorSplashMedium_1 = require("./assets/ErrorSplashMedium");
var ErrorSplashLarge_1 = require("./assets/ErrorSplashLarge");
var getClassNames = Utilities_1.classNamesFunction();
/**
 * Provides an error splash page with optional details of the cause and
 * next steps to contact self-correct or contact customer service
 * @export
 */
function ErrorBase(props) {
    var className = props.className, title = props.title, subtitle = props.subtitle, seeDetailsLinkShowLabel = props.seeDetailsLinkShowLabel, seeDetailsLinkHideLabel = props.seeDetailsLinkHideLabel, learnMoreLink = props.learnMoreLink, callToActionLink = props.callToActionLink, errorDetails = props.errorDetails, styles = props.styles, _a = props.theme, theme = _a === void 0 ? Styling_1.getTheme() : _a, rest = tslib_1.__rest(props, ["className", "title", "subtitle", "seeDetailsLinkShowLabel", "seeDetailsLinkHideLabel", "learnMoreLink", "callToActionLink", "errorDetails", "styles", "theme"]);
    var _b = React.useState(false), isContentVisible = _b[0], setIsContentVisible = _b[1];
    var classNames = getClassNames(styles, {
        theme: theme,
        className: className
    });
    var toggleState = React.useCallback(function () { return setIsContentVisible(!isContentVisible); }, [isContentVisible]);
    var headerStyles = React.useCallback(function () { return Styling_1.concatStyleSets(Error_styles_1.getDefaultButtonStyles(theme), classNames.callToActionButton); }, [theme]);
    var menuIconProps = { iconName: isContentVisible ? 'ChevronUp' : 'ChevronDown' };
    var seeDetailsButtonProps = {
        menuIconProps: menuIconProps,
        ariaExpanded: isContentVisible
    };
    return (React.createElement("div", tslib_1.__assign({ role: "presentation", className: classNames.root }, rest),
        React.createElement("h1", { className: classNames.title }, title),
        (subtitle || learnMoreLink) && (React.createElement("div", { className: classNames.subtitle },
            subtitle,
            learnMoreLink && (React.createElement(Link_1.Link, { className: classNames.learnMoreLink, href: learnMoreLink.url, target: learnMoreLink.target, styles: Error_styles_1.getDefaultLinkStyles(theme) }, learnMoreLink.label)))),
        seeDetailsLinkShowLabel && seeDetailsLinkHideLabel && errorDetails && (React.createElement(React.Fragment, null,
            React.createElement(Button_1.DefaultButton, tslib_1.__assign({ "aria-expanded": isContentVisible, onClick: toggleState, onMenuClick: toggleState }, seeDetailsButtonProps, { text: isContentVisible ? seeDetailsLinkHideLabel : seeDetailsLinkShowLabel, className: classNames.seeDetailsButton })),
            isContentVisible && (React.createElement("div", { className: classNames.seeDetailsDiv },
                renderLabeledField(errorDetails.messageLabel, errorDetails.message),
                renderLabeledField(errorDetails.timeStampLabel, errorDetails.timeStamp),
                renderLabeledField(errorDetails.sessionIdLabel, errorDetails.sessionId),
                renderLabeledField(errorDetails.objectIdLabel, errorDetails.objectId),
                renderLabeledField(errorDetails.tenantIdLabel, errorDetails.tenantId),
                errorDetails.customerSupportLink && (React.createElement(Link_1.Link, { className: classNames.customerSupportLink, href: errorDetails.customerSupportLink.url, target: errorDetails.customerSupportLink.target, styles: Error_styles_1.getDefaultLinkStyles(theme) },
                    errorDetails.customerSupportLink.label,
                    " ",
                    React.createElement(Icon_1.Icon, { iconName: "OpenInNewWindow" }))))))),
        callToActionLink && (React.createElement(Button_1.PrimaryButton, { text: callToActionLink.label, href: callToActionLink.url, target: callToActionLink.target, styles: headerStyles(), className: classNames.callToActionButton })),
        React.createElement("div", { className: classNames.imageSmall }, ErrorSplashSmall_1.getSmallSplashSvg()),
        React.createElement("div", { className: classNames.imageMedium }, ErrorSplashMedium_1.getMediumSplashSvg()),
        React.createElement("div", { className: classNames.imageLarge }, ErrorSplashLarge_1.getLargeSplashSvg()),
        React.createElement("div", { className: classNames.imageFooter })));
}
exports.ErrorBase = ErrorBase;
function renderLabeledField(fieldLabel, fieldText) {
    return (React.createElement(React.Fragment, null,
        fieldLabel && fieldText && (React.createElement("div", null,
            React.createElement("b", null,
                fieldLabel,
                ":"),
            " ",
            fieldText)),
        !fieldLabel && fieldText && React.createElement("div", null, fieldText)));
}
