// @business-app/fabric/lib/components/ColorInput/examples/ColorInput.Basic.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { ColorInput } from '..';
import { isValidColor } from '../ColorInput.base';
var ColorInputBasicExample = /** @class */ (function (_super) {
    __extends(ColorInputBasicExample, _super);
    function ColorInputBasicExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.getErrorMessage = function (value) {
            return !isValidColor(value) ? "Please enter a valid color value." : '';
        };
        return _this;
    }
    ColorInputBasicExample.prototype.render = function () {
        var _this = this;
        return (React.createElement(ColorInput, { label: "Color Input", tooltipIconProps: {
                iconProps: { iconName: 'Info', styles: { root: { color: '#666666' } } },
                ariaLabel: 'New Color Input'
            }, tooltipProps: { content: 'New Color Input' }, colorBoxProps: {
                onClick: function () {
                    return;
                },
                ariaLabel: 'Show color picker',
                title: 'Show color picker',
                disabled: false
            }, colorTextFieldProps: {
                onBlur: function () {
                    return;
                },
                ariaLabel: 'Enter color code',
                disabled: false,
                onGetErrorMessage: function (value) {
                    return _this.getErrorMessage(value);
                }
            }, colorPickerProps: {
                color: 'Red'
            } }));
    };
    return ColorInputBasicExample;
}(React.Component));
export { ColorInputBasicExample };
