// @business-app/fabric/lib-cjs/components/Wizard/Wizard.test.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var Enzyme = require("enzyme");
require("jest");
var React = require("react");
var StepStatus_1 = require("./StepStatus");
var Wizard_base_1 = require("./Wizard.base");
describe('Wizard', function () {
    it('renders the placeholders correctly', function () {
        var noop = jest.fn();
        var wrapper = Enzyme.shallow(React.createElement(Wizard_base_1.WizardBase, { onRenderStep: function () { return React.createElement("div", { className: "step-body" }); }, onRenderRightPane: function () { return React.createElement("div", { className: "right-pane" }); }, onRenderHeaderItems: function () { return React.createElement("div", { className: "header-items" }); }, onRenderHeaderFarItems: function () { return React.createElement("div", { className: "header-far-items" }); }, progressProps: {
                stepNames: ['Step1', 'Step2'],
                stepStatuses: [StepStatus_1.StepStatus.Valid, StepStatus_1.StepStatus.Empty],
                stepSummaries: [['Done'], []],
                currentIndex: 0,
                computeAriaDescription: function () { return 'description'; },
                expandButtonLabel: 'Expand timeline',
                collapseButtonLabel: 'Collapse timeline',
                onProgressStepClick: function (stepNo) { return null; }
            }, footerProps: {
                backButtonText: 'Back',
                continueButtonText: 'Next',
                canGoToPreviousStep: true,
                canGoToNextStep: true,
                shouldDisplayBackButton: true,
                goToNextStep: noop,
                goToPreviousStep: noop
            } }));
        expect(wrapper.find('.step-body').length).toBe(1);
        expect(wrapper.find('.right-pane').length).toBe(1);
        expect(wrapper.find('.header-items').length).toBe(1);
        expect(wrapper.find('.header-far-items').length).toBe(1);
    });
});
