// @business-app/fabric/src/components/Card/NotificationCard/NotificationCard.styles.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IStyleFunction } from '@fluentui/react/lib/Utilities';

import { NotificationCardClassNames, NotificationCardProps } from './NotificationCard.types';

export const styles: IStyleFunction<NotificationCardProps, NotificationCardClassNames> = ({
  theme,
  className,
  isNew,
  notificationType
}) => {
  return {
    root: [
      'baf-notification-card',
      {
        width: '100%',
        padding: '6px 12px 6px 0',
        marginBottom: 16,
        display: 'flex',
        flexDirection: 'row',
        minHeight: 34,
        selectors: {
          ':before': {
            width: isNew ? 4 : 0,
            content: "''",
            display: 'block',
            position: 'absolute',
            top: 0,
            left: 0,
            bottom: 0,
            backgroundColor: theme.palette.themePrimary,
            transition: 'all 250ms ease-in-out',
            borderTopLeftRadius: 3,
            borderBottomLeftRadius: 3
          }
        }
      },
      className
    ],
    timeDismissWrapper: [
      {
        display: 'flex',
        flexDirection: 'row-reverse',
        position: 'absolute',
        top: 6,
        right: 6
      }
    ],
    closeButton: [
      {
        color: theme.palette.neutralDark
      }
    ],
    imageWrapper: [
      {
        minWidth: 24,
        marginLeft: 12,
        marginRight: 12
      }
    ],
    icon: [
      {
        height: 32,
        fontSize: 24,
        textAlign: 'center',
        lineHeight: 32
      }
    ],
    persona: [
      {
        display: 'inline-block',
        selectors: {
          '.ms-Image': {
            borderRadius: notificationType === 'Alias' ? 6 : undefined
          }
        }
      }
    ],
    peerIcon: [
      {
        paddingBottom: 8,
        marginRight: 6,
        position: 'relative',
        top: 4,
        fontSize: 16
      }
    ],
    contentWrapper: [
      {
        width: '100%',
        paddingTop: 6,
        paddingBottom: 6
      }
    ],
    cardHeadingWrapper: [{}],
    cardHeading: [
      {
        display: 'block',
        fontSize: theme.fonts.medium.fontSize,
        fontWeight: 600
      }
    ],
    cardSubHeading: [
      {
        display: 'block',
        color: theme.palette.neutralDark,
        fontSize: theme.fonts.small.fontSize,
        fontWeight: 300
      }
    ],
    messageWrapper: [
      {
        display: 'block',
        maxHeight: 38,
        overflow: 'hidden',
        paddingInlineEnd: 12,
        textOverflow: 'ellipsis'
      }
    ],
    actionButtons: [
      {
        padding: '8px 12px 0 0'
      }
    ],
    timestamp: [
      {
        ...theme.fonts.smallPlus,
        paddingInlineEnd: 8,
        position: 'relative',
        top: 8,
        color: theme.palette.neutralSecondary
      }
    ]
  };
};
