// @business-app/fabric/lib-cjs/components/Wizard/TipsAndHelp/WizardTipsAndHelp.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
var getStyles = function (props) {
    var _a;
    var palette = props.theme.palette;
    var smallScreenSelector = '@media (max-width: 767px)';
    return {
        root: [
            'ba-wizard-tips',
            props.className,
            {
                boxSizing: 'border-box',
                flex: '0 0 auto',
                width: '21vw',
                minWidth: 180,
                maxWidth: 272,
                padding: '12px',
                background: palette.neutralLighterAlt,
                borderLeft: "1px solid " + palette.neutralLight,
                overflowY: 'auto',
                selectors: {
                    '@media (max-width: 839px)': {
                        position: 'absolute',
                        top: 0,
                        bottom: 0,
                        right: 0,
                        zIndex: 10,
                        width: '40vw',
                        maxWidth: 'none',
                        boxShadow: '0 4.8px 14.4px rgba(0, 0, 0, 0.08), 0 25.6px 57.6px rgba(0, 0, 0, 0.12)'
                    },
                    '@media (max-width: 639px)': {
                        width: '66vw'
                    }
                }
            }
        ],
        rootHidden: {
            display: 'none',
            selectors: {
                '@media (min-width: 840px)': {
                    display: 'block' /* always displayed on larger devices */
                }
            }
        },
        header: {
            display: 'flex',
            alignItems: 'flex-start',
            margin: '0 0 16px 0',
            padding: '4px 0 16px 0',
            borderBottom: "1px solid " + palette.neutralLight
        },
        icon: {
            fontSize: '16px',
            margin: '0 12px 0 8px',
            color: palette.themePrimary,
            userSelect: 'none'
        },
        show: {
            display: 'none',
            position: 'absolute',
            right: 18,
            bottom: 20,
            zIndex: 2,
            background: palette.neutralLighter,
            borderRadius: '50%',
            selectors: (_a = {
                    ':hover, :focus': {
                        backgroundColor: palette.neutralLight,
                        color: palette.themePrimary
                    },
                    '@media (max-width: 839px)': {
                        display: 'block'
                    }
                },
                _a[smallScreenSelector] = {
                    bottom: 12
                },
                _a)
        },
        close: {
            margin: '-6px -8px -6px auto',
            selectors: {
                '@media (min-width: 840px)': {
                    display: 'none'
                }
            }
        },
        title: {
            fontSize: Styling_1.FontSizes.mediumPlus,
            fontWeight: Styling_1.FontWeights.semibold
        }
    };
};
exports.getStyles = getStyles;
