// @business-app/fabric/lib/components/CommandBarWithSearchBox/CommandBarWithSearchBox.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __assign, __extends, __rest } from "tslib";
import { CommandBar as FabricCommandBar } from '@fluentui/react/lib/CommandBar';
import { SearchBox } from '@fluentui/react/lib/SearchBox';
import { getTheme, HighContrastSelector } from '@fluentui/react/lib/Styling';
import { BaseComponent } from '@fluentui/react/lib/Utilities';
import * as React from 'react';
/**
 * CommandBar wraps the Fabric CommandBar with customized styling.
 *
 * The CommandBar also has a SearchBox component defined to the right.
 *
 * @export
 * @class CommandBar
 * @extends {React.Component<ICommandBarWithSearchBoxProps>}
 */
var CommandBarWithSearchBox = /** @class */ (function (_super) {
    __extends(CommandBarWithSearchBox, _super);
    function CommandBarWithSearchBox(props) {
        var _a, _b;
        var _this = _super.call(this, props) || this;
        /**
         * Set this BaseComponent._skipComponentRefResolution to true, bypassing resolution of componentRef.
         */
        _this._skipComponentRefResolution = true;
        _this.theme = getTheme();
        _this.commandBarStyles = {
            root: {
                backgroundColor: _this.theme.palette.neutralLighter,
                selectors: (_a = {
                        '.ms-Button': {
                            backgroundColor: _this.theme.palette.neutralLighter,
                            selectors: (_b = {
                                    ':hover': {
                                        backgroundColor: _this.theme.palette.neutralQuaternaryAlt
                                    }
                                },
                                _b[HighContrastSelector] = {
                                    backgroundColor: 'Window'
                                },
                                _b)
                        }
                    },
                    _a[HighContrastSelector] = {
                        border: 'solid'
                    },
                    _a)
            }
        };
        _this.searchBoxStyles = {
            root: [
                'ms-SearchBox',
                {
                    height: '100%',
                    width: 168,
                    backgroundColor: 'transparent',
                    border: 0,
                    borderLeft: "1px solid " + _this.theme.palette.neutralLight,
                    selectors: {
                        ':hover': {
                            borderLeft: "1px solid " + _this.theme.palette.neutralLight
                        },
                        '.ms-CommandBar-sideCommands': {
                            padding: 0
                        },
                        '.ms-SearchBox-icon': {
                            color: _this.theme.palette.themeDarkAlt
                        },
                        '.ms-SearchBox-field': {
                            backgroundColor: 'transparent'
                        }
                    }
                }
            ]
        };
        return _this;
    }
    CommandBarWithSearchBox.prototype.render = function () {
        var _this = this;
        var _a = this.props, isSearchBoxVisible = _a.isSearchBoxVisible, searchBoxProps = _a.searchBoxProps, farItems = _a.farItems, wrapperClassName = _a.wrapperClassName, onChangeDelay = _a.onChangeDelay, other = __rest(_a, ["isSearchBoxVisible", "searchBoxProps", "farItems", "wrapperClassName", "onChangeDelay"]);
        if (isSearchBoxVisible) {
            if (!farItems) {
                farItems = [];
            }
            searchBoxProps = __assign(__assign({}, searchBoxProps), { styles: function (props) {
                    return _this.searchBoxStyles;
                } });
            if (searchBoxProps.onChange && onChangeDelay) {
                this._asyncFunctionToRun = this._async.debounce(searchBoxProps.onChange, onChangeDelay);
                searchBoxProps = __assign(__assign({}, searchBoxProps), { onChange: this._asyncFunctionToRun });
            }
            farItems = farItems.concat({
                key: 'ba-CommandBar-SearchBox',
                onRender: function (item) {
                    return React.createElement(SearchBox, __assign({}, searchBoxProps));
                }
            });
        }
        var className = 'ba-CommandBar ';
        className += wrapperClassName ? wrapperClassName : '';
        return (React.createElement("div", { className: className },
            React.createElement(FabricCommandBar, __assign({}, other, { farItems: farItems, styles: this.commandBarStyles }))));
    };
    CommandBarWithSearchBox.prototype.componentWillUnmount = function () {
        this._async.dispose();
    };
    return CommandBarWithSearchBox;
}(BaseComponent));
export { CommandBarWithSearchBox };
