// @business-app/fabric/src/utilities/themes.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

/**
 * Common location for CAP products to define themes. So, cross product references has consistent user experience.
 */

import { IPalette, IPartialTheme } from '@fluentui/react/lib/Styling';

export const DefaultGrayscalePalette: Partial<IPalette> = {
  white: '#fff', // White
  neutralLighterAlt: '#f8f8f8', // Gray 10
  neutralLighter: '#f4f4f4', // Gray 20
  neutralLight: '#eaeaea', // Gray 30
  neutralQuaternaryAlt: '#dadada', // Gray 40
  neutralQuaternary: '#d0d0d0', //  Gray 50
  neutralTertiaryAlt: '#c8c8c8', //  Gray 60
  neutralTertiary: '#bab8b7', //  Gray 70
  neutralSecondary: '#a3a2a0', //  Gray 90
  neutralPrimaryAlt: '#8d8b8a', //  Gray 110
  neutralDark: '#605e5d', // Gray 130
  neutralPrimary: '#323130', // Gray 160
  black: '#000' // Black
};

export const PowerAppsGrayscalePalette: Partial<IPalette> = {
  white: '#fff', // White
  neutralLighterAlt: '#FAF9F8', // Gray 10
  neutralLighter: '#F3F2F1', // Gray 20
  neutralLight: '#EDEBE9', // Gray 30
  neutralQuaternaryAlt: '#E1DFDD', // Gray 40
  neutralQuaternary: '#D2D0CE', //  Gray 50
  neutralTertiaryAlt: '#C8C6C4', //  Gray 60
  neutralTertiary: '#BEBBB8', //  Gray 70
  neutralSecondary: '#A19F9D', //  Gray 90
  neutralPrimaryAlt: '#8A8886', //  Gray 110
  neutralDark: '#605E5C', // Gray 130
  neutralPrimary: '#323130', // Gray 160
  black: '#000' // Black
};

export const PowerAppsGrayscaleDarkPalette: Partial<IPalette> = {
  white: '#1b1a19', // White
  neutralLighterAlt: '#292828', // Gray 10
  neutralLighter: 'rgba(255, 255, 255, 0.05)', // Gray 20
  neutralLight: '#2d2c2c', // Gray 30
  neutralQuaternaryAlt: '#3b3a3a', // Gray 40
  neutralQuaternary: '#3b3a3a', //  Gray 50
  neutralPrimaryAlt: '#f3f2f1', //  Gray 110
  neutralDark: '#c8c6c4', // Gray 130
  neutralPrimary: '#f3f2f1', // Gray 160
  black: '#fff' // Black
};

export const PowerAppsGrayscaleContrastPalette: Partial<IPalette> = {
  white: '#000', // White
  neutralLighterAlt: '#000', // Gray 10
  neutralLighter: 'fff', // Gray 20
  neutralLight: '#000', // Gray 30
  neutralQuaternaryAlt: '#000', // Gray 40
  neutralQuaternary: '#000', //  Gray 50
  neutralPrimaryAlt: '#fff', //  Gray 110
  neutralDark: '#fff', // Gray 130
  neutralPrimary: '#fff', // Gray 160
  black: '#fff' // Black
};

export const PowerAutomateGrayscalePalette: Partial<IPalette> = {
  black: '#000', // Black
  neutralDark: '#201f1e', // Gray 190
  neutralPrimary: '#323130', // Gray 160
  neutralPrimaryAlt: '#3b3a39', // Gray 150
  neutralSecondary: '#605e5c', //  Gray 130
  neutralSecondaryAlt: '#8a8886', //  Gray 110
  neutralTertiary: '#a19f9d', //  Gray 90
  neutralTertiaryAlt: '#c8c6c4', //  Gray 60
  neutralQuaternary: '#d2d0ce', //  Gray 50
  neutralQuaternaryAlt: '#e1dfdd', // Gray 40
  neutralLight: '#edebe9', // Gray 30
  neutralLighter: '#f3f2f1', // Gray 20
  neutralLighterAlt: '#faf9f8', // Gray 10
  white: '#fff' // White
};

export const PowerVirtualAgentGrayscalePalette: Partial<IPalette> = {
  black: '#000', // Black
  neutralDark: '#201f1e', // Gray 190
  neutralPrimary: '#323130', // Gray 160
  neutralPrimaryAlt: '#3b3a39', // Gray 150
  neutralSecondary: '#605e5c', //  Gray 130
  neutralSecondaryAlt: '#8a8886', //  Gray 110
  neutralTertiary: '#a19f9d', //  Gray 90
  neutralTertiaryAlt: '#c8c6c4', //  Gray 60
  neutralQuaternary: '#d2d0ce', //  Gray 50
  neutralQuaternaryAlt: '#e1dfdd', // Gray 40
  neutralLight: '#edebe9', // Gray 30
  neutralLighter: '#f3f2f1', // Gray 20
  neutralLighterAlt: '#faf9f8', // Gray 10
  white: '#fff' // White
};

export const PowerVirtualAgentGrayscaleDarkPalette: Partial<IPalette> = {
  black: '#fff', // Black
  neutralDark: '#faf9f8', // Gray 190
  neutralPrimary: '#f3f2f1', // Gray 160
  neutralPrimaryAlt: '#c8c6c4', // Gray 150
  neutralSecondary: '#a19f9d', //  Gray 130
  neutralSecondaryAlt: '#979693', //  Gray 110
  neutralTertiary: '#797775', //  Gray 90
  neutralTertiaryAlt: '#484644', //  Gray 60
  neutralQuaternary: '#3b3a39', //  Gray 50
  neutralQuaternaryAlt: '#323130', // Gray 40
  neutralLight: '#292827', // Gray 30
  neutralLighter: '#252423', // Gray 20
  neutralLighterAlt: '#201f1e', // Gray 10
  white: '#1b1a19' // White
};

export const PowerVirtualAgentGrayscaleContrastPalette: Partial<IPalette> = {
  black: '#ffffff',
  neutralDark: '#ffffff',
  neutralPrimary: '#ffffff',
  neutralPrimaryAlt: '#ffffff',
  neutralSecondary: '#ffffff',
  neutralSecondaryAlt: '#ffffff',
  neutralTertiary: '#2ef149',
  neutralTertiaryAlt: '#ffffff',
  neutralQuaternary: '#ffffff',
  neutralQuaternaryAlt: '#ffffff',
  neutralLight: '#ffffff',
  neutralLighter: '#000000',
  neutralLighterAlt: '#000000',
  white: '#000000'
};

export const DefaultTheme: IPartialTheme = {
  palette: {
    themePrimary: '#0078d4',
    themeLighterAlt: '#f3f9fd',
    themeLighter: '#d0e7f8',
    themeLight: '#a9d3f2',
    themeTertiary: '#5ca9e5',
    themeSecondary: '#1a86d9',
    themeDarkAlt: '#006cbe',
    themeDark: '#005ba1',
    themeDarker: '#004377',
    ...DefaultGrayscalePalette
  }
};

const DefaultDarkPalette: Partial<IPalette> = {
  themePrimary: '#2899f5',
  themeSecondary: '#0078d4',
  themeTertiary: '#235a85',
  themeLight: '#004c87',
  themeLighter: '#043862',
  themeLighterAlt: '#092c47',
  themeDarker: '#82c7ff',
  themeDark: '#6cb8f6',
  themeDarkAlt: '#3aa0f3',
  neutralDark: '#faf9f8',
  neutralPrimary: '#f3f2f1',
  neutralPrimaryAlt: '#c8c6c4',
  neutralSecondary: '#a19f9d',
  neutralSecondaryAlt: '#979693',
  neutralTertiary: '#797775',
  neutralTertiaryAlt: '#484644',
  neutralQuaternary: '#3b3a39',
  neutralQuaternaryAlt: '#323130',
  neutralLight: '#292827',
  neutralLighter: '#252423',
  neutralLighterAlt: '#201f1e',
  black: '#ffffff',
  white: '#1b1a19',
  redDark: '#F1707B'
};

export const DefaultDarkTheme: IPartialTheme = {
  palette: DefaultDarkPalette,
  semanticColors: {
    buttonText: DefaultDarkPalette.black,
    buttonTextPressed: DefaultDarkPalette.neutralDark,
    buttonTextHovered: DefaultDarkPalette.neutralPrimary,
    bodySubtext: DefaultDarkPalette.white,
    disabledBackground: DefaultDarkPalette.neutralQuaternaryAlt,
    inputBackgroundChecked: DefaultDarkPalette.themePrimary,
    menuBackground: DefaultDarkPalette.neutralLighter,
    menuItemBackgroundHovered: DefaultDarkPalette.neutralQuaternaryAlt,
    menuItemBackgroundPressed: DefaultDarkPalette.neutralQuaternary,
    menuDivider: DefaultDarkPalette.neutralTertiaryAlt,
    menuIcon: DefaultDarkPalette.themeDarkAlt,
    menuHeader: DefaultDarkPalette.black,
    menuItemText: DefaultDarkPalette.neutralPrimary,
    menuItemTextHovered: DefaultDarkPalette.neutralDark
  }
};

export const DesignDocsTheme: IPartialTheme = {
  palette: {
    themePrimary: '#9a4bd8',
    themeSecondary: '#a65fdd',
    themeLighterAlt: '#fbf7fd',
    themeLighter: '#eedff9',
    themeLight: '#dfc4f4',
    themeTertiary: '#c18ee8',
    themeDarkAlt: '#8c44c3',
    themeDark: '#763aa5',
    themeDarker: '#572a79'
  },
  semanticColors: {
    errorBackground: '#fde7e9',
    blockingBackground: '#fed9cc',
    warningBackground: '#fff4ce',
    successBackground: '#dff6dd',
    link: '#a65fdd',
    linkHovered: '#763aa5',
    messageLink: '#a65fdd',
    messageLinkHovered: '#763aa5'
  }
};

export const PowerAppsTheme: IPartialTheme = {
  palette: {
    themePrimary: '#742774',
    themeSecondary: '#8d308d',
    themeTertiary: '#d98fd9',
    themeLighterAlt: '#faf2fa',
    themeLighter: '#f6e4f6',
    themeLight: '#edcaed',
    themeDarkAlt: '#672367',
    themeDark: '#501b50',
    themeDarker: '#3f153f',
    neutralTertiaryAlt: '#666',
    neutralTertiary: '#666',
    ...PowerAppsGrayscalePalette
  },
  semanticColors: {
    link: '#8D308D',
    linkHovered: '#501b50',
    messageLink: '#8d308d',
    messageLinkHovered: '#501b50'
  }
};

export const PowerAppsDarkTheme: IPartialTheme = {
  palette: {
    themePrimary: '#742774',
    themeSecondary: '#8d308d',
    ...PowerAppsGrayscaleDarkPalette
  }
};

export const PowerAppsContrastTheme: IPartialTheme = {
  palette: {
    themePrimary: '#742774',
    themeSecondary: '#8d308d',
    ...PowerAppsGrayscaleContrastPalette
  }
};

export const PowerBITheme: IPartialTheme = {
  palette: {
    themePrimary: '#f2c811',
    themeSecondary: '#f4cf2c',
    themeTertiary: '#f7de6d',
    themeLighterAlt: '#fefdf5',
    themeLighter: '#fdf6d7',
    themeLight: '#fbeeb5',
    themeDarkAlt: '#dab50f',
    themeDark: '#a98c0c',
    themeDarker: '#7a6816',
    ...PowerAppsGrayscalePalette,
    neutralTertiaryAlt: '#666',
    neutralTertiary: '#666'
  },
  semanticColors: {
    link: '#f4cf2c',
    linkHovered: '#a98c0c',
    messageLink: '#f4cf2c',
    messageLinkHovered: '#a98c0c'
  }
};

export const PowerAutomateTheme: IPartialTheme = {
  palette: {
    themePrimary: '#0066ff',
    themeSecondary: '#1975ff',
    themeTertiary: '#94bfff',
    themeLight: '#cce0ff',
    themeLighter: '#e6f0ff',
    themeLighterAlt: '#f2f7ff',
    themeDarker: '#00388c',
    themeDark: '#0047b3',
    themeDarkAlt: '#005ce6',
    ...PowerAppsGrayscalePalette,
    neutralTertiaryAlt: '#666',
    neutralTertiary: '#666'
  },
  semanticColors: {
    link: '#1975ff',
    linkHovered: '#0047b3',
    messageLink: '#1975ff',
    messageLinkHovered: '#0047b3'
  }
};

export const PowerVirtualAgentTheme: IPartialTheme = ((): IPartialTheme => {
  const palette = {
    themeDarker: '#062d37', // Shade 30
    themeDark: '#073845', // Shade 20
    themeDarkAlt: '#0a4a5c', // Shade 10
    themePrimary: '#0b556a', // Primary
    themeSecondary: '#0f677b', // Secondary Tint 10
    themeTertiary: '#14848f', // Tertiary
    themeLight: '#73e3df', // Tint 30
    themeLighter: '#cff6f9', // Tint 40
    themeLighterAlt: '#f1fbfe', // Tint 50
    purple: '#672367', // Purple
    blue: '#118DFF', // Blue
    green: '#6BB700', // Green
    magenta: '#D98FD9', // Magenta
    teal: '#73E3DF', // Teal
    purpleLight: '#F0E9F0', // Purple Light
    blueLight: '#E7F4FF', // Blue Light
    greenLight: '#F0F8E6', // Green light
    ...PowerVirtualAgentGrayscalePalette
  };
  return {
    palette,
    semanticColors: {
      link: '#10738e',
      linkHovered: palette.themeDark,
      messageLink: '#10738e',
      messageLinkHovered: palette.themeDark,
      primaryButtonBackground: palette.themePrimary,
      primaryButtonText: palette.white,
      primaryButtonBorder: palette.white,
      primaryButtonBackgroundHovered: palette.themeDarkAlt,
      primaryButtonTextHovered: palette.white,
      primaryButtonBackgroundPressed: '#edebe9',
      primaryButtonTextPressed: palette.white,
      focusBorder: palette.black
    }
  };
})();

export const PowerVirtualAgentDarkTheme: IPartialTheme = ((): IPartialTheme => {
  const palette = {
    themeDarker: '#69ebed', // Shade 30
    themeDark: '#64e0e2', // Shade 20
    themeDarkAlt: '#34d6d8', // Shade 10
    themePrimary: '#00c9cc', // Primary
    themeSecondary: '#00b1b4', // Secondary Tint 10
    themeTertiary: '#00787a', // Tertiary
    themeLight: '#003c3d', // Tint 30
    themeLighter: '#002021', // Tint 40
    themeLighterAlt: '#000808', // Tint 50
    purple: '#D25FD2', // Purple
    blue: '#2899F5', // Blue
    green: '#92C353', // Green
    magenta: '#7C2E7C', // Magenta
    teal: '#00787A', // Teal
    purpleLight: '#3D2A3B', // Purple Light
    blueLight: '#293F50', // Blue Light
    greenLight: '#3E4730', // Green light
    ...PowerVirtualAgentGrayscaleDarkPalette
  };
  return {
    palette,
    semanticColors: {
      link: palette.themePrimary,
      linkHovered: palette.themeDark,
      bodyBackgroundChecked: palette.neutralQuaternaryAlt,
      bodyBackground: palette.white,
      bodyBackgroundHovered: palette.neutralLight,
      bodyTextChecked: palette.neutralPrimary,
      bodyText: palette.black,
      primaryButtonBackground: palette.themePrimary,
      primaryButtonText: palette.white,
      primaryButtonBorder: palette.white,
      primaryButtonBackgroundHovered: palette.themeDarkAlt,
      primaryButtonTextHovered: palette.white,
      primaryButtonBackgroundPressed: palette.themeDark,
      primaryButtonTextPressed: palette.white,
      primaryButtonBackgroundDisabled: palette.neutralLighterAlt,
      primaryButtonTextDisabled: palette.neutralTertiary,
      focusBorder: palette.black,
      inputBackground: palette.white,
      inputText: palette.neutralPrimary,
      inputPlaceholderText: palette.black,
      inputFocusBorderAlt: palette.themePrimary,
      warningIcon: palette.neutralPrimaryAlt,
      warningBackground: '#433519',
      severeWarningIcon: '#FCE100',
      severeWarningBackground: '#4F2A0F',
      errorText: '#F1707B',
      errorIcon: '#F1707B',
      errorBackground: '#442726',
      successIcon: palette.green,
      successBackground: '#393D1B'
    }
  };
})();

export const PowerVirtualAgentContrastTheme: IPartialTheme = ((): IPartialTheme => {
  const palette = {
    themeDarker: '#000000', // Shade 30
    themeDark: '#000000', // Shade 20
    themeDarkAlt: '#000000', // Shade 10
    themePrimary: '#000000', // Primary
    themeSecondary: '#000000', // Secondary Tint 10
    themeTertiary: '#000000', // Tertiary
    themeLight: '#000000', // Tint 30
    themeLighter: '#000000', // Tint 40
    themeLighterAlt: '#000000', // Tint 50
    purple: '#000000', // Purple
    blue: '#000000', // Blue
    green: '#000000', // Green
    magenta: '#000000', // Magenta
    teal: '#000000', // Teal
    purpleLight: '#000000', // Purple Light
    blueLight: '#000000', // Blue Light
    greenLight: '#000000', // Green light
    ...PowerVirtualAgentGrayscaleContrastPalette
  };

  const yellow = '#fffd39';
  const green = '#2ef149';
  const cyan = '#00ebfd';
  return {
    palette,
    semanticColors: {
      link: yellow,
      linkHovered: yellow,
      bodyBackgroundChecked: cyan,
      bodyBackground: palette.white,
      bodyBackgroundHovered: cyan,
      bodyTextChecked: palette.white,
      bodyText: palette.black,
      primaryButtonBackground: palette.black,
      primaryButtonText: palette.white,
      primaryButtonBorder: palette.black,
      primaryButtonBackgroundHovered: cyan,
      primaryButtonTextHovered: palette.white,
      primaryButtonBackgroundPressed: cyan,
      primaryButtonTextPressed: palette.white,
      primaryButtonBackgroundDisabled: green,
      primaryButtonTextDisabled: green,
      focusBorder: palette.black,
      inputBackground: palette.white,
      inputText: palette.black,
      inputPlaceholderText: palette.black,
      inputFocusBorderAlt: palette.black,
      disabledText: green,
      errorText: palette.black,
      errorIcon: palette.black,
      successIcon: palette.black,
      severeWarningIcon: palette.black,
      successBackground: palette.white,
      severeWarningBackground: palette.white,
      warningBackground: palette.white,
      errorBackground: palette.white
    }
  };
})();

export const Dynamics365Theme: IPartialTheme = {
  palette: {
    themePrimary: '#2266e3', // Primary
    themeSecondary: '#1f6cf9', // tint 10
    themeTertiary: '#9cbeff', // tint 20
    themeLight: '#ccdeff', // Tint 30
    themeLighter: '#e7efff', // Tint 40
    themeLighterAlt: '#f0f5ff', // tint 50
    themeDarker: '#053385', // Shade 30
    themeDark: '#0742ab', // Shade 20
    themeDarkAlt: '#0b53ce', // Shade 10
    ...PowerAppsGrayscalePalette,
    neutralTertiaryAlt: '#666',
    neutralTertiary: '#666'
  },
  semanticColors: {
    link: '#1f6cf9',
    linkHovered: '#0742ab',
    messageLink: '#1f6cf9',
    messageLinkHovered: '#0742ab'
  }
};

export const FlowTheme: IPartialTheme = {
  palette: {
    themePrimary: '#0066ff',
    themeSecondary: '#1975ff',
    themeLighterAlt: '#f2f7ff',
    themeLighter: '#e6f0ff',
    themeLight: '#cce0ff',
    themeTertiary: '#94bfff',
    themeDarkAlt: '#005ce6',
    ...PowerAutomateGrayscalePalette,
    themeDark: '#0047b3',
    themeDarker: '#00388c'
  }
};

// export new palettes with old export to avoid breaking partner teams who have used these values
export const PowerAppsPurplePalette: Partial<IPalette> = PowerAppsTheme.palette;
export const FlowBluePalette: Partial<IPalette> = FlowTheme.palette;
