// @business-app/fabric/lib-cjs/components/Form/inputs/textInput/FormTextInput.test.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/* eslint-disable @typescript-eslint/no-explicit-any */
var React = require("react");
var ReactDom = require("react-dom");
var ReactTestUtils = require("react-dom/test-utils");
// Controls
var Enzyme = require("enzyme");
var TextField_1 = require("@fluentui/react/lib/TextField");
var Form_1 = require("../../Form");
var FormTextInput_1 = require("./FormTextInput");
// Utilities
var Validators = require("../../validators/Validators");
var FormBaseInput_1 = require("../../FormBaseInput");
describe('FormTextInput Unit Tests', function () {
    describe('Renders for all combinations of props', function () {
        var formProps;
        var formTextInputProps;
        beforeEach(function () {
            formProps = {
                onSubmit: undefined
            };
            formTextInputProps = {
                inputKey: null,
                value: undefined
            };
        });
        it('Null name throws error', function () {
            var consoleMock = jest.spyOn(console, 'error');
            consoleMock.mockImplementation(function () { return undefined; });
            var errorFunction = function () {
                Enzyme.mount(React.createElement(Form_1.Form, tslib_1.__assign({}, formProps),
                    React.createElement(FormTextInput_1.FormTextInput, tslib_1.__assign({}, formTextInputProps))));
            };
            expect(errorFunction).toThrow();
            expect(consoleMock.mock.calls[0][0]).toMatch('Uncaught [Error: FormBaseInput: name must be defined on all form inputs]');
            consoleMock.mockRestore();
        });
        it('Null props still render', function () {
            formTextInputProps.inputKey = 'name';
            var wrapper = Enzyme.mount(React.createElement(Form_1.Form, tslib_1.__assign({}, formProps),
                React.createElement(FormTextInput_1.FormTextInput, tslib_1.__assign({}, formTextInputProps))));
            expect(wrapper.find(TextField_1.TextFieldBase).length).toBe(1);
        });
        it('With initial value', function () {
            formTextInputProps.inputKey = 'name';
            formTextInputProps.value = 'Value';
            var wrapper = Enzyme.mount(React.createElement(Form_1.Form, tslib_1.__assign({}, formProps),
                React.createElement(FormTextInput_1.FormTextInput, tslib_1.__assign({}, formTextInputProps))));
            expect(wrapper.find(TextField_1.TextFieldBase).length).toBe(1);
        });
    });
    describe('Common validations', function () {
        xit('Validators run properly', function () {
            var result;
            var formRef = React.createRef();
            ReactTestUtils.renderIntoDocument(React.createElement(Form_1.Form, { ref: formRef, onSubmit: function (value) {
                    result = value;
                } },
                React.createElement(FormTextInput_1.FormTextInput, { inputKey: "field", value: "", validators: [Validators.required('Message')] })));
            var form = ReactTestUtils.findRenderedDOMComponentWithTag(formRef.current, 'form');
            ReactTestUtils.Simulate.submit(form);
            // Find the TextField component
            // TODO: this test is not working as intended as even nonsense state names pass.
            //        commented out for now since test isn't executing when named 'xit'
            // const field = ReactTestUtils.findRenderedComponentWithType(renderedForm, TextField);
            // expect(field.state.errorMessage).toBeTruthy();
            expect(result).toBeFalsy();
        });
    });
    describe('Textbox update tests', function () {
        jest.useFakeTimers();
        var ExtendsTextInput = /** @class */ (function (_super) {
            tslib_1.__extends(ExtendsTextInput, _super);
            function ExtendsTextInput() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            ExtendsTextInput.prototype.setValue = function (value) {
                _super.prototype.setValue.call(this, value);
            };
            return ExtendsTextInput;
        }(FormTextInput_1.FormTextInput));
        // Disabled because of intermittent issues with timer logic.
        it.skip('TextInput is only trailing debounced', function (done) {
            var updateStub = jest.fn();
            var formRef = React.createRef();
            ReactTestUtils.renderIntoDocument(React.createElement(Form_1.Form, { ref: formRef, onUpdated: updateStub },
                React.createElement(ExtendsTextInput, { inputKey: "name", value: '' })));
            var datePicker = ReactTestUtils.findRenderedComponentWithType(formRef.current, ExtendsTextInput);
            datePicker.setValue('0');
            expect(updateStub).not.toHaveBeenCalled();
            datePicker.setValue('1');
            expect(updateStub).not.toHaveBeenCalled();
            jest.runAllTimers();
            expect(updateStub).toHaveBeenCalled();
            done();
        });
        it('TextInput state updates from props value change', function () {
            var updatedValue = 'updated';
            var formRef = React.createRef();
            var parent = document.createElement('div');
            ReactDom.render(React.createElement(Form_1.Form, { ref: formRef },
                React.createElement(FormTextInput_1.FormTextInput, { inputKey: "name", key: "key", value: 'old value' })), parent);
            var textboxElement = ReactTestUtils.findRenderedComponentWithType(formRef.current, FormTextInput_1.FormTextInput);
            var propsUpdateSpy = jest.spyOn(FormBaseInput_1.FormBaseInput, 'getDerivedStateFromProps');
            ReactDom.render(React.createElement(Form_1.Form, null,
                React.createElement(FormTextInput_1.FormTextInput, { inputKey: "name", key: "key", value: updatedValue })), parent);
            expect(textboxElement.state.currentValue).toEqual(updatedValue);
            expect(propsUpdateSpy).toHaveBeenCalledTimes(1);
        });
        it("TextInput state doesn't update with no new props value", function () {
            var oldValue = 'old value';
            var formRef = React.createRef();
            var parent = document.createElement('div');
            ReactDom.render(React.createElement(Form_1.Form, { ref: formRef },
                React.createElement(FormTextInput_1.FormTextInput, { inputKey: "name", key: "key", value: oldValue })), parent);
            var textboxElement = ReactTestUtils.findRenderedComponentWithType(formRef.current, FormTextInput_1.FormTextInput);
            var setStateSpy = jest.spyOn(textboxElement, 'setState');
            ReactDom.render(React.createElement(Form_1.Form, null,
                React.createElement(FormTextInput_1.FormTextInput, { inputKey: "name", key: "key", value: oldValue })), parent);
            expect(setStateSpy).toHaveBeenCalledTimes(0);
        });
    });
});
