// @business-app/fabric/lib-cjs/components/Card/RichCard/RichCard.base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RichCardBase = void 0;
var tslib_1 = require("tslib");
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
var React = require("react");
var Utilities_1 = require("@fluentui/react/lib/Utilities");
var Card_1 = require("./../Card");
var CardFooter_1 = require("./../CardFooter");
var CardHeader_1 = require("./../CardHeader");
var getClassNames = Utilities_1.classNamesFunction();
var RichCardBase = /** @class */ (function (_super) {
    tslib_1.__extends(RichCardBase, _super);
    function RichCardBase(props) {
        var _this = _super.call(this, props) || this;
        var _a = _this.props, headerProps = _a.headerProps, footerProps = _a.footerProps, renderFooter = _a.renderFooter;
        _this.state = {
            hasHeader: !!headerProps,
            hasFooter: !!(footerProps || renderFooter)
        };
        return _this;
    }
    RichCardBase.prototype.render = function () {
        var _a = this.props, width = _a.width, height = _a.height, children = _a.children, className = _a.className, headerProps = _a.headerProps, footerProps = _a.footerProps, renderContentArea = _a.renderContentArea, renderFooter = _a.renderFooter, renderHeader = _a.renderHeader, styles = _a.styles, theme = _a.theme;
        var _b = this.state, hasHeader = _b.hasHeader, hasFooter = _b.hasFooter;
        var classNames = getClassNames(styles, {
            theme: theme,
            className: className,
            hasHeader: hasHeader,
            hasFooter: hasFooter
        });
        return (React.createElement(Card_1.Card, { className: classNames.root, height: height, width: width },
            (headerProps || renderHeader) && React.createElement(CardHeader_1.CardHeader, tslib_1.__assign({}, headerProps), renderHeader === null || renderHeader === void 0 ? void 0 : renderHeader()),
            React.createElement("div", { className: classNames.contentArea }, renderContentArea ? renderContentArea() : children),
            (footerProps || renderFooter) && React.createElement(CardFooter_1.CardFooter, tslib_1.__assign({}, footerProps), renderFooter === null || renderFooter === void 0 ? void 0 : renderFooter())));
    };
    return RichCardBase;
}(React.Component));
exports.RichCardBase = RichCardBase;
