// @business-app/fabric/lib-cjs/components/Accordion/examples/Accordion.Basic.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccordionBasicExample = void 0;
var tslib_1 = require("tslib");
var Button_1 = require("@fluentui/react/lib/Button");
var React = require("react");
var __1 = require("..");
var AccordionBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(AccordionBasicExample, _super);
    function AccordionBasicExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AccordionBasicExample.prototype.render = function () {
        var exampleStyles = {
            divStyle: {
                marginBottom: '20px',
                width: 220
            },
            buttonStyle: {
                root: {
                    width: '100%',
                    minWidth: 0,
                    background: 'transparent',
                    padding: '12px 0',
                    textAlign: 'left'
                },
                icon: { fontSize: 22, padding: '0' },
                menuIcon: { fontSize: 20, padding: '0 8px' }
            }
        };
        return (React.createElement("div", null,
            React.createElement("div", { style: exampleStyles.divStyle },
                React.createElement(__1.Accordion, { text: "Stuff here", iconProps: { iconName: 'CollapseMenu' }, onRenderContent: function () {
                        return (React.createElement("div", null,
                            React.createElement(Button_1.DefaultButton, { key: "item1", text: "Link 1", iconProps: { iconName: 'CompassNW' }, ariaLabel: "Link 1", onClick: function () {
                                    return;
                                }, styles: exampleStyles.buttonStyle }),
                            React.createElement(Button_1.DefaultButton, { key: "item2", text: "Link 2", iconProps: { iconName: 'BuildQueue' }, onClick: function () {
                                    return;
                                }, styles: exampleStyles.buttonStyle }),
                            React.createElement(Button_1.DefaultButton, { key: "item3", text: "Link 3", iconProps: { iconName: 'AutoRacing' }, onClick: function () {
                                    return;
                                }, styles: exampleStyles.buttonStyle })));
                    } })),
            React.createElement("div", { style: exampleStyles.divStyle },
                React.createElement(__1.Accordion, { text: "isHiddenOnDismiss set to true", iconProps: { iconName: 'CollapseMenu' }, isHiddenOnDismiss: true, onRenderContent: function () {
                        return (React.createElement("div", null,
                            React.createElement(Button_1.DefaultButton, { key: "item1", text: "Link 1", iconProps: { iconName: 'CompassNW' }, ariaLabel: "Link 1", onClick: function () {
                                    return;
                                }, styles: exampleStyles.buttonStyle })));
                    } }))));
    };
    return AccordionBasicExample;
}(React.Component));
exports.AccordionBasicExample = AccordionBasicExample;
