// @business-app/fabric/src/components/Wizard/Wizard.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { getScreenSelector, ScreenWidthMaxSmall } from '@fluentui/react/lib/Styling';
import { WizardStyleProps, WizardStyles } from './Wizard.types';
export const getStyles = (props: WizardStyleProps): WizardStyles => {
  const { palette } = props.theme;
  const smallScreenSelector = '@media (max-width: 767px)';
  const smallOnlySelector = getScreenSelector(0, ScreenWidthMaxSmall);
  return {
    root: [
      'ba-wizard',
      props.className,
      {
        display: 'flex',
        flex: '1 1 auto',
        flexDirection: 'row',
        overflow: 'hidden',
        background: palette.white,
        selectors: {
          [smallScreenSelector]: {
            flexDirection: 'column'
          }
        }
      }
    ],
    bodyWrapper: {
      display: 'flex',
      flex: '1 1 auto',
      flexDirection: 'column',
      overflow: 'hidden'
    },
    header: {
      position: 'relative',
      display: 'flex',
      flex: '0 0 40px',
      justifyContent: 'space-between',
      alignItems: 'center',
      background: palette.white,
      zIndex: 1,
      selectors: {
        '::after': {
          content: '""',
          position: 'absolute',
          top: '100%',
          right: 0,
          left: 0,
          borderBottom: `1px solid ${palette.neutralLight}`
        },
        '.ms-Button-textContainer': {
          whiteSpace: 'nowrap'
        },
        [smallScreenSelector]: {
          background: palette.neutralLighterAlt,
          selectors: {
            '.ms-Button-textContainer': {
              display: 'none'
            }
          }
        }
      }
    },
    headerItems: {
      display: 'flex',
      alignItems: 'center',
      flex: '1 1 auto'
    },
    headerFarItems: {
      display: 'flex',
      alignItems: 'center',
      overflow: 'hidden',
      selectors: {
        '> .ms-TooltipHost': {
          whiteSpace: 'noWrap',
          overflow: 'hidden',
          textOverflow: 'ellipsis'
        }
      }
    },
    headerFarItem: {
      paddingLeft: '12px',
      selectors: {
        ':not(:last-child)': {
          borderRight: `1px solid ${palette.neutralTertiaryAlt}`,
          paddingRight: '12px'
        },
        '& .ms-Button-icon': {
          marginLeft: 0
        }
      }
    },
    section: {
      display: 'flex',
      flex: '1 1 auto',
      overflowY: 'hidden'
    },
    stepContainer: {
      flex: '1 1 auto',
      display: 'flex',
      flexDirection: 'column',
      position: 'relative',
      overflowY: 'hidden'
    },
    stepBody: {
      display: 'flex',
      flexDirection: 'column',
      flex: '1 1 auto',
      overflowY: 'hidden',
      selectors: {
        [smallOnlySelector]: {
          flex: '1 0 25px'
        }
      }
    }
  };
};
