// @business-app/fabric/lib-cjs/components/Resizer/examples/Resizer.Basic.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResizerBasicExample = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var Resizer_1 = require("../Resizer");
var Button_1 = require("@fluentui/react/lib/Button");
var Resizer_types_1 = require("../Resizer.types");
function ResizerBasicExample() {
    return React.createElement(ResizableButton, { initialSize: 350 });
}
exports.ResizerBasicExample = ResizerBasicExample;
var hiddenAccessibleDivStyles = {
    overflow: 'hidden',
    position: 'absolute',
    top: -1000,
    left: -1000,
    width: 0,
    height: 0
};
var ResizableButton = /** @class */ (function (_super) {
    tslib_1.__extends(ResizableButton, _super);
    function ResizableButton(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            width: props.initialSize,
            height: 148
        };
        return _this;
    }
    ResizableButton.prototype.render = function () {
        var _this = this;
        return (React.createElement(React.Fragment, null,
            React.createElement("div", { style: { position: 'relative', width: this.state.width, height: 150 } },
                React.createElement(Button_1.PrimaryButton, { style: { width: '100%', height: '100%' }, text: "Hover on the right edge to see the resizer" }),
                React.createElement(Resizer_1.Resizer, { ariaLabel: 'Use the left and right arrow keys to resize the example', location: Resizer_types_1.ResizerLocation.Right, size: this.state.width, onSizeChanged: function (width) { return _this.setState({ width: width }); }, minSize: 100, maxSize: 1000 })),
            React.createElement("div", { style: {
                    marginTop: 20,
                    position: 'relative',
                    width: this.props.initialSize,
                    height: this.state.height
                } },
                React.createElement(Button_1.PrimaryButton, { style: { width: '100%', height: '100%' }, text: "Hover on the bottom edge to see the resizer" }),
                React.createElement(Resizer_1.Resizer, { ariaLabel: 'Use the up and down arrow keys to resize the example', location: Resizer_types_1.ResizerLocation.Bottom, size: this.state.height, onSizeChanged: function (height) { return _this.setState({ height: height }); }, minSize: 100, maxSize: 1000 })),
            React.createElement("div", { "aria-live": "assertive", "aria-atomic": "true", style: hiddenAccessibleDivStyles }, "Example resized to " + this.state.height + " height"),
            React.createElement("div", { "aria-live": "assertive", "aria-atomic": "true", style: hiddenAccessibleDivStyles }, "Example resized to " + this.state.width + " width")));
    };
    return ResizableButton;
}(React.Component));
