// @business-app/fabric/lib-cjs/components/Toolbox/ToolboxItem.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolboxItem = void 0;
var tslib_1 = require("tslib");
var Styling_1 = require("@fluentui/react/lib/Styling");
var Utilities_1 = require("@fluentui/react/lib/Utilities");
var React = require("react");
var ToolboxItem = /** @class */ (function (_super) {
    tslib_1.__extends(ToolboxItem, _super);
    function ToolboxItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ToolboxItem.prototype.render = function () {
        var className = Styling_1.mergeStyles({
            minHeight: 0,
            display: 'flex',
            flexGrow: 1,
            selectors: {
                // For IE 10+ overflow property with fieldheaderregion minheight
                '@media all and (-ms-high-contrast: none)': {
                    minHeight: 'calc(100% - 48px)'
                }
            }
        }, this.props.className);
        return (React.createElement("div", tslib_1.__assign({}, Utilities_1.getNativeProps(this.props, Utilities_1.divProperties), { className: className }), this.props.children));
    };
    return ToolboxItem;
}(Utilities_1.BaseComponent));
exports.ToolboxItem = ToolboxItem;
