// @business-app/fabric/src/components/Accordion/examples/Accordion.Basic.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { DefaultButton } from '@fluentui/react/lib/Button';
import * as React from 'react';
import { Accordion } from '..';

export class AccordionBasicExample extends React.Component {
  public render(): JSX.Element {
    const exampleStyles = {
      divStyle: {
        marginBottom: '20px',
        width: 220
      },
      buttonStyle: {
        root: {
          width: '100%',
          minWidth: 0,
          background: 'transparent',
          padding: '12px 0',
          textAlign: 'left'
        },
        icon: { fontSize: 22, padding: '0' },
        menuIcon: { fontSize: 20, padding: '0 8px' }
      }
    };

    return (
      <div>
        <div style={exampleStyles.divStyle}>
          <Accordion
            text="Stuff here"
            iconProps={{ iconName: 'CollapseMenu' }}
            onRenderContent={() => {
              return (
                <div>
                  <DefaultButton
                    key="item1"
                    text="Link 1"
                    iconProps={{ iconName: 'CompassNW' }}
                    ariaLabel="Link 1"
                    onClick={() => {
                      return;
                    }}
                    styles={exampleStyles.buttonStyle}
                  />
                  <DefaultButton
                    key="item2"
                    text="Link 2"
                    iconProps={{ iconName: 'BuildQueue' }}
                    onClick={() => {
                      return;
                    }}
                    styles={exampleStyles.buttonStyle}
                  />
                  <DefaultButton
                    key="item3"
                    text="Link 3"
                    iconProps={{ iconName: 'AutoRacing' }}
                    onClick={() => {
                      return;
                    }}
                    styles={exampleStyles.buttonStyle}
                  />
                </div>
              );
            }}
          />
        </div>

        <div style={exampleStyles.divStyle}>
          <Accordion
            text="isHiddenOnDismiss set to true"
            iconProps={{ iconName: 'CollapseMenu' }}
            isHiddenOnDismiss={true}
            onRenderContent={() => {
              return (
                <div>
                  <DefaultButton
                    key="item1"
                    text="Link 1"
                    iconProps={{ iconName: 'CompassNW' }}
                    ariaLabel="Link 1"
                    onClick={() => {
                      return;
                    }}
                    styles={exampleStyles.buttonStyle}
                  />
                </div>
              );
            }}
          />
        </div>
      </div>
    );
  }
}
