// @business-app/fabric/lib-cjs/styler/Styler.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Styler = exports.FabricDarkStyles = exports.BusinessAppStyles = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var Styling_1 = require("@fluentui/react/lib/Styling");
var Utilities_1 = require("@fluentui/react/lib/Utilities");
var Themes = require("../utilities/themes");
var BusinessAppStyles_1 = require("./BusinessAppStyles");
var Icons_1 = require("./Icons");
var BusinessAppStyles_2 = require("./BusinessAppStyles");
Object.defineProperty(exports, "BusinessAppStyles", { enumerable: true, get: function () { return BusinessAppStyles_2.BusinessAppStyles; } });
var FabricDarkStyles_1 = require("./FabricDarkStyles");
Object.defineProperty(exports, "FabricDarkStyles", { enumerable: true, get: function () { return FabricDarkStyles_1.FabricDarkStyles; } });
var Styler = /** @class */ (function (_super) {
    tslib_1.__extends(Styler, _super);
    function Styler(props) {
        var _this = _super.call(this, props) || this;
        _this.render = function () {
            var _a = _this.props, children = _a.children, theme = _a.theme, themeOverrides = _a.themeOverrides, scopedSettings = _a.scopedSettings;
            var mergedThemeOptions = Utilities_1.merge(theme, themeOverrides);
            var mergedScopedSettings = Utilities_1.merge(BusinessAppStyles_1.BusinessAppStyles, scopedSettings);
            var customizations = {
                settings: { theme: Styling_1.createTheme(mergedThemeOptions) },
                scopedSettings: mergedScopedSettings
            };
            return React.createElement(Utilities_1.Customizer, tslib_1.__assign({}, customizations), children);
        };
        var _a = _this.props, iconSubset = _a.iconSubset, iconOptions = _a.iconOptions;
        Icons_1.icons(iconSubset, iconOptions);
        return _this;
    }
    Styler.defaultProps = {
        theme: Themes.DefaultTheme,
        themeOverrides: {}
    };
    return Styler;
}(React.Component));
exports.Styler = Styler;
