// @business-app/fabric/lib-cjs/styler/FabricDarkStyles/index.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FabricDarkStyles = void 0;
var themes_1 = require("./../../utilities/themes");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
exports.FabricDarkStyles = {
    DetailsList: {
        styles: {
            headerWrapper: {
                selectors: {
                    '.ms-DetailsHeader': {
                        borderColor: themes_1.DefaultDarkTheme.palette.neutralQuaternary
                    }
                }
            }
        }
    },
    DetailsRow: {
        styles: {
            root: {
                borderColor: themes_1.DefaultDarkTheme.palette.neutralQuaternaryAlt
            }
        }
    },
    VerticalDivider: {
        styles: {
            divider: {
                backgroundColor: themes_1.DefaultDarkTheme.palette.neutralQuaternaryAlt
            },
            wrapper: {
                Backgroundcolor: themes_1.DefaultDarkTheme.palette.green
            }
        }
    },
    DocumentCard: {
        styles: {
            root: {
                border: "1px solid " + themes_1.DefaultDarkTheme.palette.neutralQuaternaryAlt,
                selectors: {
                    '.ms-DocumentCardPreview': {
                        borderRight: "1px solid " + themes_1.DefaultDarkTheme.palette.neutralQuaternaryAlt
                    }
                }
            }
        }
    },
    DocumentCardPreview: {
        styles: {
            root: {
                borderBottom: "1px solid " + themes_1.DefaultDarkTheme.palette.neutralQuaternaryAlt,
                borderRight: "1px solid " + themes_1.DefaultDarkTheme.palette.neutralQuaternaryAlt
            }
        }
    },
    Persona: {
        styles: {
            intials: {
                color: themes_1.DefaultDarkTheme.palette.black
            }
        }
    },
    PersonaCoin: {
        color: themes_1.DefaultDarkTheme.palette.black,
        styles: {
            intials: {
                color: themes_1.DefaultDarkTheme.palette.black
            },
            initials: {
                'ms-Persona-initials': {
                    color: themes_1.DefaultDarkTheme.palette.black
                }
            }
        }
    },
    Panel: {
        styles: {
            main: {
                backgroundColor: themes_1.DefaultDarkTheme.palette.neutralLighter
            },
            closeButton: {
                color: themes_1.DefaultDarkTheme.palette.neutralSecondary,
                selectors: {
                    ':hover': {
                        color: themes_1.DefaultDarkTheme.palette.neutralPrimary
                    }
                }
            }
        }
    },
    PeoplePickerItem: {
        styles: {
            root: {
                background: themes_1.DefaultDarkTheme.palette.neutralQuaternaryAlt,
                selectors: {
                    ':hover': {
                        background: themes_1.DefaultDarkTheme.palette.neutralQuaternary
                    }
                }
            },
            removeButton: {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                tokens: function (props) {
                    var selected = props.selected, theme = props.theme;
                    var palette = theme.palette;
                    return [
                        {
                            background: 'transparent',
                            selectors: {
                                ':active': {
                                    color: palette.white,
                                    backgroundColor: palette.themeDark
                                }
                            }
                        },
                        !selected && {
                            color: palette.neutralPrimary
                        },
                        selected && {
                            color: palette.white
                        }
                    ];
                }
            }
        }
    },
    SelectedPersona: {
        styles: {
            main: {
                backgroundColor: themes_1.DefaultDarkTheme.palette.neutralLighter
            },
            closeButton: {
                color: themes_1.DefaultDarkTheme.palette.neutralSecondary,
                selectors: {
                    ':hover': {
                        color: themes_1.DefaultDarkTheme.palette.neutralPrimary
                    }
                }
            }
        }
    },
    SpinButton: {
        styles: {
            inputTextSelected: {
                color: themes_1.DefaultDarkTheme.palette.black,
                background: themes_1.DefaultDarkTheme.palette.themePrimary
            }
        }
    }
};
