// @business-app/fabric/lib/components/Card/RichCard/RichCard.base.js
import { __assign, __extends } from "tslib";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import * as React from 'react';
import { classNamesFunction } from '@fluentui/react/lib/Utilities';
import { Card } from './../Card';
import { CardFooter } from './../CardFooter';
import { CardHeader } from './../CardHeader';
var getClassNames = classNamesFunction();
var RichCardBase = /** @class */ (function (_super) {
    __extends(RichCardBase, _super);
    function RichCardBase(props) {
        var _this = _super.call(this, props) || this;
        var _a = _this.props, headerProps = _a.headerProps, footerProps = _a.footerProps, renderFooter = _a.renderFooter;
        _this.state = {
            hasHeader: !!headerProps,
            hasFooter: !!(footerProps || renderFooter)
        };
        return _this;
    }
    RichCardBase.prototype.render = function () {
        var _a = this.props, width = _a.width, height = _a.height, children = _a.children, className = _a.className, headerProps = _a.headerProps, footerProps = _a.footerProps, renderContentArea = _a.renderContentArea, renderFooter = _a.renderFooter, renderHeader = _a.renderHeader, styles = _a.styles, theme = _a.theme;
        var _b = this.state, hasHeader = _b.hasHeader, hasFooter = _b.hasFooter;
        var classNames = getClassNames(styles, {
            theme: theme,
            className: className,
            hasHeader: hasHeader,
            hasFooter: hasFooter
        });
        return (React.createElement(Card, { className: classNames.root, height: height, width: width },
            (headerProps || renderHeader) && React.createElement(CardHeader, __assign({}, headerProps), renderHeader === null || renderHeader === void 0 ? void 0 : renderHeader()),
            React.createElement("div", { className: classNames.contentArea }, renderContentArea ? renderContentArea() : children),
            (footerProps || renderFooter) && React.createElement(CardFooter, __assign({}, footerProps), renderFooter === null || renderFooter === void 0 ? void 0 : renderFooter())));
    };
    return RichCardBase;
}(React.Component));
export { RichCardBase };
