// @business-app/fabric/lib-cjs/components/Card/CardFooter/CardFooter.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.styles = void 0;
var styles = function (_a) {
    var theme = _a.theme, className = _a.className;
    var palette = theme.palette;
    return {
        root: [
            'ba-CardFooter',
            {
                paddingInlineStart: 16,
                paddingInlineEnd: 16,
                paddingBlockEnd: 16,
                display: 'flex',
                alignItems: 'center'
            },
            className
        ],
        link: [
            {
                display: 'inline-block'
            }
        ],
        buttonHolder: [
            'ba-CardFooter__ButtonHolder',
            {
                display: 'flex',
                flexGrow: '1',
                justifyContent: 'flex-end',
                selectors: {
                    '> button, > a': {
                        marginLeft: 14,
                        display: 'block'
                    }
                }
            }
        ]
    };
};
exports.styles = styles;
