// @business-app/fabric/src/components/NavigationController/examples/NavigationController.Custom.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IconButton, PrimaryButton } from '@fluentui/react/lib/Button';
import { Label } from '@fluentui/react/lib/Label';
import { IStyle, mergeStyles } from '@fluentui/react/lib/Styling';
import * as React from 'react';
import { NavigationController } from '..';
import { INavigationController, INavigationControllerProps } from '../NavigationController.types';

import { NavigationControllerView } from '../NavigationControllerView';
import { INavigationControllerViewProps } from '../NavigationControllerView.types';

const stepKeys = {
  intro: 'intro',
  end: 'end'
};

export class NavigationControllerCustomExample extends React.Component<{}, {}> {
  public render() {
    const containerStyles: IStyle = {
      border: '1px solid black',
      height: '450px',
      maxWidth: '800px',
      backgroundColor: '#FFF',
      fontSize: '16px',
      lineHeight: '20px'
    };

    return (
      <div className={mergeStyles(containerStyles)}>
        <NavigationController
          initialActiveViewKey={stepKeys.intro}
          showBackButton={true}
          onRenderHeader={this.onRenderHeader}
          backButtonAriaLabel={'Back'}
        >
          {NavigationControllerContext => [
            <NavigationControllerView key={stepKeys.intro} viewKey={stepKeys.intro} headerText="First step">
              <div
                style={{
                  display: 'flex',
                  height: '100%',
                  flexDirection: 'column',
                  justifyContent: 'space-around'
                }}
              >
                <div>
                  <Label>Look at that fancy header.</Label>
                  <Label>There is no footer for this step.</Label>
                </div>
                <PrimaryButton
                  onClick={() => NavigationControllerContext.goTo(stepKeys.end)}
                  text="Next Step"
                  ariaLabel="Next Step"
                />
              </div>
            </NavigationControllerView>,
            <NavigationControllerView
              key={stepKeys.end}
              viewKey={stepKeys.end}
              headerText="Last step"
              footerButtons={[
                {
                  primary: false,
                  onClick: () => NavigationControllerContext.reset(stepKeys.intro),
                  text: 'Start again'
                }
              ]}
            >
              <div>This is some more text.</div>
            </NavigationControllerView>
          ]}
        </NavigationController>
      </div>
    );
  }

  private onRenderHeader = (
    NavigationControllerProps: INavigationControllerProps,
    viewProps: INavigationControllerViewProps,
    context: INavigationController
  ): JSX.Element => {
    return (
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <IconButton
          iconProps={{
            iconName: 'CaretSolidLeft'
          }}
          disabled={!context.canGoBack()}
          onClick={context.back}
          ariaLabel="Back"
        />
        <div>
          <b>{viewProps.headerText}</b>
        </div>
        <IconButton
          iconProps={{
            iconName: 'CaretSolidRight'
          }}
          disabled={!context.canGoForward()}
          onClick={context.forward}
          ariaLabel="Forward"
        />
      </div>
    );
  };
}
