// @business-app/fabric/src/components/NavigationController/examples/NavigationController.Basic.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IStyle, mergeStyles } from '@fluentui/react/lib/Styling';
import * as React from 'react';
import { NavigationController } from '..';
import { NavigationControllerView } from '../NavigationControllerView';

const stepKeys = {
  intro: 'intro',
  ingredients: 'ingredients',
  directions: 'directions',
  end: 'end'
};

export class NavigationControllerBasicExample extends React.Component<{}, {}> {
  public render() {
    const containerStyles: IStyle = {
      border: '1px solid black',
      height: '500px',
      width: '300px',
      backgroundColor: '#FFF',
      fontSize: '16px',
      lineHeight: '20px'
    };

    return (
      <div className={mergeStyles(containerStyles)}>
        <NavigationController
          backButtonAriaLabel={'Back'}
          initialActiveViewKey={stepKeys.intro}
          showBackButton={true}
        >
          {NavigationControllerContext => [
            <NavigationControllerView
              key={stepKeys.intro}
              viewKey={stepKeys.intro}
              headerText="One pot garlic parmesan pasta"
              footerButtons={[
                {
                  primary: true,
                  onClick: () => NavigationControllerContext.goTo(stepKeys.ingredients),
                  text: 'Ready'
                }
              ]}
            >
              <div>
                Garlic and parmesan. A match made in heaven. And with the help of a single pan, you can make
                the easiest, most epic pasta ever.
              </div>
            </NavigationControllerView>,
            <NavigationControllerView
              key={stepKeys.ingredients}
              viewKey={stepKeys.ingredients}
              headerText="Ingredients"
              footerButtons={[
                {
                  primary: true,
                  onClick: () => NavigationControllerContext.goTo(stepKeys.directions),
                  text: 'Done'
                }
              ]}
            >
              <ol style={{ padding: '0 24px', margin: 0 }}>
                <li>2 tablespoons unsalted butter</li>
                <li>4 cloves garlic, minced</li>
                <li>2 cups chicken broth</li>
                <li>8 ounces uncooked fettuccine</li>
                <li>Kosher salt and freshly ground black pepper, to taste</li>
                <li>1/4 cup freshly grated Parmesan cheese</li>
                <li>2 tablespoons chopped fresh parsley leaves</li>
              </ol>
            </NavigationControllerView>,
            <NavigationControllerView
              key={stepKeys.directions}
              viewKey={stepKeys.directions}
              headerText="Directions"
              footerButtons={[
                {
                  primary: true,
                  onClick: () => NavigationControllerContext.goTo(stepKeys.end),
                  text: 'Done'
                }
              ]}
            >
              <ol style={{ padding: '0 24px', margin: 0 }}>
                <li>
                  Melt butter in a large skillet over medium high heat. Add garlic and cook, stirring
                  frequently, until fragrant, about 1-2 minutes.
                </li>
                <li>Stir in chicken broth, milk and fettuccine; season with salt and pepper, to taste.</li>
                <li>
                  Bring to a boil; reduce heat and simmer, stirring occasionally, until pasta is cooked
                  through, about 18-20 minutes. Stir in Parmesan. If the mixture is too thick, add more milk
                  as needed until desired consistency is reached.
                </li>
                <li>Serve immediately, garnished with parsley, if desired.</li>
              </ol>
            </NavigationControllerView>,
            <NavigationControllerView
              key={stepKeys.end}
              viewKey={stepKeys.end}
              headerText="Hurrah!"
              footerButtons={[
                {
                  primary: false,
                  onClick: () => NavigationControllerContext.reset(stepKeys.intro),
                  text: 'Start again'
                }
              ]}
            >
              <div>You did it. Bon Appétit!</div>
            </NavigationControllerView>
          ]}
        </NavigationController>
      </div>
    );
  }
}
