// @business-app/fabric/src/components/NavigationController/examples/NavigationController.Callout.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { DefaultButton } from '@fluentui/react/lib/Button';
import { Callout } from '@fluentui/react/lib/Callout';
import { Image } from '@fluentui/react/lib/Image';
import { IStyle, mergeStyles } from '@fluentui/react/lib/Styling';
import * as React from 'react';
import { NavigationController } from '..';
import { NavigationControllerView } from '../NavigationControllerView';
import { FocusTrapZone } from '@fluentui/react/lib/FocusTrapZone';

const stepKeys = {
  step1: 'step1',
  step2: 'step2',
  step3: 'step3'
};

export interface INavigationControllerCalloutExampleState {
  calloutOpen: boolean;
}

export class NavigationControllerCalloutExample extends React.Component<
  {},
  INavigationControllerCalloutExampleState
> {
  public state: INavigationControllerCalloutExampleState = {
    calloutOpen: false
  };
  private _menuButtonElement = React.createRef<HTMLDivElement>();
  public render() {
    const calloutContainerStyles: IStyle = {
      height: '500px',
      width: '300px',
      backgroundColor: '#FFF',
      fontSize: '16px',
      lineHeight: '20px'
    };
    const buttonAreaStyles: IStyle = {
      verticalAlign: 'top',
      display: 'inline-block',
      textAlign: 'center'
    };

    return (
      <div>
        <div className={mergeStyles(buttonAreaStyles)} ref={this._menuButtonElement}>
          <DefaultButton
            onClick={() => this.setState({ calloutOpen: !this.state.calloutOpen })}
            text={this.state.calloutOpen ? 'Hide callout' : 'Show callout'}
          />
        </div>

        <Callout
          gapSpace={0}
          target={this._menuButtonElement.current}
          onDismiss={() => this.setState({ calloutOpen: !this.state.calloutOpen })}
          setInitialFocus={true}
          hidden={!this.state.calloutOpen}
          styles={{
            calloutMain: calloutContainerStyles
          }}
        >
          <FocusTrapZone disabled={!this.state.calloutOpen}>
            <NavigationController
              initialActiveViewKey={stepKeys.step1}
              showBackButton={true}
              backButtonAriaLabel={'Back'}
            >
              {NavigationControllerContext => [
                <NavigationControllerView
                  key={stepKeys.step1}
                  viewKey={stepKeys.step1}
                  headerText="Step 1"
                  footerButtons={[
                    {
                      primary: true,
                      onClick: () => NavigationControllerContext.goTo(stepKeys.step2),
                      text: 'Next'
                    }
                  ]}
                >
                  <div>Get Cape.</div>
                </NavigationControllerView>,
                <NavigationControllerView
                  key={stepKeys.step2}
                  viewKey={stepKeys.step2}
                  headerText="Step 2"
                  footerButtons={[
                    {
                      primary: true,
                      onClick: () => NavigationControllerContext.goTo(stepKeys.step3),
                      text: 'Next'
                    }
                  ]}
                >
                  <div>Wear Cape.</div>
                </NavigationControllerView>,
                <NavigationControllerView
                  key={stepKeys.step3}
                  viewKey={stepKeys.step3}
                  headerText="Step 3"
                  footerButtons={[
                    {
                      primary: true,
                      onClick: () => this.setState({ calloutOpen: !this.state.calloutOpen }),
                      text: 'Hide callout'
                    }
                  ]}
                >
                  <div>Fly.</div>
                  <div>
                    <Image
                      styles={{
                        root: {
                          margin: '20px auto'
                        }
                      }}
                      height={300}
                      src="https://thefancy-media-ec5.thefancy.com/1280/20120209/288502717_157326640339.jpg"
                    />
                  </div>
                </NavigationControllerView>
              ]}
            </NavigationController>
          </FocusTrapZone>
        </Callout>
      </div>
    );
  }
}
