// @business-app/fabric/src/components/Error/Error.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IButtonStyles } from '@fluentui/react/lib/Button';
import { IStyleFunction } from '@fluentui/react/lib/Utilities';
import { ITheme } from '@fluentui/react/lib/Styling';

import { IErrorStyleProps, IErrorStyles } from './Error.types';

export const styles: IStyleFunction<IErrorStyleProps, IErrorStyles> = ({ theme, className }) => {
  const { black, neutralLighter, neutralSecondary, themePrimary, themeTertiary } = theme.palette;

  const reflowSelectors = {
    selectors: {
      '@media (max-width: 480px)': {
        'max-width': '100%'
      },
      '@media (min-width: 481px)': {
        'max-width': 420
      }
    }
  };

  return {
    root: [
      'ba-Error',
      className,
      {
        color: neutralSecondary,
        paddingTop: 188,
        minHeight: '100%',
        display: 'flex',
        alignItems: 'center',
        flexDirection: 'column'
      }
    ],
    title: [
      'ba-Error-Title',
      {
        fontSize: 28,
        fontWeight: 'bold',
        textAlign: 'center',
        margin: 0,
        marginBottom: 16,
        paddingLeft: 10,
        paddingRight: 10,
        color: black,
        overflowWrap: 'break-word',
        ...reflowSelectors
      }
    ],
    subtitle: [
      'ba-Error-Subtitle',
      {
        fontSize: 12,
        lineHeight: 16,
        textAlign: 'justify',
        margin: 0,
        marginBottom: 16,
        paddingLeft: 10,
        paddingRight: 10,
        color: black,
        overflowWrap: 'break-word',
        ...reflowSelectors
      }
    ],
    learnMoreLink: ['ba-Error-LearnMoreLink'],
    customerSupportLink: ['ba-Error-CustomerSupportLink'],
    seeDetailsButton: [
      'ba-Error-SeeDetailsButton',
      {
        marginTop: 8,
        borderColor: 'transparent',
        color: themePrimary
      }
    ],
    seeDetailsDiv: [
      'ba-Error-SeeDetailsDiv',
      {
        fontSize: 12,
        lineHeight: 19,
        backgroundColor: neutralLighter,
        marginTop: 4,
        padding: 8,
        overflowWrap: 'break-word',
        ...reflowSelectors
      }
    ],
    callToActionButton: [
      'ba-Error-CallToActionButton',
      {
        marginTop: 8
      }
    ],
    imageSmall: [
      'ba-Error-ImageSmall',
      {
        selectors: {
          '.themeSecondary': {
            fill: themeTertiary
          },
          '.themePrimary': {
            fill: themePrimary
          },
          '.themePrimaryGradient': {
            stopColor: themePrimary
          },
          '@media (max-width: 639px)': {
            marginTop: 90,
            width: 200,
            height: 200
          },
          '@media (min-width: 640px)': {
            display: 'none'
          }
        }
      }
    ],
    imageMedium: [
      'ba-Error-ImageMedium',
      {
        selectors: {
          '.themeSecondary': {
            fill: themeTertiary
          },
          '.themePrimary': {
            fill: themePrimary
          },
          '.themeSecondaryGradient': {
            stopColor: themeTertiary
          },
          '.themePrimaryGradient': {
            stopColor: themePrimary
          },
          '@media (max-width: 639px) or (min-width: 1024px)': {
            display: 'none'
          },
          '@media (min-width: 640px) and (max-width: 1023px)': {
            marginTop: 50,
            width: '65%'
          }
        }
      }
    ],
    imageLarge: [
      'ba-Error-ImageLarge',
      {
        selectors: {
          '.themeSecondary': {
            fill: themeTertiary
          },
          '.themePrimary': {
            fill: themePrimary
          },
          '.themeSecondaryGradient': {
            stopColor: themeTertiary
          },
          '.themePrimaryGradient': {
            stopColor: themePrimary
          },
          '@media (max-width: 1023px)': {
            display: 'none'
          },
          '@media (min-width: 1024px)': {
            marginTop: 'auto',
            marginBottom: '-4px',
            width: '65%'
          }
        }
      }
    ],
    imageFooter: [
      'ba-Error-ImageFooter',
      {
        selectors: {
          '@media (max-width: 1023px)': {
            display: 'none'
          },
          '@media (min-width: 1024px)': {
            justifySelf: 'flex-end',
            backgroundColor: themePrimary,
            width: '100%',
            height: 20
          }
        }
      }
    ]
  };
};

export function getDefaultButtonStyles(theme: ITheme): IButtonStyles {
  return {
    root: {
      borderColor: 'transparent',
      backgroundColor: theme.palette.themePrimary
    },
    rootHovered: {
      backgroundColor: theme.palette.neutralQuaternaryAlt
    }
  };
}

export function getDefaultLinkStyles(theme: ITheme): IButtonStyles {
  return {
    root: {
      marginLeft: 5,
      color: theme.palette.themePrimary,
      textDecoration: 'underline'
    }
  };
}
