// @business-app/fabric/src/components/Error/Error.types.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IStyle, ITheme } from '@fluentui/react/lib/Styling';
import { IStyleFunctionOrObject } from '@fluentui/react/lib/Utilities';

export interface IErrorProps extends Readonly<React.HTMLAttributes<HTMLDivElement>> {
  /**
   * The primary error message label.
   * @type {string}
   */
  readonly title: string;

  /**
   * The secondary error message label.
   * @type {string}
   */
  readonly subtitle?: string;

  /**
   * The "Learn more" link.
   * @type {IErrorLink}
   */
  readonly learnMoreLink?: IErrorLink;

  /**
   * The label to be put on the "See details" link. This and seeDetailsProps must both be set to have a "See details"
   * link.
   * @type {string}
   */
  readonly seeDetailsLinkShowLabel?: string;

  /**
   * The label to be put on the "See details" link. This and seeDetailsProps must both be set to have a "See details"
   * link.
   * @type {string}
   */
  readonly seeDetailsLinkHideLabel?: string;

  /**
   * The specific error details responsible for displaying the error.
   */
  readonly errorDetails?: IErrorDetailsProps;

  /**
   * The "Call to Action" link.
   * @type {IErrorLink}
   */
  readonly callToActionLink?: IErrorLink;

  /**
   * Theme provided by High-Order Component.
   * @type {ITheme}
   */
  theme?: ITheme;

  /**
   * Call to provide customized styling that will layer on top of the variant rules.
   */
  styles?: IStyleFunctionOrObject<Pick<IErrorProps, 'theme'>, IErrorStyles>;
}

export interface IErrorDetailsProps {
  /**
   * The label to be put on the "Message" field. This and message must both be set to have a "Message" field.
   * @type {string}
   */
  messageLabel?: string;

  /**
   * The text to be displayed as "Message".
   * @type {string}
   */
  message?: string;

  /**
   * The label to be put on the "Timestamp" field. This and timeStamp must both be set to have a "Timestamp" field.
   * @type {string}
   */
  timeStampLabel?: string;

  /**
   * The text to be displayed as "Timestamp".
   * @type {string}
   */
  timeStamp?: string;

  /**
   * The label to be put on the "Session ID" field. This and sessionId must both be set to have a "Session ID" field.
   * @type {string}
   */
  sessionIdLabel?: string;

  /**
   * The text to be displayed as "Session ID".
   * @type {string}
   */
  sessionId?: string;

  /**
   * The label to be put on the "ObjectId" field. This and objectId must both be set to have a "ObjectId" field.
   * @type {string}
   */
  objectIdLabel?: string;

  /**
   * The text to be displayed as "ObjectId".
   * @type {string}
   */
  objectId?: string;

  /**
   * The label to be put on the "Tenant" field. This and tenantId must both be set to have a "Tenant" field.
   * @type {string}
   */
  tenantIdLabel?: string;

  /**
   * The text to be displayed as "Tenant".
   * @type {string}
   */
  tenantId?: string;

  /**
   * The "Customer Support" link to be shown inside "See details" section.
   * @type {IErrorLink}
   */
  readonly customerSupportLink?: IErrorLink;
}

export interface IErrorLink {
  /**
   * The label to be put on the link. This and customerSupportUrl must both be set to have a link.
   * @type {string}
   */
  readonly label: string;

  /**
   * The url for link.
   * @type {string}
   */
  readonly url: string;

  /**
   * The target for link. (Ex: "_blank" or "_self")
   * @type {string}
   */
  readonly target?: string;
}

export interface IErrorStyleProps extends Pick<IErrorProps, 'theme' | 'className'> {
  seeDetailsButtonClassName?: string;
  callToActionButtonClassName?: string;
}

export interface IErrorStyles {
  root: IStyle;
  title: IStyle;
  subtitle: IStyle;
  learnMoreLink: IStyle;
  customerSupportLink: IStyle;
  seeDetailsButton: IStyle;
  seeDetailsDiv: IStyle;
  callToActionButton: IStyle;
  imageSmall: IStyle;
  imageMedium: IStyle;
  imageLarge: IStyle;
  imageFooter: IStyle;
}
