// @business-app/fabric/lib-cjs/components/NavigationController/examples/NavigationController.Custom.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NavigationControllerCustomExample = void 0;
var tslib_1 = require("tslib");
var Button_1 = require("@fluentui/react/lib/Button");
var Label_1 = require("@fluentui/react/lib/Label");
var Styling_1 = require("@fluentui/react/lib/Styling");
var React = require("react");
var __1 = require("..");
var NavigationControllerView_1 = require("../NavigationControllerView");
var stepKeys = {
    intro: 'intro',
    end: 'end'
};
var NavigationControllerCustomExample = /** @class */ (function (_super) {
    tslib_1.__extends(NavigationControllerCustomExample, _super);
    function NavigationControllerCustomExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onRenderHeader = function (NavigationControllerProps, viewProps, context) {
            return (React.createElement("div", { style: { display: 'flex', justifyContent: 'space-between' } },
                React.createElement(Button_1.IconButton, { iconProps: {
                        iconName: 'CaretSolidLeft'
                    }, disabled: !context.canGoBack(), onClick: context.back, ariaLabel: "Back" }),
                React.createElement("div", null,
                    React.createElement("b", null, viewProps.headerText)),
                React.createElement(Button_1.IconButton, { iconProps: {
                        iconName: 'CaretSolidRight'
                    }, disabled: !context.canGoForward(), onClick: context.forward, ariaLabel: "Forward" })));
        };
        return _this;
    }
    NavigationControllerCustomExample.prototype.render = function () {
        var containerStyles = {
            border: '1px solid black',
            height: '450px',
            maxWidth: '800px',
            backgroundColor: '#FFF',
            fontSize: '16px',
            lineHeight: '20px'
        };
        return (React.createElement("div", { className: Styling_1.mergeStyles(containerStyles) },
            React.createElement(__1.NavigationController, { initialActiveViewKey: stepKeys.intro, showBackButton: true, onRenderHeader: this.onRenderHeader, backButtonAriaLabel: 'Back' }, function (NavigationControllerContext) { return [
                React.createElement(NavigationControllerView_1.NavigationControllerView, { key: stepKeys.intro, viewKey: stepKeys.intro, headerText: "First step" },
                    React.createElement("div", { style: {
                            display: 'flex',
                            height: '100%',
                            flexDirection: 'column',
                            justifyContent: 'space-around'
                        } },
                        React.createElement("div", null,
                            React.createElement(Label_1.Label, null, "Look at that fancy header."),
                            React.createElement(Label_1.Label, null, "There is no footer for this step.")),
                        React.createElement(Button_1.PrimaryButton, { onClick: function () { return NavigationControllerContext.goTo(stepKeys.end); }, text: "Next Step", ariaLabel: "Next Step" }))),
                React.createElement(NavigationControllerView_1.NavigationControllerView, { key: stepKeys.end, viewKey: stepKeys.end, headerText: "Last step", footerButtons: [
                        {
                            primary: false,
                            onClick: function () { return NavigationControllerContext.reset(stepKeys.intro); },
                            text: 'Start again'
                        }
                    ] },
                    React.createElement("div", null, "This is some more text."))
            ]; })));
    };
    return NavigationControllerCustomExample;
}(React.Component));
exports.NavigationControllerCustomExample = NavigationControllerCustomExample;
