// @business-app/fabric/lib-cjs/components/NavigationController/examples/NavigationController.Callout.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NavigationControllerCalloutExample = void 0;
var tslib_1 = require("tslib");
var Button_1 = require("@fluentui/react/lib/Button");
var Callout_1 = require("@fluentui/react/lib/Callout");
var Image_1 = require("@fluentui/react/lib/Image");
var Styling_1 = require("@fluentui/react/lib/Styling");
var React = require("react");
var __1 = require("..");
var NavigationControllerView_1 = require("../NavigationControllerView");
var FocusTrapZone_1 = require("@fluentui/react/lib/FocusTrapZone");
var stepKeys = {
    step1: 'step1',
    step2: 'step2',
    step3: 'step3'
};
var NavigationControllerCalloutExample = /** @class */ (function (_super) {
    tslib_1.__extends(NavigationControllerCalloutExample, _super);
    function NavigationControllerCalloutExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            calloutOpen: false
        };
        _this._menuButtonElement = React.createRef();
        return _this;
    }
    NavigationControllerCalloutExample.prototype.render = function () {
        var _this = this;
        var calloutContainerStyles = {
            height: '500px',
            width: '300px',
            backgroundColor: '#FFF',
            fontSize: '16px',
            lineHeight: '20px'
        };
        var buttonAreaStyles = {
            verticalAlign: 'top',
            display: 'inline-block',
            textAlign: 'center'
        };
        return (React.createElement("div", null,
            React.createElement("div", { className: Styling_1.mergeStyles(buttonAreaStyles), ref: this._menuButtonElement },
                React.createElement(Button_1.DefaultButton, { onClick: function () { return _this.setState({ calloutOpen: !_this.state.calloutOpen }); }, text: this.state.calloutOpen ? 'Hide callout' : 'Show callout' })),
            React.createElement(Callout_1.Callout, { gapSpace: 0, target: this._menuButtonElement.current, onDismiss: function () { return _this.setState({ calloutOpen: !_this.state.calloutOpen }); }, setInitialFocus: true, hidden: !this.state.calloutOpen, styles: {
                    calloutMain: calloutContainerStyles
                } },
                React.createElement(FocusTrapZone_1.FocusTrapZone, { disabled: !this.state.calloutOpen },
                    React.createElement(__1.NavigationController, { initialActiveViewKey: stepKeys.step1, showBackButton: true, backButtonAriaLabel: 'Back' }, function (NavigationControllerContext) { return [
                        React.createElement(NavigationControllerView_1.NavigationControllerView, { key: stepKeys.step1, viewKey: stepKeys.step1, headerText: "Step 1", footerButtons: [
                                {
                                    primary: true,
                                    onClick: function () { return NavigationControllerContext.goTo(stepKeys.step2); },
                                    text: 'Next'
                                }
                            ] },
                            React.createElement("div", null, "Get Cape.")),
                        React.createElement(NavigationControllerView_1.NavigationControllerView, { key: stepKeys.step2, viewKey: stepKeys.step2, headerText: "Step 2", footerButtons: [
                                {
                                    primary: true,
                                    onClick: function () { return NavigationControllerContext.goTo(stepKeys.step3); },
                                    text: 'Next'
                                }
                            ] },
                            React.createElement("div", null, "Wear Cape.")),
                        React.createElement(NavigationControllerView_1.NavigationControllerView, { key: stepKeys.step3, viewKey: stepKeys.step3, headerText: "Step 3", footerButtons: [
                                {
                                    primary: true,
                                    onClick: function () { return _this.setState({ calloutOpen: !_this.state.calloutOpen }); },
                                    text: 'Hide callout'
                                }
                            ] },
                            React.createElement("div", null, "Fly."),
                            React.createElement("div", null,
                                React.createElement(Image_1.Image, { styles: {
                                        root: {
                                            margin: '20px auto'
                                        }
                                    }, height: 300, src: "https://thefancy-media-ec5.thefancy.com/1280/20120209/288502717_157326640339.jpg" })))
                    ]; })))));
    };
    return NavigationControllerCalloutExample;
}(React.Component));
exports.NavigationControllerCalloutExample = NavigationControllerCalloutExample;
