// @business-app/fabric/lib-cjs/components/NavigationController/examples/NavigationController.Basic.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NavigationControllerBasicExample = void 0;
var tslib_1 = require("tslib");
var Styling_1 = require("@fluentui/react/lib/Styling");
var React = require("react");
var __1 = require("..");
var NavigationControllerView_1 = require("../NavigationControllerView");
var stepKeys = {
    intro: 'intro',
    ingredients: 'ingredients',
    directions: 'directions',
    end: 'end'
};
var NavigationControllerBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(NavigationControllerBasicExample, _super);
    function NavigationControllerBasicExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NavigationControllerBasicExample.prototype.render = function () {
        var containerStyles = {
            border: '1px solid black',
            height: '500px',
            width: '300px',
            backgroundColor: '#FFF',
            fontSize: '16px',
            lineHeight: '20px'
        };
        return (React.createElement("div", { className: Styling_1.mergeStyles(containerStyles) },
            React.createElement(__1.NavigationController, { backButtonAriaLabel: 'Back', initialActiveViewKey: stepKeys.intro, showBackButton: true }, function (NavigationControllerContext) { return [
                React.createElement(NavigationControllerView_1.NavigationControllerView, { key: stepKeys.intro, viewKey: stepKeys.intro, headerText: "One pot garlic parmesan pasta", footerButtons: [
                        {
                            primary: true,
                            onClick: function () { return NavigationControllerContext.goTo(stepKeys.ingredients); },
                            text: 'Ready'
                        }
                    ] },
                    React.createElement("div", null, "Garlic and parmesan. A match made in heaven. And with the help of a single pan, you can make the easiest, most epic pasta ever.")),
                React.createElement(NavigationControllerView_1.NavigationControllerView, { key: stepKeys.ingredients, viewKey: stepKeys.ingredients, headerText: "Ingredients", footerButtons: [
                        {
                            primary: true,
                            onClick: function () { return NavigationControllerContext.goTo(stepKeys.directions); },
                            text: 'Done'
                        }
                    ] },
                    React.createElement("ol", { style: { padding: '0 24px', margin: 0 } },
                        React.createElement("li", null, "2 tablespoons unsalted butter"),
                        React.createElement("li", null, "4 cloves garlic, minced"),
                        React.createElement("li", null, "2 cups chicken broth"),
                        React.createElement("li", null, "8 ounces uncooked fettuccine"),
                        React.createElement("li", null, "Kosher salt and freshly ground black pepper, to taste"),
                        React.createElement("li", null, "1/4 cup freshly grated Parmesan cheese"),
                        React.createElement("li", null, "2 tablespoons chopped fresh parsley leaves"))),
                React.createElement(NavigationControllerView_1.NavigationControllerView, { key: stepKeys.directions, viewKey: stepKeys.directions, headerText: "Directions", footerButtons: [
                        {
                            primary: true,
                            onClick: function () { return NavigationControllerContext.goTo(stepKeys.end); },
                            text: 'Done'
                        }
                    ] },
                    React.createElement("ol", { style: { padding: '0 24px', margin: 0 } },
                        React.createElement("li", null, "Melt butter in a large skillet over medium high heat. Add garlic and cook, stirring frequently, until fragrant, about 1-2 minutes."),
                        React.createElement("li", null, "Stir in chicken broth, milk and fettuccine; season with salt and pepper, to taste."),
                        React.createElement("li", null, "Bring to a boil; reduce heat and simmer, stirring occasionally, until pasta is cooked through, about 18-20 minutes. Stir in Parmesan. If the mixture is too thick, add more milk as needed until desired consistency is reached."),
                        React.createElement("li", null, "Serve immediately, garnished with parsley, if desired."))),
                React.createElement(NavigationControllerView_1.NavigationControllerView, { key: stepKeys.end, viewKey: stepKeys.end, headerText: "Hurrah!", footerButtons: [
                        {
                            primary: false,
                            onClick: function () { return NavigationControllerContext.reset(stepKeys.intro); },
                            text: 'Start again'
                        }
                    ] },
                    React.createElement("div", null, "You did it. Bon App\u00E9tit!"))
            ]; })));
    };
    return NavigationControllerBasicExample;
}(React.Component));
exports.NavigationControllerBasicExample = NavigationControllerBasicExample;
