// @business-app/fabric/lib/components/Card/CardHeader/CardHeader.styles.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
export var styles = function (_a) {
    var className = _a.className, height = _a.height, title = _a.title, subTitle = _a.subTitle, theme = _a.theme;
    var palette = theme.palette;
    var headerHeight = (height !== null && height !== void 0 ? height : subTitle) ? 46 : 34;
    var objectHeights = subTitle ? 36 : 24;
    var titleImageSize = title && subTitle ? 36 : 24;
    var subTitleImageSize = 12;
    var textColor = palette.black;
    return {
        headerContainer: [
            'ba-CardHeader',
            {
                height: headerHeight,
                paddingBlockStart: 10,
                paddingInlineStart: 16,
                paddingInlineEnd: 16,
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'space-between',
                boxSizing: 'border-box'
            },
            className
        ],
        titleContainer: {
            display: 'block',
            flexDirection: 'column',
            flexGrow: '1'
        },
        titleIcon: {
            width: titleImageSize,
            height: titleImageSize,
            marginInlineEnd: 8,
            color: textColor,
            fontSize: titleImageSize
        },
        title: {
            padding: 0,
            display: 'block',
            color: textColor,
            fontSize: 14,
            fontWeight: 'bold',
            lineHeight: 20
        },
        subTitleContainer: {
            marginTop: 2,
            display: 'flex',
            alignItems: 'center',
            flexGrow: '1'
        },
        subTitleIcon: {
            width: subTitleImageSize,
            height: subTitleImageSize,
            marginInlineEnd: 5,
            color: textColor,
            fontSize: subTitleImageSize
        },
        subTitle: {
            padding: 0,
            display: 'inline-block',
            color: textColor,
            fontSize: 12,
            fontWeight: 'normal',
            lineHeight: 16
        },
        actionMenuContainer: {
            height: '100%',
            display: 'flex',
            selectors: {
                a: {
                    lineHeight: objectHeights
                }
            }
        },
        actionMenuItem: {
            marginRight: 5,
            selectors: {
                ':last-child': {
                    marginRight: 0
                }
            }
        },
        actionMenuLink: {
            lineHeight: objectHeights
        },
        actionMenuButton: {
            width: objectHeights,
            maxWidth: objectHeights,
            height: objectHeights,
            maxHeight: objectHeights,
            padding: 0
        }
    };
};
