// @business-app/fabric/lib-cjs/components/Card/Card.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.styles = void 0;
var styles = function (_a) {
    var width = _a.width, height = _a.height, theme = _a.theme, className = _a.className;
    var effects = theme.effects, palette = theme.palette;
    return {
        root: [
            'ba-Card',
            {
                height: height,
                maxHeight: height,
                width: width,
                maxWidth: width,
                display: 'flex',
                position: 'relative',
                flexDirection: 'column',
                border: "1px solid " + palette.neutralLight,
                borderRadius: '3px',
                boxSizing: 'border-box',
                boxShadow: effects.elevation4,
                transition: 'box-shadow 250ms ease-in-out',
                selectors: {
                    ':hover': {
                        boxShadow: effects.elevation16
                    }
                }
            },
            className
        ]
    };
};
exports.styles = styles;
