// @business-app/fabric/lib-cjs/components/Carousel/examples/Carousel.ControlButtons.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CarouselControlButtonsExample = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var __1 = require("..");
var Carousel_Example_Data_1 = require("./Carousel.Example.Data");
var CarouselControlButtonsExample = /** @class */ (function (_super) {
    tslib_1.__extends(CarouselControlButtonsExample, _super);
    function CarouselControlButtonsExample(props) {
        return _super.call(this, props) || this;
    }
    CarouselControlButtonsExample.prototype.render = function () {
        var customCarouselStyles = {
            root: {
                flexDirection: 'column'
            },
            subComponentStyles: {
                tooltipControl: {
                    root: {
                        position: 'initial',
                        transform: 'initial',
                        marginRight: 8
                    }
                },
                control: {
                    root: {
                        borderRadius: 'initial'
                    }
                },
                indicator: {},
                toggle: {}
            },
            indicatorsWrapper: {
                position: 'initial',
                transform: 'initial'
            },
            controlsWrapper: {
                margin: 24
            }
        };
        return (React.createElement(__1.Carousel, { styles: customCarouselStyles, items: Carousel_Example_Data_1.getMockContent(), useControlLabels: true, disableNavigationLoop: true, navPreviousButtonAriaLabel: "Previous", navNextButtonAriaLabel: "Next", indicatorButtonsAriaLabel: ['Item 1', 'Item 2', 'Item 3', 'Item 4'], playButtonAriaLabel: "Play", pauseButtonAriaLabel: "Pause", orderIndicatorsBeforeControls: true }));
    };
    return CarouselControlButtonsExample;
}(React.Component));
exports.CarouselControlButtonsExample = CarouselControlButtonsExample;
