// @business-app/fabric/lib-cjs/components/Carousel/examples/Carousel.Example.Data.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMockContent = void 0;
var React = require("react");
var getMockContent = function () {
    var itemsProps = [
        {
            heading: 'Heading 1',
            paragraph: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
            background: 'linear-gradient(to bottom, #F25022, #FFFFFF)'
        },
        {
            heading: 'Heading 2',
            paragraph: 'Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
            background: 'linear-gradient(to bottom, #7FBA00, #FFFFFF)'
        },
        {
            heading: 'Heading 3',
            paragraph: 'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.',
            background: 'linear-gradient(to bottom, #00A4EF, #FFFFFF)'
        },
        {
            heading: 'Heading 4',
            paragraph: 'Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
            background: 'linear-gradient(to bottom, #FFB900, #FFFFFF)'
        }
    ];
    return itemsProps.map(function (_a, index) {
        var heading = _a.heading, paragraph = _a.paragraph, background = _a.background;
        return (React.createElement("div", { key: heading + "-" + index, style: {
                boxSizing: 'border-box',
                padding: 100,
                height: 400,
                background: background
            } },
            React.createElement("h1", { className: "ms-font-xl-plus" }, heading),
            React.createElement("p", { className: "ms-font-m" }, paragraph)));
    });
};
exports.getMockContent = getMockContent;
