// @business-app/fabric/lib/components/HighlightedText/examples/HighlightedText.Basic.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { HighlightedText } from '../HighlightedText';
var HighlightedTextBasicExample = /** @class */ (function (_super) {
    __extends(HighlightedTextBasicExample, _super);
    function HighlightedTextBasicExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HighlightedTextBasicExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(HighlightedText, { text: 'Hello world. It is a marvelous time to be a highlight.', highlights: [
                    {
                        startIndex: 0,
                        length: 11
                    },
                    {
                        startIndex: 21,
                        length: 9
                    }
                ], labels: {
                    beginHighlightedText: 'Begin highlighted text',
                    endHighlightedText: 'End highlighted text'
                } })));
    };
    return HighlightedTextBasicExample;
}(React.Component));
export { HighlightedTextBasicExample };
