// @business-app/fabric/lib/components/HighlightedText/examples/HighlightedText.Styled.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { HighlightedText } from '../HighlightedText';
var HighlightedTextStyledExample = /** @class */ (function (_super) {
    __extends(HighlightedTextStyledExample, _super);
    function HighlightedTextStyledExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HighlightedTextStyledExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(HighlightedText, { text: 'Highlights can be styled, too!', highlights: [
                    {
                        startIndex: 18,
                        length: 6
                    }
                ], labels: {
                    beginHighlightedText: 'Begin highlighted text',
                    endHighlightedText: 'End highlighted text'
                }, styles: {
                    highlights: {
                        backgroundColor: 'yellow'
                    }
                } })));
    };
    return HighlightedTextStyledExample;
}(React.Component));
export { HighlightedTextStyledExample };
