// @business-app/fabric/src/components/FilterableComboBox/examples/FilterableComboBox.Grouped.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IComboBoxOption } from '@fluentui/react/lib/ComboBox';
import { Label } from '@fluentui/react/lib/Label';
import { SelectableOptionMenuItemType } from '@fluentui/react/lib/SelectableOption';
import { Stack } from '@fluentui/react/lib/Stack';
import * as React from 'react';
import { FilterableComboBox } from '../FilterableComboBox';

const tags: IComboBoxOption[] = [
  '_Colors',
  'Azure',
  'Black',
  'Cerulean',
  'Gray',
  'Green',
  'Lilac',
  'Rose',
  'Ruby',
  'Yellow',
  'オレンジ',
  '青い',
  '&div1',
  '_Flavors',
  'Berry',
  'Chocolate',
  'Pineapple',
  'Strawberry',
  'Vanilla',
  '&div2',
  '_Emotions',
  'Angry',
  'Bored',
  'Happy',
  'Lethargic',
  'Nonplussed',
  'Peaceful',
  'Restless',
  'Sad'
].map<IComboBoxOption>(name => ({
  key: name.toLowerCase(),
  text: name.indexOf('_') === 0 ? name.substr(1) : name,
  itemType:
    name.indexOf('_') === 0
      ? SelectableOptionMenuItemType.Header
      : name.indexOf('&') === 0
      ? SelectableOptionMenuItemType.Divider
      : SelectableOptionMenuItemType.Normal
}));

export class FilterableComboBoxGroupedExample extends React.Component<
  {},
  { selectedKey: string; selectedOption: IComboBoxOption | undefined }
> {
  public state = {
    selectedKey: 'vanilla',
    selectedOption: { key: 'vanilla', text: 'Vanilla' }
  };

  public render(): JSX.Element {
    return (
      <Stack maxWidth={400} gap={16}>
        <FilterableComboBox
          label={'Choose an option'}
          options={tags}
          noResultsMessage={query => (
            <Label
              styles={{ root: { padding: '8px 16px' } }}
            >{`No results found matching "${query}".`}</Label>
          )}
          calloutProps={{ calloutMaxHeight: 400 }}
          revertToLastValidOptionOnInvalidFilter={true}
          selectedKey={this.state.selectedKey}
          useComboBoxAsMenuWidth={true}
          onChange={(evt, option) =>
            option && this.setState({ selectedKey: option.key as string, selectedOption: option })
          }
        />
        <p>
          <Label>Last selected option: {this.state.selectedOption && this.state.selectedOption.text}</Label>
        </p>
      </Stack>
    );
  }
}
