// @business-app/fabric/lib-cjs/utilities/themes.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowBluePalette = exports.PowerAppsPurplePalette = exports.FlowTheme = exports.Dynamics365Theme = exports.PowerVirtualAgentContrastTheme = exports.PowerVirtualAgentDarkTheme = exports.PowerVirtualAgentTheme = exports.PowerAutomateTheme = exports.PowerBITheme = exports.PowerAppsContrastTheme = exports.PowerAppsDarkTheme = exports.PowerAppsTheme = exports.DesignDocsTheme = exports.DefaultDarkTheme = exports.DefaultTheme = exports.PowerVirtualAgentGrayscaleContrastPalette = exports.PowerVirtualAgentGrayscaleDarkPalette = exports.PowerVirtualAgentGrayscalePalette = exports.PowerAutomateGrayscalePalette = exports.PowerAppsGrayscaleContrastPalette = exports.PowerAppsGrayscaleDarkPalette = exports.PowerAppsGrayscalePalette = exports.DefaultGrayscalePalette = void 0;
var tslib_1 = require("tslib");
exports.DefaultGrayscalePalette = {
    white: '#fff',
    neutralLighterAlt: '#f8f8f8',
    neutralLighter: '#f4f4f4',
    neutralLight: '#eaeaea',
    neutralQuaternaryAlt: '#dadada',
    neutralQuaternary: '#d0d0d0',
    neutralTertiaryAlt: '#c8c8c8',
    neutralTertiary: '#bab8b7',
    neutralSecondary: '#a3a2a0',
    neutralPrimaryAlt: '#8d8b8a',
    neutralDark: '#605e5d',
    neutralPrimary: '#323130',
    black: '#000' // Black
};
exports.PowerAppsGrayscalePalette = {
    white: '#fff',
    neutralLighterAlt: '#FAF9F8',
    neutralLighter: '#F3F2F1',
    neutralLight: '#EDEBE9',
    neutralQuaternaryAlt: '#E1DFDD',
    neutralQuaternary: '#D2D0CE',
    neutralTertiaryAlt: '#C8C6C4',
    neutralTertiary: '#BEBBB8',
    neutralSecondary: '#A19F9D',
    neutralPrimaryAlt: '#8A8886',
    neutralDark: '#605E5C',
    neutralPrimary: '#323130',
    black: '#000' // Black
};
exports.PowerAppsGrayscaleDarkPalette = {
    white: '#1b1a19',
    neutralLighterAlt: '#292828',
    neutralLighter: 'rgba(255, 255, 255, 0.05)',
    neutralLight: '#2d2c2c',
    neutralQuaternaryAlt: '#3b3a3a',
    neutralQuaternary: '#3b3a3a',
    neutralPrimaryAlt: '#f3f2f1',
    neutralDark: '#c8c6c4',
    neutralPrimary: '#f3f2f1',
    black: '#fff' // Black
};
exports.PowerAppsGrayscaleContrastPalette = {
    white: '#000',
    neutralLighterAlt: '#000',
    neutralLighter: 'fff',
    neutralLight: '#000',
    neutralQuaternaryAlt: '#000',
    neutralQuaternary: '#000',
    neutralPrimaryAlt: '#fff',
    neutralDark: '#fff',
    neutralPrimary: '#fff',
    black: '#fff' // Black
};
exports.PowerAutomateGrayscalePalette = {
    black: '#000',
    neutralDark: '#201f1e',
    neutralPrimary: '#323130',
    neutralPrimaryAlt: '#3b3a39',
    neutralSecondary: '#605e5c',
    neutralSecondaryAlt: '#8a8886',
    neutralTertiary: '#a19f9d',
    neutralTertiaryAlt: '#c8c6c4',
    neutralQuaternary: '#d2d0ce',
    neutralQuaternaryAlt: '#e1dfdd',
    neutralLight: '#edebe9',
    neutralLighter: '#f3f2f1',
    neutralLighterAlt: '#faf9f8',
    white: '#fff' // White
};
exports.PowerVirtualAgentGrayscalePalette = {
    black: '#000',
    neutralDark: '#201f1e',
    neutralPrimary: '#323130',
    neutralPrimaryAlt: '#3b3a39',
    neutralSecondary: '#605e5c',
    neutralSecondaryAlt: '#8a8886',
    neutralTertiary: '#a19f9d',
    neutralTertiaryAlt: '#c8c6c4',
    neutralQuaternary: '#d2d0ce',
    neutralQuaternaryAlt: '#e1dfdd',
    neutralLight: '#edebe9',
    neutralLighter: '#f3f2f1',
    neutralLighterAlt: '#faf9f8',
    white: '#fff' // White
};
exports.PowerVirtualAgentGrayscaleDarkPalette = {
    black: '#fff',
    neutralDark: '#faf9f8',
    neutralPrimary: '#f3f2f1',
    neutralPrimaryAlt: '#c8c6c4',
    neutralSecondary: '#a19f9d',
    neutralSecondaryAlt: '#979693',
    neutralTertiary: '#797775',
    neutralTertiaryAlt: '#484644',
    neutralQuaternary: '#3b3a39',
    neutralQuaternaryAlt: '#323130',
    neutralLight: '#292827',
    neutralLighter: '#252423',
    neutralLighterAlt: '#201f1e',
    white: '#1b1a19' // White
};
exports.PowerVirtualAgentGrayscaleContrastPalette = {
    black: '#ffffff',
    neutralDark: '#ffffff',
    neutralPrimary: '#ffffff',
    neutralPrimaryAlt: '#ffffff',
    neutralSecondary: '#ffffff',
    neutralSecondaryAlt: '#ffffff',
    neutralTertiary: '#2ef149',
    neutralTertiaryAlt: '#ffffff',
    neutralQuaternary: '#ffffff',
    neutralQuaternaryAlt: '#ffffff',
    neutralLight: '#ffffff',
    neutralLighter: '#000000',
    neutralLighterAlt: '#000000',
    white: '#000000'
};
exports.DefaultTheme = {
    palette: tslib_1.__assign({ themePrimary: '#0078d4', themeLighterAlt: '#f3f9fd', themeLighter: '#d0e7f8', themeLight: '#a9d3f2', themeTertiary: '#5ca9e5', themeSecondary: '#1a86d9', themeDarkAlt: '#006cbe', themeDark: '#005ba1', themeDarker: '#004377' }, exports.DefaultGrayscalePalette)
};
var DefaultDarkPalette = {
    themePrimary: '#2899f5',
    themeSecondary: '#0078d4',
    themeTertiary: '#235a85',
    themeLight: '#004c87',
    themeLighter: '#043862',
    themeLighterAlt: '#092c47',
    themeDarker: '#82c7ff',
    themeDark: '#6cb8f6',
    themeDarkAlt: '#3aa0f3',
    neutralDark: '#faf9f8',
    neutralPrimary: '#f3f2f1',
    neutralPrimaryAlt: '#c8c6c4',
    neutralSecondary: '#a19f9d',
    neutralSecondaryAlt: '#979693',
    neutralTertiary: '#797775',
    neutralTertiaryAlt: '#484644',
    neutralQuaternary: '#3b3a39',
    neutralQuaternaryAlt: '#323130',
    neutralLight: '#292827',
    neutralLighter: '#252423',
    neutralLighterAlt: '#201f1e',
    black: '#ffffff',
    white: '#1b1a19',
    redDark: '#F1707B'
};
exports.DefaultDarkTheme = {
    palette: DefaultDarkPalette,
    semanticColors: {
        buttonText: DefaultDarkPalette.black,
        buttonTextPressed: DefaultDarkPalette.neutralDark,
        buttonTextHovered: DefaultDarkPalette.neutralPrimary,
        bodySubtext: DefaultDarkPalette.white,
        disabledBackground: DefaultDarkPalette.neutralQuaternaryAlt,
        inputBackgroundChecked: DefaultDarkPalette.themePrimary,
        menuBackground: DefaultDarkPalette.neutralLighter,
        menuItemBackgroundHovered: DefaultDarkPalette.neutralQuaternaryAlt,
        menuItemBackgroundPressed: DefaultDarkPalette.neutralQuaternary,
        menuDivider: DefaultDarkPalette.neutralTertiaryAlt,
        menuIcon: DefaultDarkPalette.themeDarkAlt,
        menuHeader: DefaultDarkPalette.black,
        menuItemText: DefaultDarkPalette.neutralPrimary,
        menuItemTextHovered: DefaultDarkPalette.neutralDark
    }
};
exports.DesignDocsTheme = {
    palette: {
        themePrimary: '#9a4bd8',
        themeSecondary: '#a65fdd',
        themeLighterAlt: '#fbf7fd',
        themeLighter: '#eedff9',
        themeLight: '#dfc4f4',
        themeTertiary: '#c18ee8',
        themeDarkAlt: '#8c44c3',
        themeDark: '#763aa5',
        themeDarker: '#572a79'
    },
    semanticColors: {
        errorBackground: '#fde7e9',
        blockingBackground: '#fed9cc',
        warningBackground: '#fff4ce',
        successBackground: '#dff6dd',
        link: '#a65fdd',
        linkHovered: '#763aa5',
        messageLink: '#a65fdd',
        messageLinkHovered: '#763aa5'
    }
};
exports.PowerAppsTheme = {
    palette: tslib_1.__assign({ themePrimary: '#742774', themeSecondary: '#8d308d', themeTertiary: '#d98fd9', themeLighterAlt: '#faf2fa', themeLighter: '#f6e4f6', themeLight: '#edcaed', themeDarkAlt: '#672367', themeDark: '#501b50', themeDarker: '#3f153f', neutralTertiaryAlt: '#666', neutralTertiary: '#666' }, exports.PowerAppsGrayscalePalette),
    semanticColors: {
        link: '#8D308D',
        linkHovered: '#501b50',
        messageLink: '#8d308d',
        messageLinkHovered: '#501b50'
    }
};
exports.PowerAppsDarkTheme = {
    palette: tslib_1.__assign({ themePrimary: '#742774', themeSecondary: '#8d308d' }, exports.PowerAppsGrayscaleDarkPalette)
};
exports.PowerAppsContrastTheme = {
    palette: tslib_1.__assign({ themePrimary: '#742774', themeSecondary: '#8d308d' }, exports.PowerAppsGrayscaleContrastPalette)
};
exports.PowerBITheme = {
    palette: tslib_1.__assign(tslib_1.__assign({ themePrimary: '#f2c811', themeSecondary: '#f4cf2c', themeTertiary: '#f7de6d', themeLighterAlt: '#fefdf5', themeLighter: '#fdf6d7', themeLight: '#fbeeb5', themeDarkAlt: '#dab50f', themeDark: '#a98c0c', themeDarker: '#7a6816' }, exports.PowerAppsGrayscalePalette), { neutralTertiaryAlt: '#666', neutralTertiary: '#666' }),
    semanticColors: {
        link: '#f4cf2c',
        linkHovered: '#a98c0c',
        messageLink: '#f4cf2c',
        messageLinkHovered: '#a98c0c'
    }
};
exports.PowerAutomateTheme = {
    palette: tslib_1.__assign(tslib_1.__assign({ themePrimary: '#0066ff', themeSecondary: '#1975ff', themeTertiary: '#94bfff', themeLight: '#cce0ff', themeLighter: '#e6f0ff', themeLighterAlt: '#f2f7ff', themeDarker: '#00388c', themeDark: '#0047b3', themeDarkAlt: '#005ce6' }, exports.PowerAppsGrayscalePalette), { neutralTertiaryAlt: '#666', neutralTertiary: '#666' }),
    semanticColors: {
        link: '#1975ff',
        linkHovered: '#0047b3',
        messageLink: '#1975ff',
        messageLinkHovered: '#0047b3'
    }
};
exports.PowerVirtualAgentTheme = (function () {
    var palette = tslib_1.__assign({ themeDarker: '#062d37', themeDark: '#073845', themeDarkAlt: '#0a4a5c', themePrimary: '#0b556a', themeSecondary: '#0f677b', themeTertiary: '#14848f', themeLight: '#73e3df', themeLighter: '#cff6f9', themeLighterAlt: '#f1fbfe', purple: '#672367', blue: '#118DFF', green: '#6BB700', magenta: '#D98FD9', teal: '#73E3DF', purpleLight: '#F0E9F0', blueLight: '#E7F4FF', greenLight: '#F0F8E6' }, exports.PowerVirtualAgentGrayscalePalette);
    return {
        palette: palette,
        semanticColors: {
            link: '#10738e',
            linkHovered: palette.themeDark,
            messageLink: '#10738e',
            messageLinkHovered: palette.themeDark,
            primaryButtonBackground: palette.themePrimary,
            primaryButtonText: palette.white,
            primaryButtonBorder: palette.white,
            primaryButtonBackgroundHovered: palette.themeDarkAlt,
            primaryButtonTextHovered: palette.white,
            primaryButtonBackgroundPressed: '#edebe9',
            primaryButtonTextPressed: palette.white,
            focusBorder: palette.black
        }
    };
})();
exports.PowerVirtualAgentDarkTheme = (function () {
    var palette = tslib_1.__assign({ themeDarker: '#69ebed', themeDark: '#64e0e2', themeDarkAlt: '#34d6d8', themePrimary: '#00c9cc', themeSecondary: '#00b1b4', themeTertiary: '#00787a', themeLight: '#003c3d', themeLighter: '#002021', themeLighterAlt: '#000808', purple: '#D25FD2', blue: '#2899F5', green: '#92C353', magenta: '#7C2E7C', teal: '#00787A', purpleLight: '#3D2A3B', blueLight: '#293F50', greenLight: '#3E4730' }, exports.PowerVirtualAgentGrayscaleDarkPalette);
    return {
        palette: palette,
        semanticColors: {
            link: palette.themePrimary,
            linkHovered: palette.themeDark,
            bodyBackgroundChecked: palette.neutralQuaternaryAlt,
            bodyBackground: palette.white,
            bodyBackgroundHovered: palette.neutralLight,
            bodyTextChecked: palette.neutralPrimary,
            bodyText: palette.black,
            primaryButtonBackground: palette.themePrimary,
            primaryButtonText: palette.white,
            primaryButtonBorder: palette.white,
            primaryButtonBackgroundHovered: palette.themeDarkAlt,
            primaryButtonTextHovered: palette.white,
            primaryButtonBackgroundPressed: palette.themeDark,
            primaryButtonTextPressed: palette.white,
            primaryButtonBackgroundDisabled: palette.neutralLighterAlt,
            primaryButtonTextDisabled: palette.neutralTertiary,
            focusBorder: palette.black,
            inputBackground: palette.white,
            inputText: palette.neutralPrimary,
            inputPlaceholderText: palette.black,
            inputFocusBorderAlt: palette.themePrimary,
            warningIcon: palette.neutralPrimaryAlt,
            warningBackground: '#433519',
            severeWarningIcon: '#FCE100',
            severeWarningBackground: '#4F2A0F',
            errorText: '#F1707B',
            errorIcon: '#F1707B',
            errorBackground: '#442726',
            successIcon: palette.green,
            successBackground: '#393D1B'
        }
    };
})();
exports.PowerVirtualAgentContrastTheme = (function () {
    var palette = tslib_1.__assign({ themeDarker: '#000000', themeDark: '#000000', themeDarkAlt: '#000000', themePrimary: '#000000', themeSecondary: '#000000', themeTertiary: '#000000', themeLight: '#000000', themeLighter: '#000000', themeLighterAlt: '#000000', purple: '#000000', blue: '#000000', green: '#000000', magenta: '#000000', teal: '#000000', purpleLight: '#000000', blueLight: '#000000', greenLight: '#000000' }, exports.PowerVirtualAgentGrayscaleContrastPalette);
    var yellow = '#fffd39';
    var green = '#2ef149';
    var cyan = '#00ebfd';
    return {
        palette: palette,
        semanticColors: {
            link: yellow,
            linkHovered: yellow,
            bodyBackgroundChecked: cyan,
            bodyBackground: palette.white,
            bodyBackgroundHovered: cyan,
            bodyTextChecked: palette.white,
            bodyText: palette.black,
            primaryButtonBackground: palette.black,
            primaryButtonText: palette.white,
            primaryButtonBorder: palette.black,
            primaryButtonBackgroundHovered: cyan,
            primaryButtonTextHovered: palette.white,
            primaryButtonBackgroundPressed: cyan,
            primaryButtonTextPressed: palette.white,
            primaryButtonBackgroundDisabled: green,
            primaryButtonTextDisabled: green,
            focusBorder: palette.black,
            inputBackground: palette.white,
            inputText: palette.black,
            inputPlaceholderText: palette.black,
            inputFocusBorderAlt: palette.black,
            disabledText: green,
            errorText: palette.black,
            errorIcon: palette.black,
            successIcon: palette.black,
            severeWarningIcon: palette.black,
            successBackground: palette.white,
            severeWarningBackground: palette.white,
            warningBackground: palette.white,
            errorBackground: palette.white
        }
    };
})();
exports.Dynamics365Theme = {
    palette: tslib_1.__assign(tslib_1.__assign({ themePrimary: '#2266e3', themeSecondary: '#1f6cf9', themeTertiary: '#9cbeff', themeLight: '#ccdeff', themeLighter: '#e7efff', themeLighterAlt: '#f0f5ff', themeDarker: '#053385', themeDark: '#0742ab', themeDarkAlt: '#0b53ce' }, exports.PowerAppsGrayscalePalette), { neutralTertiaryAlt: '#666', neutralTertiary: '#666' }),
    semanticColors: {
        link: '#1f6cf9',
        linkHovered: '#0742ab',
        messageLink: '#1f6cf9',
        messageLinkHovered: '#0742ab'
    }
};
exports.FlowTheme = {
    palette: tslib_1.__assign(tslib_1.__assign({ themePrimary: '#0066ff', themeSecondary: '#1975ff', themeLighterAlt: '#f2f7ff', themeLighter: '#e6f0ff', themeLight: '#cce0ff', themeTertiary: '#94bfff', themeDarkAlt: '#005ce6' }, exports.PowerAutomateGrayscalePalette), { themeDark: '#0047b3', themeDarker: '#00388c' })
};
// export new palettes with old export to avoid breaking partner teams who have used these values
exports.PowerAppsPurplePalette = exports.PowerAppsTheme.palette;
exports.FlowBluePalette = exports.FlowTheme.palette;
