// @business-app/fabric/lib-cjs/components/FilterSearchViews/examples/FilterSearchViewsExampleData.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldFilterItem = exports.filters = exports.flatItems = exports.groupedItems = exports.charts = exports.arrows = void 0;
exports.arrows = [
    { id: '1', text: 'Up', iconName: 'Up' },
    { id: '2', text: 'Down', iconName: 'Down' },
    { id: '3', text: 'Right', iconName: 'Forward' },
    { id: '4', text: 'Left', iconName: 'Back' },
    { id: '5', text: 'Chevron down', iconName: 'ChevronDown' },
    { id: '6', text: 'Chevron up', iconName: 'ChevronUp' },
    { id: '7', text: 'Chevron left', iconName: 'ChevronLeft' },
    { id: '8', text: 'Chevron right', iconName: 'ChevronRight' }
];
exports.charts = [
    { id: '9', text: 'Chart', iconName: 'Chart' },
    { id: '10', text: 'Area chart', iconName: 'AreaChart' },
    { id: '11', text: 'Pivot chart', iconName: 'PivotChart' },
    { id: '12', text: 'Line chart', iconName: 'LineChart' },
    { id: '13', text: 'Horizontal bar chart', iconName: 'BarChartHorizontal' },
    { id: '14', text: 'Vertical bar chart', iconName: 'BarChartVertical' },
    { id: '15', text: 'Chart series', iconName: 'ChartSeries' },
    { id: '16', text: 'Donut chart', iconName: 'DonutChart' }
];
exports.groupedItems = [
    { id: 'arrowsId', text: 'Arrows', iconName: '', children: exports.arrows },
    { id: 'chartsId', text: 'Charts', iconName: '', children: exports.charts }
];
exports.flatItems = exports.arrows.concat(exports.charts);
exports.filters = [
    { key: 'arrows', text: 'Arrows' },
    { key: 'charts', text: 'Charts' }
];
var shouldFilterItem = function (item, text, filter) {
    // Groups should be visible only if any children match filter
    if (item.children) {
        return item.children.length > 0;
    }
    if (!!text && item.text.toLowerCase().indexOf(text.toLowerCase()) === -1) {
        return false;
    }
    if (!!filter) {
        var list = filter.key === 'arrows' ? exports.arrows : exports.charts;
        for (var i = 0; i < list.length; i++) {
            if (list[i].id === item.id) {
                return true;
            }
        }
        return false;
    }
    return true;
};
exports.shouldFilterItem = shouldFilterItem;
