// @business-app/fabric/src/components/Card/examples/RichCard.With.Custom.Title.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import * as React from 'react';

import { RichCard } from '@business-app/fabric/lib/components/Card/RichCard';
import { FontIcon } from '@fluentui/react/lib/Icon';
import { mergeStyles, mergeStyleSets } from '@fluentui/react/lib/Styling';

const iconClass = mergeStyles({
  fontSize: 50,
  height: 50,
  width: 50
});

export class RichCardWithCustomTitleExample extends React.Component {
  public render() {
    return (
      <RichCard
        height={150}
        width={300}
        headerProps={{
          onRenderTitle: () => {
            return (
              <>
                <FontIcon iconName="FavoriteStarFill" />
                <FontIcon iconName="FavoriteStarFill" />
                <FontIcon iconName="FavoriteStarFill" />
                <FontIcon iconName="FavoriteStar" />
                <FontIcon iconName="FavoriteStar" />
              </>
            );
          }
        }}
      >
        Title replaced with stars
      </RichCard>
    );
  }
}
