// @business-app/fabric/src/components/Card/examples/RichCard.With.Sub.Components.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import * as React from 'react';

import { ContextualMenuItemType } from '@fluentui/react/lib/ContextualMenu';

import { ActionItemType, RichCard } from '@business-app/fabric/lib/components/Card';

export class RichCardWithSubComponentsExample extends React.Component {
  public render() {
    return (
      <RichCard
        headerProps={{
          title: 'Card title',
          subTitle: 'Card subtitle',
          titleIconProps: {
            iconName: 'Dictionary'
          },
          subTitleIconProps: {
            iconName: 'Dictionary'
          },
          actionItemProps: [
            {
              type: ActionItemType.Link,
              props: {
                href: 'https"//bing.com',
                target: '_blank',
                title: 'Hyperlink'
              }
            },
            {
              type: ActionItemType.Button,
              props: {
                iconProps: {
                  iconName: 'Attach'
                }
              }
            },
            {
              type: ActionItemType.Button,
              props: {
                iconProps: {
                  iconName: 'Add'
                }
              }
            },
            {
              type: ActionItemType.ContextualMenu,
              props: {
                text: 'Content',
                menuProps: {
                  items: [
                    {
                      key: 'emailMessage',
                      text: 'Email message',
                      iconProps: { iconName: 'Mail' }
                    },
                    {
                      key: 'calendarEvent',
                      text: 'Calendar event',
                      iconProps: { iconName: 'Calendar' }
                    }
                  ]
                }
              }
            }
          ],
          contextualMenuProps: {
            items: [
              {
                key: 'newItem',
                text: 'New',
                onClick: () => alert('New clicked')
              },
              {
                key: 'divider_1',
                itemType: ContextualMenuItemType.Divider
              },
              {
                key: 'rename',
                text: 'Rename',
                onClick: () => alert('Rename clicked')
              },
              {
                key: 'edit',
                text: 'Edit',
                onClick: () => alert('Edit clicked')
              },
              {
                key: 'properties',
                text: 'Properties',
                onClick: () => alert('Properties clicked')
              },
              {
                key: 'linkNoTarget',
                text: 'Link same window',
                href: 'http://bing.com'
              }
            ]
          }
        }}
        footerProps={{
          link: {
            text: 'Hyperlink',
            href: 'https://bing.com'
          },
          buttons: {
            primaryButton: {
              text: 'First Butjon'
            },
            secondaryButton: {
              text: 'Second Butjon'
            }
          }
        }}
      >
        <div style={{ width: '100%', height: '100%', backgroundColor: '#EEE' }}>
          <img style={{ width: '100%' }} src={'https://placehold.it/600x400/EEE'} />
        </div>
      </RichCard>
    );
  }
}
