// @business-app/fabric/lib/components/Carousel/examples/Carousel.ButtonClick.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { Carousel } from '..';
import { getMockContent } from './Carousel.Example.Data';
var CarouselButtonClickExample = /** @class */ (function (_super) {
    __extends(CarouselButtonClickExample, _super);
    function CarouselButtonClickExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onSkip = function () {
            var skipCount = _this.state.skipCount;
            _this.setState({ skipCount: skipCount + 1 });
        };
        _this._onDismiss = function () {
            var doneCount = _this.state.doneCount;
            _this.setState({ doneCount: doneCount + 1 });
        };
        _this._onNext = function () {
            var nextCount = _this.state.nextCount;
            _this.setState({ nextCount: nextCount + 1 });
        };
        _this._onPrevious = function () {
            var previousCount = _this.state.previousCount;
            _this.setState({ previousCount: previousCount + 1 });
        };
        _this.state = {
            skipCount: 0,
            doneCount: 0,
            nextCount: 0,
            previousCount: 0
        };
        return _this;
    }
    CarouselButtonClickExample.prototype.render = function () {
        var _a = this.state, skipCount = _a.skipCount, doneCount = _a.doneCount, nextCount = _a.nextCount, previousCount = _a.previousCount;
        var customCarouselStyles = {
            root: {
                flexDirection: 'column'
            },
            subComponentStyles: {
                tooltipControl: {
                    root: {
                        position: 'initial',
                        transform: 'initial',
                        marginRight: 8
                    }
                },
                control: {
                    root: {
                        borderRadius: 'initial',
                        paddingLeft: 24,
                        paddingRight: 24,
                        selectors: {
                            '&.ba-Carousel-control-next': {
                                borderRadius: 2
                            },
                            '&.ba-Carousel-control-previous': {
                                borderRadius: 2
                            }
                        }
                    }
                },
                indicator: {},
                toggle: {}
            },
            indicatorsWrapper: {
                position: 'initial',
                transform: 'initial'
            },
            controlsWrapper: {
                margin: 24
            }
        };
        return (React.createElement(React.Fragment, null,
            React.createElement("div", null, "Skip button count: " + skipCount),
            React.createElement("div", null, "Done button count: " + doneCount),
            React.createElement("div", null, "Next button count: " + nextCount),
            React.createElement("div", null, "Previous button count: " + previousCount),
            React.createElement(Carousel, { styles: customCarouselStyles, items: getMockContent(), useControlLabels: true, disableNavigationLoop: true, navPreviousButtonAriaLabel: "Previous", navNextButtonAriaLabel: "Next", indicatorButtonsAriaLabel: ['Item 1', 'Item 2', 'Item 3', 'Item 4'], playButtonAriaLabel: "Play", pauseButtonAriaLabel: "Pause", skipButton: {
                    onSkip: this._onSkip,
                    skipButtonAriaLabel: 'Skip',
                    skipButtonLabel: 'Skip'
                }, doneButton: {
                    onDone: this._onDismiss,
                    doneButtonAriaLabel: 'Done',
                    doneButtonLabel: 'Done'
                }, onNext: this._onNext, onPrevious: this._onPrevious, orderIndicatorsBeforeControls: true })));
    };
    return CarouselButtonClickExample;
}(React.Component));
export { CarouselButtonClickExample };
