// @business-app/fabric/lib/components/Carousel/examples/Carousel.ControlButtons.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { Carousel } from '..';
import { getMockContent } from './Carousel.Example.Data';
var CarouselControlButtonsExample = /** @class */ (function (_super) {
    __extends(CarouselControlButtonsExample, _super);
    function CarouselControlButtonsExample(props) {
        return _super.call(this, props) || this;
    }
    CarouselControlButtonsExample.prototype.render = function () {
        var customCarouselStyles = {
            root: {
                flexDirection: 'column'
            },
            subComponentStyles: {
                tooltipControl: {
                    root: {
                        position: 'initial',
                        transform: 'initial',
                        marginRight: 8
                    }
                },
                control: {
                    root: {
                        borderRadius: 'initial'
                    }
                },
                indicator: {},
                toggle: {}
            },
            indicatorsWrapper: {
                position: 'initial',
                transform: 'initial'
            },
            controlsWrapper: {
                margin: 24
            }
        };
        return (React.createElement(Carousel, { styles: customCarouselStyles, items: getMockContent(), useControlLabels: true, disableNavigationLoop: true, navPreviousButtonAriaLabel: "Previous", navNextButtonAriaLabel: "Next", indicatorButtonsAriaLabel: ['Item 1', 'Item 2', 'Item 3', 'Item 4'], playButtonAriaLabel: "Play", pauseButtonAriaLabel: "Pause", orderIndicatorsBeforeControls: true }));
    };
    return CarouselControlButtonsExample;
}(React.Component));
export { CarouselControlButtonsExample };
