// @business-app/fabric/lib-cjs/components/Resizer/Resizer.types.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResizerAttachment = exports.ResizeDirection = exports.ResizerLocation = void 0;
/**
 * Specifies the location of the {@link Resizer} with respect to the resizable element.
 */
var ResizerLocation;
(function (ResizerLocation) {
    ResizerLocation[ResizerLocation["Top"] = 0] = "Top";
    ResizerLocation[ResizerLocation["Right"] = 1] = "Right";
    ResizerLocation[ResizerLocation["Bottom"] = 2] = "Bottom";
    ResizerLocation[ResizerLocation["Left"] = 3] = "Left";
})(ResizerLocation = exports.ResizerLocation || (exports.ResizerLocation = {}));
/**
 * Specifies whether the resize direction is horizontal or vertical.
 */
var ResizeDirection;
(function (ResizeDirection) {
    ResizeDirection[ResizeDirection["Vertical"] = 0] = "Vertical";
    ResizeDirection[ResizeDirection["Horizontal"] = 1] = "Horizontal";
})(ResizeDirection = exports.ResizeDirection || (exports.ResizeDirection = {}));
/**
 * Specifies if the {@link Resizer} handler attaches inside or outside the edge of the element to be resized.
 */
var ResizerAttachment;
(function (ResizerAttachment) {
    ResizerAttachment[ResizerAttachment["Inside"] = 0] = "Inside";
    ResizerAttachment[ResizerAttachment["Outside"] = 1] = "Outside";
})(ResizerAttachment = exports.ResizerAttachment || (exports.ResizerAttachment = {}));
