// @business-app/fabric/lib-cjs/components/Resizer/Resizer.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
var getStyles = function (props) {
    var palette = props.theme.palette;
    var activeResizer = {
        opacity: 1,
        backgroundColor: palette.neutralTertiary
    };
    return {
        root: {
            position: 'absolute',
            backgroundColor: palette.neutralSecondary,
            borderRadius: 0,
            transition: '0.5s opacity, 0.5s background-color',
            opacity: 0,
            zIndex: Styling_1.ZIndexes.Layer,
            selectors: {
                ':hover': {
                    transition: '0.2s opacity, 0.2s background-color',
                    opacity: 1
                },
                ':focus': activeResizer,
                ':active': activeResizer
            }
        }
    };
};
exports.getStyles = getStyles;
