// @business-app/fabric/src/components/TreeView/examples/TreeView.CustomStyles.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IContextualMenuItem } from '@fluentui/react/lib/ContextualMenu';
import * as React from 'react';
import { TreeView } from '../TreeView.base';
import { ITreeStyles, ITreeViewItem } from '../TreeView.types';

export class TreeViewCustomStylesExample extends React.Component {
  public render(): JSX.Element {
    const items: ITreeViewItem[] = [
      {
        label: 'Group1',
        key: '0',
        children: [
          {
            label: 'Group1-1',
            key: '1',
            children: [
              {
                label: 'Item 1-1-1',
                key: '2'
              },
              {
                label: 'Item 1-1-2',
                key: '3'
              },
              {
                label: 'Item 1-1-3',
                key: '4'
              }
            ]
          },
          {
            label: 'Item 1-1',
            key: '5'
          },
          {
            label: 'Item 1-2',
            key: '6'
          }
        ]
      },
      {
        label: 'Group2',
        key: '7',
        children: [
          {
            label: 'Item 2-1',
            key: '8'
          },
          {
            label: 'Item 2-2',
            key: '9'
          }
        ]
      }
    ];

    const noBackground: Partial<ITreeStyles> = {
      root: { backgroundColor: 'transparent' }
    };

    return (
      <div style={{ width: '220px', height: '320px', overflow: 'auto', userSelect: 'none' }}>
        <TreeView
          items={items}
          strings={{
            listAriaLabel: 'list',
            moreCommandsAriaLabel: 'more commands'
          }}
          getItemCommands={this._getCommands}
          styles={noBackground}
        />
      </div>
    );
  }

  private _getCommands = (): IContextualMenuItem[] => {
    return [
      {
        key: 'New',
        text: 'New',
        iconProps: { iconName: 'Add' }
      },
      {
        key: 'Mail',
        text: 'Mail',
        iconProps: { iconName: 'Mail' }
      },
      {
        key: 'Calendar',
        text: 'Calendar',
        iconProps: { iconName: 'Calendar' }
      }
    ];
  };
}
