// @business-app/fabric/lib/components/TreeView/examples/TreeView.CustomStyles.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { TreeView } from '../TreeView.base';
var TreeViewCustomStylesExample = /** @class */ (function (_super) {
    __extends(TreeViewCustomStylesExample, _super);
    function TreeViewCustomStylesExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._getCommands = function () {
            return [
                {
                    key: 'New',
                    text: 'New',
                    iconProps: { iconName: 'Add' }
                },
                {
                    key: 'Mail',
                    text: 'Mail',
                    iconProps: { iconName: 'Mail' }
                },
                {
                    key: 'Calendar',
                    text: 'Calendar',
                    iconProps: { iconName: 'Calendar' }
                }
            ];
        };
        return _this;
    }
    TreeViewCustomStylesExample.prototype.render = function () {
        var items = [
            {
                label: 'Group1',
                key: '0',
                children: [
                    {
                        label: 'Group1-1',
                        key: '1',
                        children: [
                            {
                                label: 'Item 1-1-1',
                                key: '2'
                            },
                            {
                                label: 'Item 1-1-2',
                                key: '3'
                            },
                            {
                                label: 'Item 1-1-3',
                                key: '4'
                            }
                        ]
                    },
                    {
                        label: 'Item 1-1',
                        key: '5'
                    },
                    {
                        label: 'Item 1-2',
                        key: '6'
                    }
                ]
            },
            {
                label: 'Group2',
                key: '7',
                children: [
                    {
                        label: 'Item 2-1',
                        key: '8'
                    },
                    {
                        label: 'Item 2-2',
                        key: '9'
                    }
                ]
            }
        ];
        var noBackground = {
            root: { backgroundColor: 'transparent' }
        };
        return (React.createElement("div", { style: { width: '220px', height: '320px', overflow: 'auto', userSelect: 'none' } },
            React.createElement(TreeView, { items: items, strings: {
                    listAriaLabel: 'list',
                    moreCommandsAriaLabel: 'more commands'
                }, getItemCommands: this._getCommands, styles: noBackground })));
    };
    return TreeViewCustomStylesExample;
}(React.Component));
export { TreeViewCustomStylesExample };
