// @business-app/fabric/lib-cjs/components/Tree/examples/AsyncTree.PaginationExample.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncTreePaginationExample = void 0;
var tslib_1 = require("tslib");
var Toggle_1 = require("@fluentui/react/lib/Toggle");
var Styling_1 = require("@fluentui/react/lib/Styling");
var React = require("react");
var AsyncTree_1 = require("../AsyncTree");
var wait = function (ms) { return new Promise(function (r, j) { return setTimeout(r, ms); }); };
var strings = {
    listAriaLabel: 'list',
    moreCommandsAriaLabel: 'more commands'
};
var AsyncTreePaginationExample = function () {
    var throwErrorsRef = React.useRef(false);
    var onResolveItems = React.useCallback(function (item, continuationToken) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var startIndex, endIndex, items, i, i;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, wait(1500)];
                case 1:
                    _a.sent();
                    if (throwErrorsRef.current) {
                        throw new Error('this is an error');
                    }
                    startIndex = continuationToken ? parseInt(continuationToken, 10) : 0;
                    endIndex = startIndex + 5;
                    items = [];
                    if (item === null) {
                        for (i = startIndex; i < endIndex; i++) {
                            items.push({
                                key: i.toString(),
                                label: i.toString()
                            });
                        }
                    }
                    else {
                        for (i = startIndex; i < endIndex; i++) {
                            items.push({
                                key: item.key + '/' + i.toString(),
                                label: item.label + '/' + i.toString(),
                                isFolder: i % 2 === 0
                            });
                        }
                    }
                    return [2 /*return*/, [items, endIndex.toString()]];
            }
        });
    }); }, []);
    var onToggleChange = React.useCallback(function (ev, checked) {
        throwErrorsRef.current = checked;
    }, []);
    return (React.createElement(React.Fragment, null,
        React.createElement("div", null,
            React.createElement(Toggle_1.Toggle, { label: 'Throw errors', defaultChecked: throwErrorsRef.current, onChange: onToggleChange })),
        React.createElement("div", { className: exampleStyles.exampleContainer },
            React.createElement(AsyncTree_1.AsyncTree, { onResolveItems: onResolveItems, strings: strings }))));
};
exports.AsyncTreePaginationExample = AsyncTreePaginationExample;
var exampleStyles = Styling_1.mergeStyleSets({
    exampleContainer: {
        // 'display: flex' is important for the Tree to layout correctly.
        display: 'flex',
        // Just an example to prove that row direction also works for the Tree.
        flexDirection: 'row',
        width: 320,
        height: 200,
        userSelect: 'none'
    }
});
