// @business-app/fabric/lib-cjs/components/Tree/examples/Tree.Example.Data.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.cloneSet = exports.getFolderKeys = exports.produceItemsWithIcons = exports.produceItems = void 0;
var tslib_1 = require("tslib");
exports.produceItems = [
    {
        label: 'Fruits',
        key: 'fruits',
        isFolder: true,
        level: 0,
        setSize: 2,
        indexInParent: 0
    },
    {
        label: 'Apples',
        key: 'apples',
        isFolder: true,
        level: 1,
        setSize: 0,
        indexInParent: 0
    },
    {
        label: 'Granny Smith',
        key: 'granny smith',
        level: 2,
        setSize: 2,
        indexInParent: 0
    },
    {
        label: 'Red Delicious',
        key: 'red delicious',
        level: 2,
        setSize: 2,
        indexInParent: 1
    },
    {
        label: 'Vegetables',
        key: 'vegetables',
        isFolder: true,
        level: 0,
        setSize: 2,
        indexInParent: 1
    },
    {
        label: 'Carrots',
        key: 'carrots',
        level: 1,
        setSize: 3,
        indexInParent: 0
    },
    {
        label: 'Cucumbers',
        key: 'cucumbers',
        level: 1,
        setSize: 3,
        indexInParent: 1
    },
    {
        label: 'Lettuce',
        key: 'lettuce',
        level: 1,
        setSize: 3,
        indexInParent: 2
    }
];
exports.produceItemsWithIcons = exports.produceItems.map(function (item) {
    return tslib_1.__assign(tslib_1.__assign({}, item), { iconProps: { iconName: 'Info' } });
});
function getFolderKeys(items) {
    var folderKeys = new Set();
    items.forEach(function (item) {
        if (item.isFolder) {
            folderKeys.add(item.key);
        }
    });
    return folderKeys;
}
exports.getFolderKeys = getFolderKeys;
function cloneSet(set) {
    var newSet = new Set();
    set.forEach(function (key) { return newSet.add(key); });
    return newSet;
}
exports.cloneSet = cloneSet;
