// @business-app/fabric/lib-cjs/components/Tree/examples/Tree.OverflowAndLayout.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreeOverflowAndLayoutExample = exports.CommandSetting = void 0;
var tslib_1 = require("tslib");
var Button_1 = require("@fluentui/react/lib/Button");
var Announced_1 = require("@fluentui/react/lib/Announced");
var Dropdown_1 = require("@fluentui/react/lib/Dropdown");
var Styling_1 = require("@fluentui/react/lib/Styling");
var TextField_1 = require("@fluentui/react/lib/TextField");
var Toggle_1 = require("@fluentui/react/lib/Toggle");
var React = require("react");
var Tree_1 = require("../Tree");
var Tree_Example_Data_1 = require("./Tree.Example.Data");
var CommandSetting;
(function (CommandSetting) {
    CommandSetting[CommandSetting["None"] = 0] = "None";
    CommandSetting[CommandSetting["Some"] = 1] = "Some";
    CommandSetting[CommandSetting["All"] = 2] = "All";
})(CommandSetting = exports.CommandSetting || (exports.CommandSetting = {}));
var TreeOverflowAndLayoutExample = /** @class */ (function (_super) {
    tslib_1.__extends(TreeOverflowAndLayoutExample, _super);
    function TreeOverflowAndLayoutExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onClickExpandAll = function () {
            _this.setState(function (prev) {
                return {
                    expandedItemKeys: Tree_Example_Data_1.getFolderKeys(prev.items)
                };
            });
            _this.setState({ announcedMessage: 'All items expanded' });
        };
        _this._onClickCollapseAll = function () {
            _this.setState({ expandedItemKeys: new Set(), announcedMessage: 'All items collapsed' });
        };
        _this._onItemExpand = function (item, index) {
            _this.setState(function (prev) {
                var set = Tree_Example_Data_1.cloneSet(prev.expandedItemKeys);
                set.add(item.key);
                return { expandedItemKeys: set };
            });
        };
        _this._onItemCollapsed = function (item, index) {
            _this.setState(function (prev) {
                var set = Tree_Example_Data_1.cloneSet(prev.expandedItemKeys);
                set.delete(item.key);
                return { expandedItemKeys: set };
            });
        };
        _this._getItemHasCommands = function (item) {
            var commands = _this.state.commands;
            if (commands === CommandSetting.None) {
                return false;
            }
            else if (commands === CommandSetting.All) {
                return true;
            }
            else {
                return item.label.toLowerCase().indexOf('a') !== -1;
            }
        };
        _this._getCommands = function () {
            var selection = _this.state.selection;
            var keys = ['newItem', 'share', 'mobile'];
            var items = [
                {
                    key: keys[0],
                    text: 'News',
                    canCheck: true,
                    isChecked: selection[keys[0]],
                    onClick: _this._onToggleSelect,
                    icon: 'Add'
                },
                {
                    key: keys[1],
                    text: 'Share',
                    canCheck: true,
                    isChecked: selection[keys[1]],
                    onClick: _this._onToggleSelect,
                    icon: 'Mail'
                },
                {
                    key: keys[2],
                    text: 'Mobile',
                    canCheck: true,
                    isChecked: selection[keys[2]],
                    onClick: _this._onToggleSelect,
                    icon: 'Calendar'
                }
            ];
            return items;
        };
        _this._onToggleSelect = function (ev, item) {
            var selection = _this.state.selection;
            if (ev.preventDefault) {
                ev.preventDefault();
            }
            selection[item.key] = !selection[item.key];
            _this.setState({
                selection: selection
            });
        };
        _this.state = {
            commands: CommandSetting.All,
            containerWidth: '280px',
            containerHeight: '260px',
            longText: true,
            items: Tree_Example_Data_1.produceItemsWithIcons,
            longItems: Tree_Example_Data_1.produceItemsWithIcons.map(function (item) { return (tslib_1.__assign(tslib_1.__assign({}, item), { label: item.label.indexOf('Cucumber') >= 0
                    ? 'CucumbersCucumbersCucumbersCucumbersCucumbers CucumbersCucumbersCucumbersCucumbersCucumbers'
                    : item.label })); }),
            expandedItemKeys: new Set(),
            selection: {},
            announcedMessage: ''
        };
        return _this;
    }
    TreeOverflowAndLayoutExample.prototype.render = function () {
        var _this = this;
        var _a = this.state, containerWidth = _a.containerWidth, containerHeight = _a.containerHeight, items = _a.items, longItems = _a.longItems, expandedItemKeys = _a.expandedItemKeys, longText = _a.longText;
        return (React.createElement("div", { className: exampleStyles.root },
            React.createElement("div", { className: exampleStyles.optionRow },
                React.createElement(TextField_1.TextField, { className: exampleStyles.option, onChange: function (ev, value) { return _this.setState({ containerWidth: value }); }, label: 'Container Width (eg 100% or 280px)', defaultValue: containerWidth }),
                React.createElement(TextField_1.TextField, { className: exampleStyles.option, onChange: function (ev, value) { return _this.setState({ containerHeight: value }); }, label: 'Container Height (eg 100% or 320px)', defaultValue: containerHeight })),
            React.createElement("div", { className: exampleStyles.optionRow },
                React.createElement(Dropdown_1.Dropdown, { label: 'Contextual menu buttons on rows', defaultSelectedKey: CommandSetting.All, className: exampleStyles.option, options: [
                        {
                            key: CommandSetting.None,
                            text: 'None'
                        },
                        {
                            key: CommandSetting.Some,
                            text: 'Some'
                        },
                        {
                            key: CommandSetting.All,
                            text: 'All'
                        }
                    ], onChange: function (ev, option) { return _this.setState({ commands: option.key }); } }),
                React.createElement(Toggle_1.Toggle, { className: exampleStyles.option, onChange: function (ev, checked) { return _this.setState({ longText: checked }); }, defaultChecked: longText, onText: 'On', offText: 'Off', label: 'Long text' })),
            React.createElement("div", { style: { width: containerWidth, height: containerHeight, border: '1px solid gray' }, className: exampleStyles.treeContainer },
                React.createElement("div", { className: exampleStyles.optionRow },
                    React.createElement(Button_1.DefaultButton, { onClick: this._onClickExpandAll }, 'Expand all'),
                    React.createElement(Button_1.DefaultButton, { onClick: this._onClickCollapseAll }, 'Collapse all')),
                React.createElement(Tree_1.Tree, { items: longText ? longItems : items, expandedItemKeys: expandedItemKeys, strings: {
                        listAriaLabel: 'test',
                        moreCommandsAriaLabel: 'More Command'
                    }, onItemExpand: this._onItemExpand, onItemCollapse: this._onItemCollapsed, getItemCommands: this._getCommands, getItemHasCommands: this._getItemHasCommands })),
            React.createElement(Announced_1.Announced, { message: this.state.announcedMessage })));
    };
    return TreeOverflowAndLayoutExample;
}(React.Component));
exports.TreeOverflowAndLayoutExample = TreeOverflowAndLayoutExample;
var exampleStyles = Styling_1.mergeStyleSets({
    root: {
        userSelect: 'none'
    },
    option: {
        marginRight: 32
    },
    optionRow: {
        display: 'flex',
        flexDirection: 'row'
    },
    treeContainer: {
        display: 'flex',
        flexDirection: 'column'
    }
});
