// @business-app/fabric/lib-cjs/components/Tree/examples/Tree.BasicUncontrolled.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreeBasicUncontrolledExample = void 0;
var tslib_1 = require("tslib");
var Styling_1 = require("@fluentui/react/lib/Styling");
var React = require("react");
var Tree_1 = require("../Tree");
var Tree_Example_Data_1 = require("./Tree.Example.Data");
var TreeBasicUncontrolledExample = /** @class */ (function (_super) {
    tslib_1.__extends(TreeBasicUncontrolledExample, _super);
    function TreeBasicUncontrolledExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TreeBasicUncontrolledExample.prototype.render = function () {
        var items = Tree_Example_Data_1.produceItems;
        var folderKeysSet = Tree_Example_Data_1.getFolderKeys(items);
        return (React.createElement("div", { className: exampleStyles.exampleContainer },
            React.createElement(Tree_1.Tree, { items: items, initialExpandedItemKeys: folderKeysSet, strings: {
                    listAriaLabel: 'list',
                    moreCommandsAriaLabel: 'more commands'
                } })));
    };
    return TreeBasicUncontrolledExample;
}(React.Component));
exports.TreeBasicUncontrolledExample = TreeBasicUncontrolledExample;
var exampleStyles = Styling_1.mergeStyleSets({
    exampleContainer: {
        // 'display: flex' is important for the Tree to layout correctly.
        display: 'flex',
        // Just an example to prove that row direction also works for the Tree.
        flexDirection: 'row',
        width: 320,
        height: 200,
        userSelect: 'none'
    }
});
