// @business-app/fabric/lib-cjs/components/Presence/PersonaAccordionPanel.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPersonaAccordionPanelStyles = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
var getPersonaAccordionPanelStyles = function (_a) {
    var semanticColors = _a.theme.semanticColors;
    var commonIconImg = {
        fontSize: Styling_1.FontSizes.medium,
        verticalAlign: 'middle',
        marginRight: 5,
        marginLeft: 5,
        color: semanticColors.link
    };
    return {
        PersonaActionsStack: {
            overflow: 'hidden'
        },
        personaTitleButton: {
            padding: 0,
            border: 'none',
            height: 56,
            paddingLeft: 14,
            paddingRight: 8,
            width: '100%',
            '.ms-Button-flexContainer': {
                justifyContent: 'space-between'
            }
        },
        separator: {
            padding: 0,
            lineHeight: 0,
            width: '100%'
        },
        actionButton: {
            padding: 0,
            border: 'none',
            width: '100%',
            paddingLeft: 5
        },
        icon: commonIconImg,
        teamsIcon: [commonIconImg, { color: semanticColors.link }],
        actionButtonText: {
            marginLeft: 5,
            marginRight: 5,
            width: '100%',
            textAlign: 'left',
            overflow: 'hidden',
            whiteSpace: 'nowrap',
            textOverflow: 'ellipsis'
        },
        teamsActionButtonSpan: {
            width: '100%'
        },
        TeamsMenuStack: {
            overflow: 'hidden'
        }
    };
};
exports.getPersonaAccordionPanelStyles = getPersonaAccordionPanelStyles;
