// @business-app/fabric/lib-cjs/components/Presence/Presence.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Presence = exports.PersonaCoinForWithAccordionBase = exports.PresencePersona = exports.PresenceBase = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var PersonaAccordionPanel_1 = require("./PersonaAccordionPanel");
var Presence_styles_1 = require("./Presence.styles");
var Utilities_1 = require("@fluentui/react/lib/Utilities");
var Facepile_1 = require("@fluentui/react/lib/Facepile");
var Persona_1 = require("@fluentui/react/lib/Persona");
var Callout_1 = require("@fluentui/react/lib/Callout");
var FocusZone_1 = require("@fluentui/react/lib/FocusZone");
var List_1 = require("@fluentui/react/lib/List");
var Tooltip_1 = require("@fluentui/react/lib/Tooltip");
var Button_1 = require("@fluentui/react/lib/Button");
var getClassNames = Utilities_1.classNamesFunction();
var DefaultMaxDisplayableUsers = 2;
var PresenceBase = function (props) {
    var _a;
    var _b = React.useState(false), showCallout = _b[0], setShowCallout = _b[1];
    var _c = React.useState(undefined), overFlowButtonTarget = _c[0], setOverFlowButtonTarget = _c[1];
    var presentPersonas = props.presentUsers.map(function (user) { return ({ data: user }); });
    var maxDisplayableUsers = props.maxDisplayableUsers || DefaultMaxDisplayableUsers;
    var styles = getClassNames(props.styles, {
        theme: props.theme,
        showAccordion: props.showAccordion
    });
    var onRenderPersona = props.showAccordion
        ? onRenderPersonaCoinForWithAccordion(props.strings, props.signedInUser)
        : onRenderPersonaCoinForInlineAuthorPersona();
    var facepileProps = {
        personas: presentPersonas.slice(0, maxDisplayableUsers),
        overflowPersonas: presentPersonas.slice(maxDisplayableUsers),
        personaSize: Persona_1.PersonaSize.size24,
        maxDisplayablePersonas: maxDisplayableUsers,
        // Overriding OnRenderPersonaCoin doesn't work when there is only one item. onRenderPersona is a way around that.
        onRenderPersona: onRenderPersona,
        onRenderPersonaCoin: onRenderPersona,
        overflowButtonProps: (_a = {},
            _a['aria-haspopup'] = true,
            _a.ariaLabel = props.strings.overflowButtonAriaLabel,
            _a),
        ariaLabel: props.strings.ariaLabel,
        overflowButtonType: Facepile_1.OverflowButtonType.descriptive
    };
    if (facepileProps.overflowPersonas.length > 0) {
        facepileProps.overflowButtonProps.onClick = function (e) {
            e.stopPropagation();
            setOverFlowButtonTarget(e.currentTarget);
            setShowCallout(!showCallout);
        };
    }
    return (React.createElement(React.Fragment, null,
        React.createElement(Facepile_1.Facepile, tslib_1.__assign({}, facepileProps)),
        facepileProps.overflowPersonas.length > 0 && (React.createElement(Callout_1.Callout, { hidden: !showCallout, role: 'dialog', isBeakVisible: true, target: overFlowButtonTarget, onDismiss: function () { return setShowCallout(false); }, directionalHint: Callout_1.DirectionalHint.bottomCenter, setInitialFocus: true },
            React.createElement(FocusZone_1.FocusZone, { direction: FocusZone_1.FocusZoneDirection.vertical, className: styles.overFlowListContent },
                React.createElement(List_1.List, { items: facepileProps.overflowPersonas, onRenderCell: onRenderPersonaCoinForWithAccordion(props.strings, props.signedInUser, true, props.showAccordion) }))))));
};
exports.PresenceBase = PresenceBase;
var onRenderPersonaCoinForInlineAuthorPersona = function () { return function (props) {
    return React.createElement(exports.PresencePersona, tslib_1.__assign({}, props.data));
}; };
var PresencePersona = function (props) {
    var personaCommonProps = {
        imageUrl: props.imageUrl,
        text: props.displayName
    };
    var tooltipProps = {
        onRenderContent: function () { return (React.createElement(Persona_1.Persona, tslib_1.__assign({}, personaCommonProps, { size: Persona_1.PersonaSize.size32, showSecondaryText: true, secondaryText: props.presenceContext }))); }
    };
    return (React.createElement(Tooltip_1.TooltipHost, { key: props.key, tooltipProps: tooltipProps, delay: Tooltip_1.TooltipDelay.zero, directionalHint: Callout_1.DirectionalHint.bottomCenter },
        React.createElement(Persona_1.Persona, tslib_1.__assign({}, personaCommonProps, { hidePersonaDetails: true, size: Persona_1.PersonaSize.size24, key: props.key }))));
};
exports.PresencePersona = PresencePersona;
var onRenderPersonaCoinForWithAccordion = function (strings, signedInUser, isOverflowItem, isCollapsible) { return function (props) {
    var authorPersonaProps = {
        presentUser: props.data,
        strings: strings,
        signedInUser: signedInUser,
        isOverflowItem: isOverflowItem,
        isCollapsible: isCollapsible
    };
    return React.createElement(PersonaCoinForWithAccordion, tslib_1.__assign({}, authorPersonaProps));
}; };
var PersonaCoinForWithAccordionBase = function (props) {
    var _a;
    var _b = React.useState(false), showCallout = _b[0], setShowCallout = _b[1];
    var personaRef = React.useRef(null);
    var personaAccordionPanelProps = {
        showActionItemsAlways: !props.isOverflowItem,
        isCollapsible: props.isCollapsible,
        presentUser: props.presentUser,
        signedInUser: props.signedInUser,
        strings: props.strings
    };
    var styles = getClassNames(props.styles, {
        theme: props.theme,
        showAccordion: false,
        isSelected: showCallout
    });
    return props.isOverflowItem ? (React.createElement(PersonaAccordionPanel_1.PersonaAccordionPanel, tslib_1.__assign({}, personaAccordionPanelProps))) : (React.createElement(React.Fragment, null,
        React.createElement("div", { ref: personaRef, key: props.presentUser.key },
            React.createElement(Button_1.DefaultButton, { key: props.presentUser.key, "aria-haspopup": showCallout, className: styles.commandBarPersonaButton, onClick: function () { return setShowCallout(!showCallout); }, ariaLabel: props.presentUser.displayName },
                React.createElement(Persona_1.Persona, { imageUrl: props.presentUser.imageUrl, text: props.presentUser.displayName, hidePersonaDetails: true, size: Persona_1.PersonaSize.size24, key: (_a = props.presentUser) === null || _a === void 0 ? void 0 : _a.key }))),
        React.createElement(Callout_1.Callout, { hidden: !showCallout, role: 'dialog', isBeakVisible: true, target: personaRef.current, onDismiss: function () { return setShowCallout(false); }, directionalHint: Callout_1.DirectionalHint.bottomCenter, setInitialFocus: true },
            React.createElement(FocusZone_1.FocusZone, { direction: FocusZone_1.FocusZoneDirection.vertical, role: "menu" },
                React.createElement(PersonaAccordionPanel_1.PersonaAccordionPanel, tslib_1.__assign({}, personaAccordionPanelProps))))));
};
exports.PersonaCoinForWithAccordionBase = PersonaCoinForWithAccordionBase;
var PersonaCoinForWithAccordion = Utilities_1.styled(exports.PersonaCoinForWithAccordionBase, Presence_styles_1.getPresenceStyles);
/**
 * Facepile for showing users present on a resource, with additional context and contact info.
 */
exports.Presence = Utilities_1.styled(exports.PresenceBase, Presence_styles_1.getPresenceStyles);
