// @business-app/fabric/src/components/Wizard/TipsAndHelpCard/TipsAndHelpCard.types.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IStyle, ITheme } from '@fluentui/react/lib/Styling';
import { IPropsWithStyles } from '@fluentui/react/lib/Utilities';
import { ReactNode } from 'react';

export type TipsAndHelpCardStyleProps = {
  theme: ITheme;
  className?: string;
};

export interface TipsAndHelpCardProps
  extends IPropsWithStyles<TipsAndHelpCardStyleProps, TipsAndHelpCardStyles> {
  card: TipsAndHelpCardDetails;
  className?: string;
  theme?: ITheme;
}

export interface TipsAndHelpCardDetails {
  titleIcon?: TipsAndHelpIcon;
  title: string;
  content: ReactNode;
  action?: {
    name: string;
    iconName: string;
    isDisabled?: boolean;
    onClick(): void;
  };
  actionLink?: {
    name: string;
    iconName?: string;
    href: string;
    /** Default is _blank */
    target?: string;
    onClick?: () => void;
  };
}

export enum TipsAndHelpIcon {
  Idea = 'SunQuestionMark',
  ThumbsUp = 'Like'
}

export interface TipsAndHelpCardStyles {
  root: IStyle;
  content: IStyle;
  title: IStyle;
  text: IStyle;
  action: IStyle;
  button: IStyle;
}
