// @business-app/fabric/lib-cjs/components/Accordion/Accordion.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultHeaderStyles = exports.styles = void 0;
var styles = function (_a) {
    var containerClassName = _a.containerClassName, isContentVisible = _a.isContentVisible;
    return ({
        container: [
            'ba-Accordion',
            isContentVisible && 'ba-Accordion--contentVisible',
            containerClassName,
            {
                width: '100%'
            }
        ],
        header: [
            'ba-Accordion-header',
            {
                width: '100%',
                minWidth: 0,
                textAlign: 'left'
            }
        ],
        content: [
            'ba-Accordion-content',
            {
                selectors: {
                    '> *': {
                        width: '100%',
                        display: 'block'
                    }
                }
            }
        ]
    });
};
exports.styles = styles;
function getDefaultHeaderStyles(theme) {
    return {
        root: {
            borderColor: 'transparent',
            backgroundColor: theme.palette.neutralLighter
        },
        rootHovered: {
            backgroundColor: theme.palette.neutralQuaternaryAlt
        }
    };
}
exports.getDefaultHeaderStyles = getDefaultHeaderStyles;
