// @business-app/fabric/src/components/FilterSearchBox/examples/FilterSearchBox.SingleFilters.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import * as React from 'react';

import { Label } from '@fluentui/react/lib/Label';

import { FilterSearchBox, IFilterItem } from '../';

export interface IFilterSearchBoxSingleFiltersExampleState {
  searchText: string;
  selectedFilterKey: string;
}

const filters: IFilterItem[] = [{ key: 'only', text: 'Only Filter' }];

export class FilterSearchBoxSingleFiltersExample extends React.Component<
  {},
  IFilterSearchBoxSingleFiltersExampleState
> {
  public readonly state = {
    searchText: '',
    selectedFilterKey: filters[0].key
  };

  public render() {
    const { searchText, selectedFilterKey } = this.state;
    return (
      <>
        <FilterSearchBox
          strings={{
            searchBoxPlaceholder: 'Search',
            filterButtonTooltipText: 'Tooltip text'
          }}
          searchText={searchText}
          onSearchTextChange={this._onSearchTextChange}
          filters={filters}
          selectedFilterKey={selectedFilterKey}
          onFilterChange={this._onFilterChange}
          styles={{
            root: {
              marginBottom: 16
            }
          }}
        />
        <Label>{`Current search text: ${searchText}`}</Label>
        <Label>{`Current filter key: ${selectedFilterKey}`}</Label>
      </>
    );
  }

  private _onSearchTextChange = (event?: React.ChangeEvent<HTMLInputElement>, text?: string) => {
    this.setState({ searchText: text });
  };

  private _onFilterChange = (key: string) => {
    this.setState({ selectedFilterKey: key });
  };
}
