// @business-app/fabric/src/components/Form/Form.types.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { IRefObject } from '@fluentui/react/lib/Utilities';
import * as React from 'react';

/* eslint-disable @typescript-eslint/no-empty-interface */
export interface IForm {}

/* eslint-disable @typescript-eslint/no-explicit-any */
export interface IFormResult {
  [key: string]: any;
}

/**
 * The props for Form
 */
export interface IFormProps extends React.AllHTMLAttributes<HTMLFormElement> {
  componentRef?: IRefObject<IForm>;

  /** The submit handler. Passes back results keyed by the names of the inputs */
  onSubmit?: (results: IFormResult) => void;

  /** The invalid submit handler. Passes back results when the form is submitted but fails validation */
  onInvalidSubmit?: (results: { [key: string]: any }) => void;

  /** The update handler. Passes back the valid, updated value of form inputs */
  onUpdated?: (key: string, value: any) => void;

  /** Should the form show errors when it is pristine? */
  showErrorsWhenPristine?: boolean;
}
