// @business-app/fabric/lib/components/Error/examples/Error.Basic.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { Error } from '..';
var primaryMessage = 'This is primary messaging';
var secondaryMessage = 'This is secondary messaging where a description with more information is placed. This section can also include a link.';
var callToAction = 'Call to action';
var errorDetails = {
    messageLabel: 'Error/message',
    message: 'a message',
    timeStampLabel: 'Timestamp',
    timeStamp: 'Mon Nov 09 2020 14:41:25 GMT-0800 (Pacific Standard Time)',
    sessionIdLabel: 'Session ID',
    sessionId: '1181a7fc-3df9-4c2e-8eab-89c2e429cf51',
    objectIdLabel: 'Object ID/PUID',
    objectId: '297d311d-d5de-4c5a-a4c2-2c58ae24e97a',
    tenantIdLabel: 'Tenant ID',
    tenantId: 'bd6edef8-1f81-4411-8890-5f9a0152f775',
    customerSupportLink: {
        label: 'Contact customer support',
        url: '/',
        target: '_blank'
    }
};
var ErrorBasicExample = /** @class */ (function (_super) {
    __extends(ErrorBasicExample, _super);
    function ErrorBasicExample(props) {
        return _super.call(this, props) || this;
    }
    ErrorBasicExample.prototype.render = function () {
        return (React.createElement("div", { style: { display: 'flex', flexDirection: 'column', width: '100%', justifyContent: 'space-around' } },
            React.createElement(Error, { title: primaryMessage, subtitle: secondaryMessage, seeDetailsLinkShowLabel: 'See details', seeDetailsLinkHideLabel: 'Hide details', learnMoreLink: {
                    label: 'Learn more',
                    url: '/',
                    target: '_blank'
                }, callToActionLink: {
                    label: callToAction,
                    url: '/',
                    target: '_blank'
                }, errorDetails: errorDetails })));
    };
    return ErrorBasicExample;
}(React.Component));
export { ErrorBasicExample };
