// @business-app/fabric/src/components/Tree/examples/Tree.BasicUncontrolled.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { mergeStyleSets } from '@fluentui/react/lib/Styling';
import * as React from 'react';
import { Tree } from '../Tree';
import { ITreeItem } from '../Tree.types';
import { getFolderKeys, produceItems } from './Tree.Example.Data';

export class TreeBasicUncontrolledExample extends React.Component {
  public render(): JSX.Element {
    const items: ITreeItem[] = produceItems;
    const folderKeysSet = getFolderKeys(items);

    return (
      <div className={exampleStyles.exampleContainer}>
        <Tree
          items={items}
          initialExpandedItemKeys={folderKeysSet}
          strings={{
            listAriaLabel: 'list',
            moreCommandsAriaLabel: 'more commands'
          }}
        />
      </div>
    );
  }
}

const exampleStyles = mergeStyleSets({
  exampleContainer: {
    // 'display: flex' is important for the Tree to layout correctly.
    display: 'flex',
    // Just an example to prove that row direction also works for the Tree.
    flexDirection: 'row',
    width: 320,
    height: 200,
    userSelect: 'none'
  }
});
