// @business-app/fabric/lib-cjs/components/ColorInput/examples/ColorInput.Basic.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorInputBasicExample = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var __1 = require("..");
var ColorInput_base_1 = require("../ColorInput.base");
var ColorInputBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(ColorInputBasicExample, _super);
    function ColorInputBasicExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.getErrorMessage = function (value) {
            return !ColorInput_base_1.isValidColor(value) ? "Please enter a valid color value." : '';
        };
        return _this;
    }
    ColorInputBasicExample.prototype.render = function () {
        var _this = this;
        return (React.createElement(__1.ColorInput, { label: "Color Input", tooltipIconProps: {
                iconProps: { iconName: 'Info', styles: { root: { color: '#666666' } } },
                ariaLabel: 'New Color Input'
            }, tooltipProps: { content: 'New Color Input' }, colorBoxProps: {
                onClick: function () {
                    return;
                },
                ariaLabel: 'Show color picker',
                title: 'Show color picker',
                disabled: false
            }, colorTextFieldProps: {
                onBlur: function () {
                    return;
                },
                ariaLabel: 'Enter color code',
                disabled: false,
                onGetErrorMessage: function (value) {
                    return _this.getErrorMessage(value);
                }
            }, colorPickerProps: {
                color: 'Red'
            } }));
    };
    return ColorInputBasicExample;
}(React.Component));
exports.ColorInputBasicExample = ColorInputBasicExample;
