// @business-app/fabric/lib-cjs/components/Wizard/TipsAndHelpCard/TipsAndHelpCard.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
var getStyles = function (props) {
    var palette = props.theme.palette;
    return {
        root: [
            'ba-wizard-help-card',
            props.className,
            {
                border: '1px solid transparent',
                background: palette.white,
                margin: '0 0 16px 0',
                boxShadow: props.theme.effects.elevation4,
                borderRadius: 4
            }
        ],
        content: {
            padding: '12px',
            selectors: {
                '& > :last-child': {
                    marginBottom: 0
                }
            }
        },
        title: {
            fontSize: Styling_1.FontSizes.medium,
            fontWeight: Styling_1.FontWeights.semibold,
            margin: '0 0 4px 0'
        },
        text: {
            margin: '0 0 12px 0'
        },
        action: {
            padding: '12px',
            borderTop: "1px solid " + palette.neutralLighter,
            selectors: {
                '.ms-Button--action .ms-Button-icon': {
                    color: palette.neutralPrimary
                },
                '.ms-Button--action .ms-Button-label': {
                    textDecoration: 'underline',
                    color: palette.themePrimary
                },
                '.ms-Button--action:hover .ms-Button-label': {
                    textDecoration: 'none'
                }
            }
        },
        button: {
            height: 'auto',
            margin: '0 -4px',
            padding: 0
        }
    };
};
exports.getStyles = getStyles;
