// @business-app/fabric/lib-cjs/components/Sidebar/examples/Sidebar.Basic.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SidebarBasicExample = void 0;
var tslib_1 = require("tslib");
var Styling_1 = require("@fluentui/react/lib/Styling");
var React = require("react");
var __1 = require("..");
var SidebarBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(SidebarBasicExample, _super);
    function SidebarBasicExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SidebarBasicExample.prototype.render = function () {
        this.state = {
            active: true
        };
        /*
         * Basic sidebar example, with position override so that the example fits into the example page.
         * Here is the wrapper class. By default the sidebar takes all of the left height and is positioned on the left.
         *
         *   .sidebar-position-override {
         *       top: auto;
         *       bottom: auto;
         *       left: auto;
         *   }
         */
        return (React.createElement("div", { style: { height: 400 } },
            React.createElement(__1.Sidebar, { id: 'sidebar-basic', footerItems: [
                    {
                        key: 'basic-example-footer-item',
                        name: 'Sidebar Footer',
                        iconProps: { iconName: 'ProFootball' },
                        items: [
                            {
                                key: 'basic-example-footer-item-link1',
                                name: 'Footer Item 1',
                                iconProps: { iconName: 'CollegeFootball' },
                                active: false,
                                telemetryId: 'basic-example-footer-item-link1'
                            },
                            {
                                key: 'basic-example-footer-item-link2',
                                name: 'Footer Item 2',
                                iconProps: { iconName: 'Soccer' },
                                active: false,
                                telemetryId: 'basic-example-footer-item-link1'
                            }
                        ]
                    }
                ], theme: Styling_1.getTheme(), items: [
                    {
                        key: 'basic-example-item1',
                        name: 'Item 1',
                        iconProps: { iconName: 'Home' },
                        active: false,
                        href: 'https://www.bing.com',
                        target: '_blank'
                    },
                    {
                        key: 'basic-example-item2',
                        name: 'Item 2',
                        iconProps: { iconName: 'Bullseye' },
                        active: true
                    },
                    {
                        key: 'basic-example-item3',
                        name: 'Item Styled Example',
                        subMenuIconProps: { iconName: 'ArrowUpRight' },
                        active: false,
                        styles: {
                            root: {
                                backgroundColor: 'lightpink'
                            }
                        }
                    },
                    {
                        key: 'basic-example-accordion1',
                        name: 'More Stuff Here',
                        iconProps: { iconName: 'Code' },
                        items: [
                            {
                                key: 'basic-example-accordion1-link1',
                                name: 'Inner Link 1',
                                iconProps: { iconName: 'Mail' },
                                active: false
                            },
                            {
                                key: 'basic-example-accordion1-link2',
                                name: 'Inner Link 2',
                                iconProps: { iconName: 'AlarmClock' },
                                active: true
                            }
                        ]
                    },
                    {
                        key: 'basic-example-accordion2',
                        name: 'More Stuff Here Not Active',
                        iconProps: { iconName: 'Crown' },
                        items: [
                            {
                                key: 'basic-example-accordion2-link1',
                                name: 'Inner Link 1',
                                iconProps: { iconName: 'DiamondSolid' },
                                active: false
                            },
                            {
                                key: 'basic-example-accordion2-link2',
                                name: 'Inner Link 2',
                                iconProps: { iconName: 'DeliveryTruck' },
                                active: false
                            }
                        ]
                    },
                    {
                        key: 'basic-example-render1',
                        name: 'onRender Example',
                        iconProps: { iconName: 'AirTickets' },
                        onRender: function (item) {
                            return (React.createElement(__1.SidebarButton, { key: item.key, text: item.name, iconProps: item.iconProps, role: "menuitem", theme: Styling_1.getTheme() }));
                        }
                    },
                    {
                        key: 'basic-example-render2',
                        name: 'disabled Example',
                        iconProps: { iconName: 'Lock' },
                        disabled: true
                    }
                ] })));
    };
    return SidebarBasicExample;
}(React.Component));
exports.SidebarBasicExample = SidebarBasicExample;
