// @business-app/fabric/src/components/NotificationsPanel/NotificationCard/NotificationCard.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { getFocusStyle } from '@fluentui/react/lib/Styling';
import { IStyleFunction } from '@fluentui/react/lib/Utilities';
import { INotificationCardStyleProps, INotificationCardStyles } from './NotificationCard.types';

const clickActionStyle = {
  cursor: 'pointer'
};

export const getStyles: IStyleFunction<INotificationCardStyleProps, INotificationCardStyles> = ({
  hasOnClick,
  theme
}) => {
  const bodyClickActionStyle = hasOnClick
    ? ['ba-notificationCard-body--hasClickAction', clickActionStyle]
    : [];
  const textClickActionStyle = hasOnClick
    ? ['ba-notificationCard-text--hasClickAction', clickActionStyle]
    : [];
  const { palette } = theme;
  return {
    root: [
      'ba-notificationCard-container',
      {
        paddingBottom: 8
      }
    ],
    container: [
      'ba-notificationCard',
      {
        borderRadius: 3,
        boxShadow: '0px 2px 8px rgba(0, 0, 0, 0.1)',
        border: `1px solid ${palette.neutralLight}`,
        display: 'flex'
      }
    ],
    readMarker: [
      'ba-notificationCard-readMarker',
      {
        width: 4,
        alignSelf: 'stretch',
        background: palette.themePrimary,
        borderBottomLeftRadius: '3',
        borderTopLeftRadius: '3'
      }
    ],
    body: [
      'ba-notificationCard-body',
      {
        flex: 2,
        padding: 10,
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'stretch',
        wordBreak: 'break-word'
      },
      getFocusStyle(theme),
      ...bodyClickActionStyle
    ],
    header: [
      'ba-notificationCard-header',
      {
        display: 'flex'
      }
    ],
    headerTextContainer: [
      'ba-notificationCard-headerTextContainer',
      {
        flex: 2,
        fontSize: 12
      }
    ],
    headerSeparator: [
      'ba-notificationCard-separator',
      {
        marginLeft: 2,
        marginRight: 2,
        verticalAlign: 'middle'
      }
    ],
    closeIcon: [
      'ba-notificationCard-closeIcon',
      {
        alignSelf: 'center',
        fontSize: 8,
        cursor: 'pointer',
        background: 'transparent',
        border: '1px solid transparent'
      },
      getFocusStyle(theme)
    ],
    messageContainer: [
      'ba-notificationCard-messageContainer',
      {
        paddingTop: 8,
        display: 'flex'
      }
    ],
    message: ['ba-notificationCard-message'],
    messageTitle: [
      'ba-notificationCard-title',
      {
        fontWeight: 'normal',
        fontSize: 12,
        lineHeight: 14,
        padding: 0,
        color: palette.neutralPrimary,
        wordBreak: 'break-word'
      },
      ...textClickActionStyle
    ],
    messageDescription: [
      'ba-notificationCard-description',
      {
        fontWeight: 'normal',
        color: palette.neutralSecondary,
        fontSize: 12,
        lineHeight: 18,
        padding: 0,
        marginTop: 2,
        marginBottom: 6,
        wordBreak: 'break-word'
      },
      ...textClickActionStyle
    ],
    messageActionText: [
      'ba-notificationCard-actionText',
      {
        fontSize: 10
      }
    ],
    iconContainer: [
      'ba-notificationCard-iconContainer',
      {
        width: 30,
        height: 30,
        flexShrink: 0,
        overflow: 'hidden',
        fontSize: 24,
        lineHeight: 24,
        marginRight: 8,
        marginTop: 2,
        cursor: 'default'
      }
    ]
  };
};
