// @business-app/fabric/lib/components/Card/examples/RichCard.With.Custom.Title.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { RichCard } from '@business-app/fabric/lib/components/Card/RichCard';
import { FontIcon } from '@fluentui/react/lib/Icon';
import { mergeStyles } from '@fluentui/react/lib/Styling';
var iconClass = mergeStyles({
    fontSize: 50,
    height: 50,
    width: 50
});
var RichCardWithCustomTitleExample = /** @class */ (function (_super) {
    __extends(RichCardWithCustomTitleExample, _super);
    function RichCardWithCustomTitleExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RichCardWithCustomTitleExample.prototype.render = function () {
        return (React.createElement(RichCard, { height: 150, width: 300, headerProps: {
                onRenderTitle: function () {
                    return (React.createElement(React.Fragment, null,
                        React.createElement(FontIcon, { iconName: "FavoriteStarFill" }),
                        React.createElement(FontIcon, { iconName: "FavoriteStarFill" }),
                        React.createElement(FontIcon, { iconName: "FavoriteStarFill" }),
                        React.createElement(FontIcon, { iconName: "FavoriteStar" }),
                        React.createElement(FontIcon, { iconName: "FavoriteStar" })));
                }
            } }, "Title replaced with stars"));
    };
    return RichCardWithCustomTitleExample;
}(React.Component));
export { RichCardWithCustomTitleExample };
