// @business-app/fabric/lib/components/Toolbox/examples/Toolbox.Controlled.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import { PrimaryButton } from '@fluentui/react/lib/Button';
import { ChoiceGroup } from '@fluentui/react/lib/ChoiceGroup';
import { Image } from '@fluentui/react/lib/Image';
import { mergeStyles } from '@fluentui/react/lib/Styling';
import * as React from 'react';
import { Toolbox } from '../Toolbox';
import { ToolboxItem } from '../ToolboxItem';
var toolboxItem1Key = 'tool1';
var toolboxItem2Key = 'tool2';
var toolboxItem3Key = 'tool3';
var toolboxItem4Key = 'tool4';
var toolboxItem5Key = null;
var toolboxItem1Name = 'Play tool';
var toolboxItem2Name = 'Media tool';
var toolboxItem3Name = 'Settings tool';
var toolboxItem4Name = '(Invalid Toolbox Item)';
var toolboxItem5Name = '(Unselect Toolbox Item)';
var ToolboxControlledExample = /** @class */ (function (_super) {
    __extends(ToolboxControlledExample, _super);
    function ToolboxControlledExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onSettingChangedCallback = function (event, option) {
            _this.setState({ userChosenSetting: option.key });
        };
        _this._onChoiceBoxChanged = function (event, option) {
            _this._selectedItemChanged(option.key);
        };
        _this._selectedItemChanged = function (itemKey) {
            _this.setState({ selectedKey: itemKey });
        };
        _this.state = {
            selectedKey: toolboxItem1Key,
            userChosenSetting: ''
        };
        return _this;
    }
    ToolboxControlledExample.prototype.render = function () {
        var selectedKey = this.state.selectedKey;
        var toolboxItemRootStyle = mergeStyles({
            width: '100%',
            padding: 16
        });
        var buttonStyle = mergeStyles({
            marginRight: 16
        });
        return (React.createElement(React.Fragment, null,
            React.createElement(Toolbox, { strings: {
                    expandTooltip: 'Expand',
                    collapseTooltip: 'Collapse',
                    closeTooltip: 'Close'
                }, styles: {
                    root: {
                        height: 320
                    }
                }, selectedKey: this.state.selectedKey, onSelectedItemChanged: this._selectedItemChanged },
                React.createElement(ToolboxItem, { headerText: toolboxItem1Name, itemKey: toolboxItem1Key, itemIcon: 'Play' },
                    React.createElement("div", { className: toolboxItemRootStyle },
                        React.createElement(PrimaryButton, { className: buttonStyle }, 'Play'),
                        React.createElement(PrimaryButton, { className: buttonStyle }, 'Quit'))),
                React.createElement(ToolboxItem, { headerText: toolboxItem2Name, itemKey: toolboxItem2Key, itemIcon: 'MediaAdd' },
                    React.createElement("div", { className: toolboxItemRootStyle },
                        React.createElement(Image, { height: 150, width: 150, src: "https://i.imgur.com/B83RUSb.jpg" }))),
                React.createElement(ToolboxItem, { headerText: toolboxItem3Name, itemKey: toolboxItem3Key, itemIcon: 'Settings' },
                    React.createElement("div", { className: toolboxItemRootStyle },
                        React.createElement(ChoiceGroup, { selectedKey: this.state.userChosenSetting, onChange: this._onSettingChangedCallback, label: 'Choose your setting', options: [
                                {
                                    key: 'item1',
                                    text: 'Setting 1'
                                },
                                {
                                    key: 'item2',
                                    text: 'Setting 2'
                                }
                            ] })))),
            React.createElement(ChoiceGroup, { className: mergeStyles({ marginTop: 48 }), label: 'Override Toolbox selection', selectedKey: selectedKey, options: [
                    {
                        key: toolboxItem1Key,
                        text: toolboxItem1Name
                    },
                    {
                        key: toolboxItem2Key,
                        text: toolboxItem2Name
                    },
                    {
                        key: toolboxItem3Key,
                        text: toolboxItem3Name
                    },
                    {
                        key: toolboxItem4Key,
                        text: toolboxItem4Name
                    },
                    {
                        key: toolboxItem5Key,
                        text: toolboxItem5Name
                    }
                ], onChange: this._onChoiceBoxChanged })));
    };
    return ToolboxControlledExample;
}(React.Component));
export { ToolboxControlledExample };
