// @business-app/fabric/lib-cjs/components/SettingsExplorer/SettingsPage.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = void 0;
function getStyles(props) {
    var _a;
    var theme = props.theme;
    return {
        root: {
            flexGrow: 1
        },
        body: {
            padding: 16,
            flexGrow: 1,
            display: 'flex',
            flexDirection: 'column',
            flexWrap: 'nowrap',
            overflow: 'auto'
        },
        header: {
            display: 'flex',
            flexFlow: 'row wrap',
            justifyContent: 'space-between'
        },
        headerName: {
            flexGrow: 1,
            fontSize: 18,
            fontWeight: 600,
            alignContent: 'flex-start',
            paddingLeft: 8
        },
        contentContainer: {
            flexGrow: 1,
            padding: 8,
            overflow: 'auto',
            marginBottom: 16,
            borderBottom: "1px solid " + ((_a = theme === null || theme === void 0 ? void 0 : theme.palette) === null || _a === void 0 ? void 0 : _a.neutralLight)
        },
        footerButtons: {
            alignSelf: 'flex-end',
            paddingRight: 8
        }
    };
}
exports.getStyles = getStyles;
