// @business-app/fabric/lib-cjs/components/SettingsExplorer/SettingsExplorer.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.itemClassName = exports.twoTierStyles = exports.settingsStyles = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
var rootStyles = {
    display: 'flex',
    flexFlow: 'row nowrap',
    justifyContent: 'flex-start',
    height: 632,
    maxHeight: '80vh',
    borderSpacing: '0',
    borderCollapse: 'collapse',
    padding: 0,
    overflow: 'hidden'
};
var pageContainer = {
    width: 776,
    display: 'flex',
    flexGrow: 1
};
var settingsStyles = function (_a) {
    var palette = _a.theme.palette;
    return ({
        root: [
            rootStyles,
            {
                backgroundColor: palette.white
            }
        ],
        pageContainer: pageContainer
    });
};
exports.settingsStyles = settingsStyles;
var twoTierStyles = function (_a) {
    var palette = _a.theme.palette;
    return ({
        root: [rootStyles],
        pageContainer: pageContainer
    });
};
exports.twoTierStyles = twoTierStyles;
var itemClassName = function (propStyles) { return Styling_1.mergeStyles({ paddingBottom: 8 }, propStyles); };
exports.itemClassName = itemClassName;
