// @business-app/fabric/lib-cjs/components/SettingsExplorer/SettingsExplorer.test.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Enzyme = require("enzyme");
require("jest");
var Button_1 = require("@fluentui/react/lib/Button");
var Label_1 = require("@fluentui/react/lib/Label");
var React = require("react");
var ExplorerPanel_1 = require("./ExplorerPanel");
var SettingsExplorer_base_1 = require("./SettingsExplorer.base");
var SettingsItem_1 = require("./SettingsItem");
var SettingsPage_1 = require("./SettingsPage");
describe('Settings component with children', function () {
    var SettingsExplorerProps;
    var wrapper;
    var mountedWrapper;
    var onSaveButtonClick = jest.fn();
    var onCancelButtonClick = jest.fn();
    var onExitButtonClick = jest.fn();
    var onPageSelected = jest.fn();
    beforeEach(function () {
        SettingsExplorerProps = {
            name: 'General Settings',
            saveButtonProps: {
                onClick: onSaveButtonClick,
                text: 'Save'
            },
            cancelButtonProps: {
                onClick: onCancelButtonClick,
                text: 'Cancel'
            },
            exitButtonProps: {
                onClick: onExitButtonClick,
                ariaLabel: 'Exit'
            }
        };
        wrapper = Enzyme.shallow(React.createElement(SettingsExplorer_base_1.SettingsExplorerBase, tslib_1.__assign({}, SettingsExplorerProps),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'First' }),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'Second' })));
    });
    it('should initialize the state to default select the first element in the list if no initial id is provided', function () {
        expect(wrapper.state()).toHaveProperty('selectedKey');
        expect(wrapper.state().selectedKey).toBe('0');
    });
    it('should initialize the state to default select the first non-disabled element in the list if no id is provided', function () {
        wrapper = Enzyme.shallow(React.createElement(SettingsExplorer_base_1.SettingsExplorerBase, tslib_1.__assign({}, SettingsExplorerProps),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'First', id: "id1", disabled: true }),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'Second', id: "id2" })));
        expect(wrapper.state()).toHaveProperty('selectedKey');
        expect(wrapper.state().selectedKey).toBe('id2');
    });
    it('should initialize the state to default select the initialSelected page if provided', function () {
        SettingsExplorerProps.initialSelectedPageId = 'id2';
        mountedWrapper = Enzyme.mount(React.createElement(SettingsExplorer_base_1.SettingsExplorerBase, tslib_1.__assign({}, SettingsExplorerProps),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'First', id: "id1" }),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'Second', id: "id2" }),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'Third', id: "id3" })));
        expect(mountedWrapper.state()).toHaveProperty('selectedKey');
        expect(mountedWrapper.state().selectedKey).toBe('id2');
    });
    it('should update the current selected key on button click of different key', function () {
        var lastButton = mountedWrapper.find(ExplorerPanel_1.ExplorerPanel).find({ text: 'Second' }).first();
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        lastButton.props().onClick({});
        expect(mountedWrapper.state().selectedKey).toBe('id2');
    });
    it('should allow a custom key prop to override the index-based key', function () {
        var firstKey = 'firstKey';
        mountedWrapper = Enzyme.mount(React.createElement(SettingsExplorer_base_1.SettingsExplorerBase, tslib_1.__assign({}, SettingsExplorerProps),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'First', id: firstKey }),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'Second' })));
        expect(mountedWrapper.state().selectedKey).toBe(firstKey);
    });
    it('should render a single side panel', function () {
        expect(mountedWrapper.find(ExplorerPanel_1.ExplorerPanel)).toHaveLength(1);
    });
    it('should render the same number of side panel buttons as there are child props', function () {
        var panelButtons = mountedWrapper.find(ExplorerPanel_1.ExplorerPanel).find(Button_1.DefaultButton);
        expect(panelButtons.exists()).toBeTruthy();
        expect(panelButtons).toHaveLength(React.Children.count(mountedWrapper.props().children));
    });
    it('should render panel buttons with the same names as the child props', function () {
        var sidePanel = mountedWrapper.find(ExplorerPanel_1.ExplorerPanel);
        expect(sidePanel.find({ text: 'First' }).exists()).toBeTruthy();
        expect(sidePanel.find({ text: 'Second' }).exists()).toBeTruthy();
    });
    it('should render a footer with a save and cancel button', function () {
        mountedWrapper = Enzyme.mount(React.createElement(SettingsExplorer_base_1.SettingsExplorerBase, tslib_1.__assign({}, SettingsExplorerProps),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'First', id: "id1", disabled: true }),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'Second', id: "id2" })));
        var pageContents = mountedWrapper.find(SettingsPage_1.SettingsPage);
        expect(pageContents.exists()).toBeTruthy();
        expect(pageContents.find(Button_1.PrimaryButton).exists()).toBeTruthy();
        expect(pageContents.find(Button_1.DefaultButton).exists()).toBeTruthy();
        expect(pageContents.find({ text: 'Save' }).exists()).toBeTruthy();
        expect(pageContents.find({ text: 'Cancel' }).exists()).toBeTruthy();
    });
    it('should call onPageSelected for when a page is selected', function () {
        var firstKey = 'id1';
        var secondKey = 'id2';
        mountedWrapper = Enzyme.mount(React.createElement(SettingsExplorer_base_1.SettingsExplorerBase, tslib_1.__assign({}, SettingsExplorerProps, { onPageSelected: onPageSelected }),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'First', id: firstKey }),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'Second', id: secondKey })));
        // initial page selection callback
        expect(onPageSelected).toHaveBeenCalledWith(firstKey);
        var sidePanel = mountedWrapper.find(ExplorerPanel_1.ExplorerPanel);
        sidePanel
            .find(Button_1.DefaultButton)
            .last()
            .props()
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            .onClick({});
        // page selection callback
        expect(onPageSelected).toHaveBeenCalledWith(secondKey);
    });
    it('should call save and cancel when corresponding buttons are clicked', function () {
        var buttons = mountedWrapper.find(SettingsPage_1.SettingsPage).find(Button_1.DefaultButton);
        expect(buttons).toHaveLength(2);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        buttons.forEach(function (button) { return button.props().onClick({}); });
        expect(onSaveButtonClick).toHaveBeenCalled();
        expect(onCancelButtonClick).toHaveBeenCalled();
    });
    it('should render a close button to exit the explorer', function () {
        var pageHeader = mountedWrapper.find(SettingsPage_1.SettingsPage);
        expect(pageHeader.exists()).toBeTruthy();
        var exitButton = pageHeader.find(Button_1.IconButton);
        expect(exitButton.exists()).toBeTruthy();
        expect(exitButton.find({ ariaLabel: 'Exit' }).exists()).toBeTruthy();
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        exitButton.props().onClick({});
        expect(onExitButtonClick).toHaveBeenCalled();
    });
    it('should only have a single settings page being rendered at a time', function () {
        var pageContents = mountedWrapper.find(SettingsPage_1.SettingsPage);
        expect(pageContents.exists()).toBeTruthy();
        var sidePanel = mountedWrapper.find(ExplorerPanel_1.ExplorerPanel);
        sidePanel
            .find(Button_1.DefaultButton)
            .last()
            .props()
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            .onClick({});
        expect(pageContents.find(SettingsPage_1.SettingsPage)).toHaveLength(1);
    });
    it('should render the children of each settings page', function () {
        mountedWrapper = Enzyme.mount(React.createElement(SettingsExplorer_base_1.SettingsExplorerBase, tslib_1.__assign({}, SettingsExplorerProps),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'First' },
                React.createElement(SettingsItem_1.SettingsItem, { name: 'Settings Page with label' },
                    React.createElement(Label_1.Label, null, "some text"))),
            React.createElement(SettingsPage_1.SettingsPage, { name: 'Second' })));
        var label = mountedWrapper.find(SettingsPage_1.SettingsPage).find(SettingsPage_1.SettingsPage).find(SettingsItem_1.SettingsItem).find(Label_1.Label);
        expect(label.exists()).toBeTruthy();
    });
});
