// @business-app/fabric/lib-cjs/components/SettingsExplorer/TwoTierSettingsExplorer.test.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Enzyme = require("enzyme");
require("jest");
var React = require("react");
var SettingsExplorer_1 = require("./SettingsExplorer");
var SettingsPage_1 = require("./SettingsPage");
var TwoTierSettingsExplorer_base_1 = require("./TwoTierSettingsExplorer.base");
var TwoTierSettingsExplorerProps = {
    name: 'General Settings',
    saveButtonProps: {
        onClick: function () { return null; },
        text: 'Save'
    },
    cancelButtonProps: {
        onClick: function () { return null; },
        text: 'Cancel'
    },
    exitButtonProps: {
        onClick: function () { return null; },
        ariaLabel: 'Exit'
    }
};
var wrapper;
describe('Two tiered settings component', function () {
    beforeEach(function () {
        wrapper = Enzyme.shallow(React.createElement(TwoTierSettingsExplorer_base_1.TwoTierSettingsExplorerBase, tslib_1.__assign({}, TwoTierSettingsExplorerProps),
            React.createElement(SettingsExplorer_1.SettingsExplorer, { name: 'First' },
                React.createElement(SettingsPage_1.SettingsPage, { name: 'a' }),
                React.createElement(SettingsPage_1.SettingsPage, { name: 'b' })),
            React.createElement(SettingsExplorer_1.SettingsExplorer, { name: 'Second' },
                React.createElement(SettingsPage_1.SettingsPage, { name: 'c' }),
                React.createElement(SettingsPage_1.SettingsPage, { name: 'd' }))));
    });
    it('should render as expected', function () {
        expect(wrapper).toMatchSnapshot();
    });
});
describe('Two tiered settings component with custom key props', function () {
    beforeEach(function () {
        wrapper = Enzyme.shallow(React.createElement(TwoTierSettingsExplorer_base_1.TwoTierSettingsExplorerBase, tslib_1.__assign({}, TwoTierSettingsExplorerProps),
            React.createElement(SettingsExplorer_1.SettingsExplorer, { name: 'First', id: 'First' },
                React.createElement(SettingsPage_1.SettingsPage, { name: 'a', id: 'a' }),
                React.createElement(SettingsPage_1.SettingsPage, { name: 'b' })),
            React.createElement(SettingsExplorer_1.SettingsExplorer, { name: 'Second', id: 'Second' },
                React.createElement(SettingsPage_1.SettingsPage, { name: 'c' }),
                React.createElement(SettingsPage_1.SettingsPage, { name: 'd', id: 'a' }))));
    });
    it('should render as expected', function () {
        expect(wrapper).toMatchSnapshot();
    });
});
