// @business-app/fabric/src/components/Form/inputs/tagPicker/FormTagPicker.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import * as React from 'react';

// Components
import { ITag, ITagPickerProps, TagPicker } from '@fluentui/react/lib/Pickers';
import { IFormContext } from '../../Form';
import { FormBaseInput, IFormBaseInputState } from '../../FormBaseInput';
import { IFormTagPickerProps } from './FormTagPicker.types';

/**
 * TagPicker input for Form
 */
export class FormTagPicker extends FormBaseInput<ITag[], IFormTagPickerProps, IFormBaseInputState<ITag[]>> {
  constructor(props: IFormTagPickerProps, context: IFormContext) {
    super(props, context);

    this.state = {
      isValid: true,
      currentValue: props.value,
      currentError: undefined
    };

    this._validateTagPickerProps(this.props.tagPickerProps);
  }

  /**
   * Render a Fabric TagPicker
   */
  public render(): JSX.Element {
    const { currentValue } = this.state;

    return (
      <TagPicker
        {...this.props.tagPickerProps}
        // These props cannot be overridden
        key={this.props.inputKey}
        onChange={this._onChange}
        selectedItems={currentValue}
      />
    );
  }

  private _onChange = (items: ITag[]): void => {
    const { tagPickerProps } = this.props;

    tagPickerProps?.onChange?.(items);

    this.setValue(items);
  };

  private _validateTagPickerProps(props?: ITagPickerProps): void {
    if (!props) {
      return;
    }

    if (props.selectedItems !== null && props.selectedItems !== undefined) {
      // eslint-disable-next-line no-console
      console.warn(`FormTagPicker: 'selectedItems' prop was specified and will be ignored`);
    }

    if (props.onChange) {
      // eslint-disable-next-line no-console
      console.warn(`FormTagPicker: 'onChange' prop was specified and will be ignored`);
    }
  }
}
