// @business-app/fabric/lib-cjs/components/Tree/Tree.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = void 0;
var tslib_1 = require("tslib");
var Styling_1 = require("@fluentui/react/lib/Styling");
var Tree_selectors_1 = require("./Tree.selectors");
var buttonHighContrastFocus = {
    left: -2,
    top: -2,
    bottom: -2,
    right: -2,
    border: 'none',
    outlineColor: 'ButtonText'
};
var getStyles = function (props) {
    var _a, _b, _c;
    var className = props.className, theme = props.theme, nodeHeight = props.nodeHeight, iconWidth = props.iconWidth;
    var palette = theme.palette;
    var sharedIconContainerStyle = {
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        height: '100%',
        width: iconWidth,
        flex: '0 0 auto'
    };
    var sharedIconStyle = {
        flex: '0 0 auto',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        width: '1em',
        height: '1em',
        padding: 0,
        margin: 0
    };
    var blankIconBlockStyle = {
        width: iconWidth,
        flex: '0 0 auto'
    };
    // Menu button should be a little smaller than the node to account for borders.
    var menuButtonSize = nodeHeight - 2;
    var backgroundColor = palette.neutralLighter;
    var hoveredRowColor = palette.neutralQuaternaryAlt;
    var selectedRowColor = palette.neutralQuaternary;
    return {
        root: [
            className,
            Tree_selectors_1.TreeAutomationSelectors.treeRootClassName,
            {
                display: 'flex',
                boxSizing: 'border-box',
                height: '100%',
                width: '100%',
                backgroundColor: backgroundColor
            }
        ],
        innerRoot: {
            display: 'block',
            position: 'relative',
            flex: '1 1 auto',
            boxSizing: 'border-box'
        },
        focusZone: {},
        scrollView: {
            display: 'flex',
            overflow: 'auto',
            position: 'absolute',
            top: 0,
            left: 0,
            right: 0,
            bottom: 0
        },
        list: {
            display: 'flex',
            flex: '0 0 auto',
            minWidth: '100%',
            minHeight: '100%',
            selectors: (_a = {},
                _a['& .ms-List-surface'] = {
                    width: '100%',
                    height: '100%'
                },
                _a)
        },
        item: [
            Styling_1.getFocusStyle(theme, -1, 'relative', buttonHighContrastFocus),
            Tree_selectors_1.TreeAutomationSelectors.treeNodeClassName,
            {
                display: 'flex',
                flex: '1 1 auto',
                flexDirection: 'row',
                alignItems: 'center',
                whiteSpace: 'nowrap',
                height: nodeHeight,
                position: 'relative',
                outline: 'none',
                border: '1px solid transparent',
                boxSizing: 'border-box'
            }
        ],
        itemIsSelectable: {
            cursor: 'pointer'
        },
        itemIsNotSelectable: {
            selectors: {
                ':hover': {
                    backgroundColor: 'initial'
                }
            }
        },
        itemIsSelected: {
            backgroundColor: selectedRowColor,
            selectors: (_b = {},
                _b["& ." + Tree_selectors_1.TreeAutomationSelectors.menuButtonClassName] = {
                    visibility: 'visible'
                },
                _b)
        },
        itemIsNotSelected: {
            selectors: (_c = {
                    ':hover': {
                        backgroundColor: hoveredRowColor
                    }
                },
                _c[":hover& ." + Tree_selectors_1.TreeAutomationSelectors.menuButtonClassName] = {
                    visibility: 'visible'
                },
                _c["& ." + Tree_selectors_1.TreeAutomationSelectors.openMenuClassName] = {
                    visibility: 'visible'
                },
                _c)
        },
        itemLabel: {
            flex: '1 1 auto',
            whiteSpace: 'nowrap',
            maxWidth: '100%',
            marginLeft: '5px',
            outline: 'none',
            marginRight: 12
        },
        expandIconButton: [
            Tree_selectors_1.TreeAutomationSelectors.collapseExpandClassName,
            tslib_1.__assign(tslib_1.__assign({}, sharedIconContainerStyle), { color: palette.black })
        ],
        expandIconButtonEnabled: {
            cursor: 'pointer'
        },
        expandIconButtonDisabled: {
            opacity: 0.4
        },
        expandIcon: tslib_1.__assign(tslib_1.__assign({}, sharedIconStyle), { fontSize: 12, transition: '0.2s transform', transformOrigin: 'center' }),
        expandIconExpanded: {
            transform: 'rotate(90deg)'
        },
        noExpandIcon: tslib_1.__assign({}, blankIconBlockStyle),
        levelIndentSpacer: tslib_1.__assign({}, blankIconBlockStyle),
        menuButton: [
            Tree_selectors_1.TreeAutomationSelectors.menuButtonClassName,
            {
                flex: '0 0 auto',
                // Menu button should be hidden by default.
                // We will show it in the following cases, though: a row is hovered, a row is selected, or the menu is open.
                visibility: 'hidden',
                position: 'sticky',
                right: 0,
                width: menuButtonSize,
                height: menuButtonSize,
                fontSize: 16,
                border: 'none',
                cursor: 'pointer',
                selectors: {
                    ':hover': {
                        color: palette.themePrimary
                    }
                }
            }
        ],
        menuButtonSelectedRowClosedMenu: {
            backgroundColor: selectedRowColor,
            selectors: {
                ':hover': {
                    backgroundColor: hoveredRowColor
                }
            }
        },
        menuButtonSelectedRowOpenMenu: [
            Tree_selectors_1.TreeAutomationSelectors.openMenuClassName,
            {
                color: palette.themePrimary,
                backgroundColor: hoveredRowColor
            }
        ],
        menuButtonUnselectedRowClosedMenu: {
            backgroundColor: hoveredRowColor,
            selectors: {
                ':hover': {
                    backgroundColor: selectedRowColor
                }
            }
        },
        menuButtonUnselectedRowOpenMenu: [
            Tree_selectors_1.TreeAutomationSelectors.openMenuClassName,
            {
                color: palette.themePrimary,
                backgroundColor: selectedRowColor
            }
        ],
        menuButtonDisabled: {
            color: palette.neutralSecondaryAlt,
            backgroundColor: 'transparent',
            selectors: {
                ':hover': {
                    color: palette.neutralSecondaryAlt,
                    backgroundColor: 'transparent'
                }
            }
        },
        itemIconContainer: tslib_1.__assign({}, sharedIconContainerStyle),
        itemIcon: tslib_1.__assign(tslib_1.__assign({}, sharedIconStyle), { fontSize: 16 }),
        noItemIcon: {
            flex: '0 0 auto',
            // By default, don't take up any space with the no icon element (some trees may not even have icons).
            width: 0
        }
    };
};
exports.getStyles = getStyles;
