// @business-app/fabric/lib/components/FilterSearchViews/examples/FilterSearchViews.FlatListMultipleViews.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { Toggle } from '@fluentui/react/lib/Toggle';
import { FilterSearchViews } from '../index';
import { filters, flatItems, shouldFilterItem } from './FilterSearchViewsExampleData';
import { IListRenderMethod, renderItem } from './FilterSearchViewsExampleList';
var FilterSearchViewsFlatListMultipleViewsExample = /** @class */ (function (_super) {
    __extends(FilterSearchViewsFlatListMultipleViewsExample, _super);
    function FilterSearchViewsFlatListMultipleViewsExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            includeHorizontalList: true,
            includeGrid: true,
            amountFiltered: 0
        };
        _this._onHorizontalToggleChange = function (_, checked) {
            _this.setState({ includeHorizontalList: checked });
        };
        _this._onGridToggleChange = function (_, checked) {
            _this.setState({ includeGrid: checked });
        };
        return _this;
    }
    FilterSearchViewsFlatListMultipleViewsExample.prototype.render = function () {
        var _this = this;
        var _a = this.state, includeHorizontalList = _a.includeHorizontalList, includeGrid = _a.includeGrid;
        var views = [
            {
                key: 'verticalList',
                text: 'Vertical List',
                ariaLabel: 'Vertical list view',
                iconName: 'AlignVerticalBottom',
                onRenderView: function (filteredItems) {
                    if (_this.state.amountFiltered !== filteredItems.length) {
                        _this.setState({ amountFiltered: filteredItems.length });
                    }
                    return _this._renderList(filteredItems, IListRenderMethod.VerticalList);
                }
            }
        ];
        if (includeHorizontalList) {
            views.push({
                key: 'horizontalList',
                text: 'Horizontal List',
                ariaLabel: 'Horizontal list view',
                iconName: 'AlignHorizontalLeft',
                onRenderView: function (filteredItems) {
                    if (_this.state.amountFiltered !== filteredItems.length) {
                        _this.setState({ amountFiltered: filteredItems.length });
                    }
                    return _this._renderList(filteredItems, IListRenderMethod.HorizontalList);
                }
            });
        }
        if (includeGrid) {
            views.push({
                key: 'grid',
                text: 'Grid',
                ariaLabel: 'Grid view',
                iconName: 'GridViewMedium',
                onRenderView: function (filteredItems) {
                    if (_this.state.amountFiltered !== filteredItems.length) {
                        _this.setState({ amountFiltered: filteredItems.length });
                    }
                    return _this._renderList(filteredItems, IListRenderMethod.Grid);
                }
            });
        }
        return (React.createElement(React.Fragment, null,
            React.createElement(Toggle, { checked: includeHorizontalList, onChange: this._onHorizontalToggleChange, label: 'Include Horizontal List' }),
            React.createElement(Toggle, { checked: includeGrid, onChange: this._onGridToggleChange, label: 'Include Grid', styles: {
                    root: {
                        marginBottom: 12
                    }
                } }),
            React.createElement(FilterSearchViews, { strings: {
                    searchBoxPlaceholder: 'Search',
                    allItemsFilterItemText: 'All'
                }, items: flatItems, shouldFilterItem: shouldFilterItem, filters: filters, views: views }),
            React.createElement("div", { "aria-live": "assertive", "aria-atomic": "true", style: { overflow: 'hidden', position: 'absolute', top: -1000, left: -1000, width: 0, height: 0 } }, this.state.amountFiltered + ' items shown of ' + flatItems.length)));
    };
    FilterSearchViewsFlatListMultipleViewsExample.prototype._renderList = function (filteredItems, renderMethod) {
        return (React.createElement("div", { style: {
                display: 'flex',
                flexDirection: renderMethod === IListRenderMethod.VerticalList ? 'column' : 'row',
                flexWrap: renderMethod === IListRenderMethod.Grid ? 'wrap' : 'initial',
                overflowX: renderMethod === IListRenderMethod.HorizontalList ? 'auto' : 'initial'
            } }, filteredItems.map(function (item, index) { return renderItem(item, index, renderMethod); })));
    };
    return FilterSearchViewsFlatListMultipleViewsExample;
}(React.Component));
export { FilterSearchViewsFlatListMultipleViewsExample };
