// @business-app/fabric/src/components/Carousel/examples/Carousel.ControlButtons.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import * as React from 'react';
import { Carousel, ICarouselStyles } from '..';
import { getMockContent } from './Carousel.Example.Data';

export class CarouselControlButtonsExample extends React.Component<{}, {}> {
  constructor(props: {}) {
    super(props);
  }

  public render(): JSX.Element {
    const customCarouselStyles: ICarouselStyles = {
      root: {
        flexDirection: 'column'
      },
      subComponentStyles: {
        tooltipControl: {
          root: {
            position: 'initial',
            transform: 'initial',
            marginRight: 8
          }
        },
        control: {
          root: {
            borderRadius: 'initial'
          }
        },
        indicator: {},
        toggle: {}
      },
      indicatorsWrapper: {
        position: 'initial',
        transform: 'initial'
      },
      controlsWrapper: {
        margin: 24
      }
    };

    return (
      <Carousel
        styles={customCarouselStyles}
        items={getMockContent()}
        useControlLabels={true}
        disableNavigationLoop={true}
        navPreviousButtonAriaLabel="Previous"
        navNextButtonAriaLabel="Next"
        indicatorButtonsAriaLabel={['Item 1', 'Item 2', 'Item 3', 'Item 4']}
        playButtonAriaLabel="Play"
        pauseButtonAriaLabel="Pause"
        orderIndicatorsBeforeControls={true}
      />
    );
  }
}
