// @business-app/fabric/lib-cjs/components/SettingsExplorer/examples/SettingsExplorer.Basic.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SettingsExplorerBasicExample = void 0;
var tslib_1 = require("tslib");
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
var Button_1 = require("@fluentui/react/lib/Button");
var Checkbox_1 = require("@fluentui/react/lib/Checkbox");
var Dropdown_1 = require("@fluentui/react/lib/Dropdown");
var Modal_1 = require("@fluentui/react/lib/Modal");
var Overlay_1 = require("@fluentui/react/lib/Overlay");
var Spinner_1 = require("@fluentui/react/lib/Spinner");
var Styling_1 = require("@fluentui/react/lib/Styling");
var React = require("react");
var SettingsExplorer_1 = require("../SettingsExplorer");
var SettingsItem_1 = require("../SettingsItem");
var SettingsPage_1 = require("../SettingsPage");
var ExampleInput_1 = require("./ExampleInput");
var dropdownOptions = [
    { key: 'pug', text: 'pug' },
    { key: 'poodle', text: 'poodle' },
    { key: 'cocker spaniel', text: 'cocker spaniel' },
    { key: 'samoyed', text: 'samoyed' }
];
var asyncSettingPageId = 'asyncSettingPage';
var SettingsExplorerBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(SettingsExplorerBasicExample, _super);
    function SettingsExplorerBasicExample(props) {
        var _this = _super.call(this, props) || this;
        _this.buttonRef = React.createRef();
        _this.loadAsyncSetting = function () {
            setTimeout(function () {
                _this.setState({
                    asyncSettingsData: 'The quick brown fox jumps over the lazy asynchronous page.'
                });
                _this.buttonRef.current.focus();
            }, 2000);
        };
        _this._openSettingsExplorer = function (status) {
            _this.setState({ modalOpen: status });
        };
        _this.state = {
            modalOpen: false,
            asyncSettingsData: ''
        };
        return _this;
    }
    SettingsExplorerBasicExample.prototype.render = function () {
        var _this = this;
        return (React.createElement("div", null,
            React.createElement(Button_1.DefaultButton, { onClick: function () { return _this._openSettingsExplorer(true); }, text: "One Tiered", className: Styling_1.mergeStyles({
                    width: 200
                }) }),
            React.createElement(Modal_1.Modal, { isOpen: this.state.modalOpen, onDismiss: function () { return _this._openSettingsExplorer(false); }, closeButtonAriaLabel: "Cancel" },
                React.createElement(SettingsExplorer_1.SettingsExplorer, tslib_1.__assign({ name: 'General Settings', saveButtonProps: {
                        onClick: function () {
                            alert('save clicked');
                        },
                        text: 'Save'
                    }, cancelButtonProps: {
                        onClick: function () {
                            _this._openSettingsExplorer(false);
                        },
                        text: 'Discard'
                    }, exitButtonProps: {
                        onClick: function () { return _this._openSettingsExplorer(false); },
                        ariaLabel: 'Clear'
                    }, onPageSelected: function (id) {
                        if (id === asyncSettingPageId) {
                            _this.loadAsyncSetting();
                        }
                    } }, this.props),
                    React.createElement(SettingsPage_1.SettingsPage, { name: 'Language and time' },
                        React.createElement(SettingsItem_1.SettingsItem, null,
                            React.createElement(Dropdown_1.Dropdown, { label: 'Language', options: ExampleInput_1.languageOptions, defaultSelectedKey: 'en-US' })),
                        React.createElement(SettingsItem_1.SettingsItem, null,
                            React.createElement(Dropdown_1.Dropdown, { label: 'Date format', options: ExampleInput_1.dateFormatOptions, defaultSelectedKey: 'M/D/YYYY' })),
                        React.createElement(SettingsItem_1.SettingsItem, null,
                            React.createElement(Dropdown_1.Dropdown, { label: 'Time format', options: ExampleInput_1.timeFormatOptions, defaultSelectedKey: '1:01 AM - 11:59 PM' })),
                        React.createElement(SettingsItem_1.SettingsItem, null,
                            React.createElement(Dropdown_1.Dropdown, { label: 'Current timezone', options: dropdownOptions, defaultSelectedKey: 'pug' }))),
                    React.createElement(SettingsPage_1.SettingsPage, { name: 'Disabled Page', disabled: true },
                        React.createElement(SettingsItem_1.SettingsItem, { name: 'Disabled setting' },
                            React.createElement("p", null, "You should never be able to see this."))),
                    React.createElement(SettingsPage_1.SettingsPage, { name: 'Async Setting', id: asyncSettingPageId }, this.state.asyncSettingsData ? (React.createElement(React.Fragment, null,
                        React.createElement("p", null, this.state.asyncSettingsData),
                        React.createElement(Button_1.DefaultButton, { ariaLabel: 'Refresh', componentRef: this.buttonRef, onClick: function () {
                                _this.setState({
                                    asyncSettingsData: ''
                                }, function () {
                                    _this.loadAsyncSetting();
                                });
                            } }, "Refresh"))) : (React.createElement(Overlay_1.Overlay, { styles: {
                            root: {
                                backgroundColor: '#ffffff',
                                opacity: 0.7,
                                display: 'flex',
                                alignItems: 'center',
                                justifyContent: 'center'
                            }
                        } },
                        React.createElement(Spinner_1.Spinner, { size: Spinner_1.SpinnerSize.large, label: 'Loading' })))),
                    React.createElement(SettingsPage_1.SettingsPage, { name: 'Notifications' },
                        React.createElement(SettingsItem_1.SettingsItem, { name: 'Newsletter' },
                            React.createElement(Dropdown_1.Dropdown, { label: 'Choose your country/region:', options: dropdownOptions, defaultValue: 'pug' }),
                            React.createElement("br", null),
                            React.createElement(Checkbox_1.Checkbox, { label: 'Send me the latest PowerApps updates, tips, and offers' })),
                        React.createElement(SettingsItem_1.SettingsItem, { name: 'App usage emails' },
                            React.createElement(Checkbox_1.Checkbox, { label: 'Check here to receive app usage emails' })))))));
    };
    return SettingsExplorerBasicExample;
}(React.Component));
exports.SettingsExplorerBasicExample = SettingsExplorerBasicExample;
