// @business-app/fabric/src/components/Toolbox/ToolboxItem.types.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IKeytipProps } from '@fluentui/react/lib/Keytip';
import { IRenderFunction } from '@fluentui/react/lib/Utilities';
import * as React from 'react';

export interface IToolboxItemProps extends React.HTMLAttributes<HTMLDivElement> {
  /**
   * Gets the component ref.
   */
  componentRef?: () => void;

  /**
   * The text displayed of each toolbox link.
   */
  headerText?: string;

  /**
   * Props for the header command button supporting native props - data-*
   * and aria-* - for each toolbox header/link element
   */
  headerButtonProps?: { [key: string]: string | number | boolean };

  /**
   * An required key to uniquely identify a toolbox item.
   *
   * Note: The 'key' from react props cannot be used inside component.
   */
  itemKey: string;

  /**
   * The aria label of each toolbox link which will read by screen reader instead of linkText.
   *
   * Note that unless you have compelling requirements you should not override aria-label.
   */
  ariaLabel?: string;

  /**
   * An optional icon to show next to the toolbox link.
   */
  itemIcon?: string;

  /**
   * Whether the toolbox link should be disabled.
   */
  disabled?: boolean;

  /**
   * Optional custom renderer for the toolbox item link
   */
  onRenderItemLink?: IRenderFunction<IToolboxItemProps>;

  /**
   * Optional keytip for this ToolboxItem
   */
  keytipProps?: IKeytipProps;
}
