// @business-app/fabric/src/components/Page/examples/Page.Basic.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { Breadcrumb } from '@fluentui/react/lib/Breadcrumb';
import { ChoiceGroup, IChoiceGroupOption } from '@fluentui/react/lib/ChoiceGroup';
import { Label } from '@fluentui/react/lib/Label';
import { MessageBar, MessageBarType } from '@fluentui/react/lib/MessageBar';
import { PivotItem } from '@fluentui/react/lib/Pivot';
import { Toggle } from '@fluentui/react/lib/Toggle';
import * as React from 'react';

import { DetailsList, DetailsListLayoutMode, IColumn } from '@fluentui/react/lib/DetailsList';
import { Pivot } from '@fluentui/react/lib/Pivot';
import { Page } from '../';
import { items } from '../../../utilities/data/contextualMenuItems';
import { CommandBarWithSearchBox } from '../../CommandBarWithSearchBox';
import { PaneSize } from '../../Pane';

const categories = ['Cat', 'House', 'Table'];

export interface IPageBasicExampleState {
  isLoading: boolean;
  isOpenPane: boolean;
  paneType: PaneSize;
}

export interface IDetailsListBasicExampleItem {
  key: number;
  name: string;
  value: number;
}

export class PageBasicExample extends React.Component<{}, IPageBasicExampleState> {
  private columns: IColumn[];
  private detailsListItems: IDetailsListBasicExampleItem[];
  constructor(props: {}) {
    super(props);
    this.state = {
      isLoading: false,
      isOpenPane: false,
      paneType: PaneSize.small
    };

    this.detailsListItems = [];
    for (let i = 0; i < 75; i++) {
      this.detailsListItems.push({
        key: i,
        name: 'Item ' + i,
        value: i
      });
    }

    this.columns = [
      { key: 'column1', name: 'Name', fieldName: 'name', minWidth: 100, maxWidth: 200, isResizable: true },
      { key: 'column2', name: 'Value', fieldName: 'value', minWidth: 100, maxWidth: 200, isResizable: true }
    ];

    this._onLoadingSet = this._onLoadingSet.bind(this);
  }

  public render(): JSX.Element {
    return (
      <div>
        <Toggle
          label="Show loading overlay"
          checked={this.state.isLoading}
          onChanged={this._onLoadingSet}
          onText="Loading"
          offText="Loaded"
        />
        <ChoiceGroup
          className="defaultChoiceGroup"
          defaultSelectedKey="A"
          options={[
            {
              key: 'A',
              text: 'Small pane'
            },
            {
              key: 'B',
              text: 'Medium pane'
            }
          ]}
          onChange={this.onChangePaneType}
          label="Pane type"
          required={true}
        />
        <Toggle label="Show Pane" checked={this.state.isOpenPane} onChanged={this._onTogglePane} />
        <Label>The page is surrounded by a black border so you can tell where the page ends.</Label>
        <div style={{ border: '2px solid black' }}>
          <Page
            style={{ height: '500px', overflowX: 'scroll' }}
            paneProps={{
              isOpen: this.state.isOpenPane,
              headerText: 'Pane Example',
              onDismiss: this._onDismissPane,
              hasCloseButton: true,
              onRenderBody: this._onRenderPaneBody,
              type: this.state.paneType
            }}
            spinnerProps={{
              ariaLabel: 'Loading'
            }}
            showOverlay={this.state.isLoading}
            renderHeader={() => <CommandBarWithSearchBox items={items} />}
            renderMessageBar={() => (
              <MessageBar messageBarType={MessageBarType.error}>Lets display an error</MessageBar>
            )}
            renderFooter={() => <MessageBar>This is a page footer</MessageBar>}
            renderTitle={() => (
              <Breadcrumb
                items={[
                  { text: 'Test', key: 'Test' },
                  { text: 'Header', key: 'Header' }
                ]}
              />
            )}
            renderPivot={() => (
              <Pivot>
                <PivotItem linkText={'Test'} />
                <PivotItem linkText={'Another Test'} />
                <PivotItem linkText={'One More'} />
                <PivotItem linkText={'Hi'} />
              </Pivot>
            )}
            renderBody={this._renderBody}
          />
        </div>
      </div>
    );
  }

  private onChangePaneType = (ev: React.FormEvent<HTMLInputElement>, option: IChoiceGroupOption): void => {
    let paneType: PaneSize;
    switch (option.key) {
      case 'A':
        paneType = PaneSize.small;
        break;
      case 'B':
        paneType = PaneSize.medium;
        break;
      default:
        paneType = PaneSize.small;
        break;
    }
    this.setState({ paneType, isOpenPane: false });
  };

  private _onRenderPaneBody = (): JSX.Element => {
    return <div>Content goes here...</div>;
  };

  private _onDismissPane = (): void => {
    this.setState({ isOpenPane: false });
  };

  private _onTogglePane = (isOpenPane: boolean): void => {
    this.setState({ isOpenPane });
  };

  private _onLoadingSet(isLoading: boolean): void {
    this.setState({ isLoading });
  }
  private _renderBody = (): JSX.Element => {
    return (
      <div>
        <DetailsList
          items={this.detailsListItems}
          columns={this.columns}
          setKey="set"
          layoutMode={DetailsListLayoutMode.justified}
          selectionPreservedOnEmptyClick={true}
          ariaLabelForSelectionColumn="Toggle selection"
          ariaLabelForSelectAllCheckbox="Toggle selection for all items"
          checkButtonAriaLabel="Row checkbox"
        />
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore
          et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut
          aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
          cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
          culpa qui officia deserunt mollit anim id est laborum.
        </p>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore
          et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut
          aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
          cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
          culpa qui officia deserunt mollit anim id est laborum.
        </p>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore
          et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut
          aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
          cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
          culpa qui officia deserunt mollit anim id est laborum.
        </p>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore
          et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut
          aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
          cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
          culpa qui officia deserunt mollit anim id est laborum.
        </p>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore
          et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut
          aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
          cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
          culpa qui officia deserunt mollit anim id est laborum.
        </p>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore
          et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut
          aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
          cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
          culpa qui officia deserunt mollit anim id est laborum.
        </p>
      </div>
    );
  };
}
