// @business-app/fabric/lib-cjs/components/HighlightedText/examples/HighlightedText.Basic.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HighlightedTextBasicExample = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var HighlightedText_1 = require("../HighlightedText");
var HighlightedTextBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(HighlightedTextBasicExample, _super);
    function HighlightedTextBasicExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HighlightedTextBasicExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(HighlightedText_1.HighlightedText, { text: 'Hello world. It is a marvelous time to be a highlight.', highlights: [
                    {
                        startIndex: 0,
                        length: 11
                    },
                    {
                        startIndex: 21,
                        length: 9
                    }
                ], labels: {
                    beginHighlightedText: 'Begin highlighted text',
                    endHighlightedText: 'End highlighted text'
                } })));
    };
    return HighlightedTextBasicExample;
}(React.Component));
exports.HighlightedTextBasicExample = HighlightedTextBasicExample;
