// @business-app/fabric/lib-cjs/components/HighlightedText/examples/HighlightedText.Styled.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HighlightedTextStyledExample = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var HighlightedText_1 = require("../HighlightedText");
var HighlightedTextStyledExample = /** @class */ (function (_super) {
    tslib_1.__extends(HighlightedTextStyledExample, _super);
    function HighlightedTextStyledExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HighlightedTextStyledExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(HighlightedText_1.HighlightedText, { text: 'Highlights can be styled, too!', highlights: [
                    {
                        startIndex: 18,
                        length: 6
                    }
                ], labels: {
                    beginHighlightedText: 'Begin highlighted text',
                    endHighlightedText: 'End highlighted text'
                }, styles: {
                    highlights: {
                        backgroundColor: 'yellow'
                    }
                } })));
    };
    return HighlightedTextStyledExample;
}(React.Component));
exports.HighlightedTextStyledExample = HighlightedTextStyledExample;
