// @business-app/fabric/lib-cjs/components/Toolbox/examples/Toolbox.ConditionallyShowTools.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolboxConditionallyShowToolsExample = void 0;
var tslib_1 = require("tslib");
var Styling_1 = require("@fluentui/react/lib/Styling");
var Toggle_1 = require("@fluentui/react/lib/Toggle");
var React = require("react");
var Toolbox_1 = require("../Toolbox");
var ToolboxItem_1 = require("../ToolboxItem");
var ToolboxConditionallyShowToolsExample = /** @class */ (function (_super) {
    tslib_1.__extends(ToolboxConditionallyShowToolsExample, _super);
    function ToolboxConditionallyShowToolsExample(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            showSecondItem: false,
            showThirdItem: false,
            disableSecondItem: false
        };
        return _this;
    }
    ToolboxConditionallyShowToolsExample.prototype.render = function () {
        var _this = this;
        var _a = this.state, showSecondItem = _a.showSecondItem, showThirdItem = _a.showThirdItem, disableSecondItem = _a.disableSecondItem;
        var toolboxItemRootStyle = Styling_1.mergeStyles({
            width: '100%',
            padding: 16
        });
        return (React.createElement(React.Fragment, null,
            React.createElement(Toggle_1.Toggle, { onChange: function (ev, checked) { return _this.setState({ showSecondItem: checked }); }, defaultChecked: showSecondItem, onText: 'true', offText: 'false', label: 'Show second tool' }),
            React.createElement(Toggle_1.Toggle, { onChange: function (ev, checked) { return _this.setState({ showThirdItem: checked }); }, defaultChecked: showThirdItem, onText: 'true', offText: 'false', label: 'Show third tool' }),
            React.createElement(Toggle_1.Toggle, { onChange: function (ev, checked) { return _this.setState({ disableSecondItem: checked }); }, defaultChecked: showThirdItem, onText: 'true', offText: 'false', label: 'Disable second tool' }),
            React.createElement(Toolbox_1.Toolbox, { strings: {
                    expandTooltip: 'Expand',
                    collapseTooltip: 'Collapse',
                    closeTooltip: 'Close'
                }, styles: {
                    root: {
                        height: 320
                    }
                } },
                React.createElement(ToolboxItem_1.ToolboxItem, { headerText: 'header1', itemIcon: 'Play', itemKey: "key1" },
                    React.createElement("div", { className: toolboxItemRootStyle }, 'Item 1')),
                showSecondItem && (React.createElement(ToolboxItem_1.ToolboxItem, { headerText: 'header2', itemIcon: 'MediaAdd', itemKey: "key2", disabled: disableSecondItem },
                    React.createElement("div", { className: toolboxItemRootStyle }, 'Item 2'))),
                showThirdItem && (React.createElement(ToolboxItem_1.ToolboxItem, { headerText: 'header3', itemIcon: 'Settings', itemKey: "key3" },
                    React.createElement("div", { className: toolboxItemRootStyle }, 'Item 3'))))));
    };
    return ToolboxConditionallyShowToolsExample;
}(React.Component));
exports.ToolboxConditionallyShowToolsExample = ToolboxConditionallyShowToolsExample;
