// @business-app/fabric/lib-cjs/components/LoadingButton/examples/LoadingButton.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadingButtonExample = void 0;
var tslib_1 = require("tslib");
var CommandBar_1 = require("@fluentui/react/lib/CommandBar");
var Stack_1 = require("@fluentui/react/lib/Stack");
var Toggle_1 = require("@fluentui/react/lib/Toggle");
var React = require("react");
var LoadingButton_1 = require("../LoadingButton");
var LoadingButtonExample = /** @class */ (function (_super) {
    tslib_1.__extends(LoadingButtonExample, _super);
    function LoadingButtonExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            isLoading: false
        };
        return _this;
    }
    LoadingButtonExample.prototype.render = function () {
        var _this = this;
        return (React.createElement(Stack_1.Stack, { maxWidth: 400, gap: 16 },
            React.createElement(LoadingButton_1.LoadingPrimaryButton, { iconProps: { iconName: 'Save' }, text: "Save data", loading: this.state.isLoading, styles: { root: { width: 130 } } }),
            React.createElement(CommandBar_1.CommandBar, { items: [
                    {
                        key: 'add',
                        iconProps: { iconName: 'Add' },
                        text: 'Add item',
                        loading: this.state.isLoading
                    },
                    {
                        key: 'edit',
                        iconProps: { iconName: 'Edit' },
                        text: 'Edit item',
                        loading: this.state.isLoading
                    },
                    {
                        key: 'remove',
                        iconProps: { iconName: 'Delete' },
                        text: 'Remove item',
                        loading: this.state.isLoading
                    }
                ], buttonAs: LoadingButton_1.LoadingCommandButton }),
            React.createElement(Toggle_1.Toggle, { label: "Toggle loading", checked: !!this.state.isLoading, onClick: function () { return _this.setState({ isLoading: !_this.state.isLoading }); } })));
    };
    return LoadingButtonExample;
}(React.Component));
exports.LoadingButtonExample = LoadingButtonExample;
