// @business-app/fabric/src/components/Wizard/Progress/WizardProgress.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import {
  FontSizes,
  FontWeights,
  hiddenContentStyle,
  HighContrastSelector,
  IStyle
} from '@fluentui/react/lib/Styling';
import { WizardProgressStyleProps, WizardProgressStyles } from './WizardProgress.types';

export const getStyles = (props: WizardProgressStyleProps): WizardProgressStyles => {
  const { palette } = props.theme;
  const smallScreenSelector = '@media (max-width: 767px)';

  const step: IStyle = {
    position: 'relative',
    display: 'flex',
    flexDirection: 'column',
    boxSizing: 'border-box',
    width: '100%',
    minHeight: 60,
    margin: 0,
    padding: '0 24px 8px 52px',
    border: 'none',
    background: 'none',
    fontFamily: 'inherit',
    fontSize: FontSizes.medium,
    textAlign: 'left',
    selectors: {
      '::after': {
        content: '""',
        position: 'absolute',
        top: 22,
        bottom: 2,
        left: 31,
        display: 'block',
        borderRight: `2px solid ${palette.neutralLight}`,
        selectors: {
          [HighContrastSelector]: {
            borderRightWidth: 1
          }
        }
      },
      '.is-collapsed &': {
        selectors: {
          [smallScreenSelector]: {
            flexDirection: 'row',
            alignItems: 'flex-start',
            padding: '0 0 12px 0',
            minHeight: 'unset',
            selectors: {
              '::after': {
                position: 'static',
                flexGrow: 1,
                minWidth: 24,
                margin: '0.7em 0 0 0',
                borderBottom: `2px solid ${palette.neutralLight}`
              }
            }
          }
        }
      }
    }
  };

  const icon: IStyle = {
    position: 'absolute',
    left: 24,
    top: 2,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    selectors: {
      '.is-collapsed &': {
        selectors: {
          [smallScreenSelector]: {
            position: 'static',
            flex: '0 0 16px',
            margin: '2px 8px 0 8px'
          }
        }
      }
    }
  };

  const validOrEmptyIcon: IStyle = {
    border: `2px solid ${palette.neutralLight}`,
    boxSizing: 'border-box',
    borderRadius: '50%',
    width: 16,
    height: 16,
    overflow: 'hidden',
    background: 'transparent',
    fontSize: FontSizes.mini,
    textIndent: '20px' /* force hiding the icon in High Contrast (a transparent color is not sufficient) */
  };

  return {
    root: [
      'ba-wizard-progress',
      props.className,
      {
        position: 'relative',
        boxSizing: 'border-box',
        flex: '0 0 auto',
        width: '19vw',
        minWidth: 183,
        maxWidth: 280,
        padding: '32px 0 0 0',
        background: palette.white,
        borderRight: `1px solid ${palette.neutralLight}`,
        overflowY: 'auto',
        selectors: {
          [smallScreenSelector]: {
            width: '100%',
            maxWidth: 'none',
            borderRight: 'none',
            borderBottom: `1px solid ${palette.neutralLight}`,
            padding: '16px 0 0 0'
          },
          '&.is-collapsed': {
            selectors: {
              [smallScreenSelector]: {
                padding: '12px 0 4px 0'
              }
            }
          }
        }
      }
    ],
    list: {
      listStyle: 'none',
      margin: 0,
      padding: 0,
      selectors: {
        [smallScreenSelector]: {
          marginRight: 50 /* margin for the collapse button */
        },
        '.is-collapsed &': {
          selectors: {
            [smallScreenSelector]: {
              display: 'flex',
              justifyContent: 'center',
              marginRight: 0
            }
          }
        }
      }
    },
    itemWrapper: {
      margin: 0,
      padding: 0,
      selectors: {
        '.is-collapsed &:not(:last-child)': {
          selectors: {
            [smallScreenSelector]: {
              width: 64
            }
          }
        }
      }
    },
    passedStep: {
      ...step
    },
    currentStep: [
      'ba-wizard-progress-current',
      {
        ...step,
        fontWeight: FontWeights.semibold
      }
    ],
    futureStep: {
      ...step,
      color: palette.neutralSecondary
    },
    lastStep: {
      selectors: {
        '::after': {
          display: 'none'
        }
      }
    },
    label: {
      selectors: {
        '.is-collapsed &': {
          selectors: {
            [smallScreenSelector]: {
              display: 'none'
            }
          }
        }
      }
    },
    summary: {
      display: 'flex',
      fontSize: FontSizes.small,
      fontWeight: FontWeights.regular,
      color: palette.themePrimary,
      selectors: {
        ':first-of-type': {
          marginTop: 2
        },
        '.is-collapsed &': {
          selectors: {
            [smallScreenSelector]: {
              display: 'none'
            }
          }
        }
      }
    },
    validIcon: {
      ...icon,
      ...validOrEmptyIcon,
      backgroundColor: palette.themePrimary,
      borderWidth: 0,
      color: palette.white,
      textIndent: '0',
      selectors: {
        ...icon.selectors,
        [HighContrastSelector]: {
          borderColor: 'WindowText',
          backgroundColor: 'Window',
          borderWidth: 1
        }
      }
    },
    warningIcon: {
      ...icon,
      top: 0,
      fontSize: FontSizes.mediumPlus,
      color: palette.yellow
    },
    errorIcon: {
      ...icon,
      top: 0,
      fontSize: FontSizes.mediumPlus,
      color: palette.redDark
    },
    emptyIcon: {
      ...icon,
      ...validOrEmptyIcon,
      selectors: {
        ...icon.selectors,
        [HighContrastSelector]: {
          borderColor: 'GrayText',
          borderWidth: 1
        }
      }
    },
    currentEmptyIcon: {
      ...icon,
      ...validOrEmptyIcon,
      backgroundColor: palette.themeTertiary,
      borderWidth: 0,
      selectors: {
        ...icon.selectors,
        [HighContrastSelector]: {
          borderColor: 'GrayText',
          borderWidth: 1,
          backgroundColor: 'Highlight'
        }
      }
    },
    toggle: {
      display: 'none',
      boxSizing: 'border-box',
      position: 'absolute',
      right: 12,
      top: 12,
      margin: 0,
      padding: 4,
      border: 'none',
      background: 'transparent',
      cursor: 'pointer',
      selectors: {
        [smallScreenSelector]: {
          display: 'block'
        },
        i: {
          fontSize: FontSizes.mediumPlus,
          color: palette.neutralPrimary
        },
        '.is-collapsed &': {
          top: 0,
          bottom: 0,
          left: 0,
          width: '100%',
          selectors: {
            i: {
              display: 'none'
            }
          }
        }
      }
    },
    visuallyHidden: [hiddenContentStyle]
  };
};
