// @business-app/fabric/src/components/Card/CardHeader/CardHeader.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IStyleFunction } from '@fluentui/react/lib/Utilities';

import { CardHeaderProps, CardHeaderStyles } from './CardHeader.types';

export const styles: IStyleFunction<CardHeaderProps, CardHeaderStyles> = ({
  className,
  height,
  title,
  subTitle,
  theme
}) => {
  const { palette } = theme;

  const headerHeight = height ?? subTitle ? 46 : 34;
  const objectHeights = subTitle ? 36 : 24;
  const titleImageSize = title && subTitle ? 36 : 24;
  const subTitleImageSize = 12;

  const textColor = palette.black;

  return {
    headerContainer: [
      'ba-CardHeader',
      {
        height: headerHeight,
        paddingBlockStart: 10,
        paddingInlineStart: 16,
        paddingInlineEnd: 16,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'space-between',
        boxSizing: 'border-box'
      },
      className
    ],
    titleContainer: {
      display: 'block',
      flexDirection: 'column',
      flexGrow: '1'
    },
    titleIcon: {
      width: titleImageSize,
      height: titleImageSize,
      marginInlineEnd: 8,
      color: textColor,
      fontSize: titleImageSize
    },
    title: {
      padding: 0,
      display: 'block',
      color: textColor,
      fontSize: 14,
      fontWeight: 'bold',
      lineHeight: 20
    },
    subTitleContainer: {
      marginTop: 2,
      display: 'flex',
      alignItems: 'center',
      flexGrow: '1'
    },
    subTitleIcon: {
      width: subTitleImageSize,
      height: subTitleImageSize,
      marginInlineEnd: 5,
      color: textColor,
      fontSize: subTitleImageSize
    },
    subTitle: {
      padding: 0,
      display: 'inline-block',
      color: textColor,
      fontSize: 12,
      fontWeight: 'normal',
      lineHeight: 16
    },
    actionMenuContainer: {
      height: '100%',
      display: 'flex',
      selectors: {
        a: {
          lineHeight: objectHeights
        }
      }
    },
    actionMenuItem: {
      marginRight: 5,
      selectors: {
        ':last-child': {
          marginRight: 0
        }
      }
    },
    actionMenuLink: {
      lineHeight: objectHeights
    },
    actionMenuButton: {
      width: objectHeights,
      maxWidth: objectHeights,
      height: objectHeights,
      maxHeight: objectHeights,
      padding: 0
    }
  };
};
