// @business-app/fabric/lib/components/NavigationController/examples/NavigationController.Custom.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import { IconButton, PrimaryButton } from '@fluentui/react/lib/Button';
import { Label } from '@fluentui/react/lib/Label';
import { mergeStyles } from '@fluentui/react/lib/Styling';
import * as React from 'react';
import { NavigationController } from '..';
import { NavigationControllerView } from '../NavigationControllerView';
var stepKeys = {
    intro: 'intro',
    end: 'end'
};
var NavigationControllerCustomExample = /** @class */ (function (_super) {
    __extends(NavigationControllerCustomExample, _super);
    function NavigationControllerCustomExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onRenderHeader = function (NavigationControllerProps, viewProps, context) {
            return (React.createElement("div", { style: { display: 'flex', justifyContent: 'space-between' } },
                React.createElement(IconButton, { iconProps: {
                        iconName: 'CaretSolidLeft'
                    }, disabled: !context.canGoBack(), onClick: context.back, ariaLabel: "Back" }),
                React.createElement("div", null,
                    React.createElement("b", null, viewProps.headerText)),
                React.createElement(IconButton, { iconProps: {
                        iconName: 'CaretSolidRight'
                    }, disabled: !context.canGoForward(), onClick: context.forward, ariaLabel: "Forward" })));
        };
        return _this;
    }
    NavigationControllerCustomExample.prototype.render = function () {
        var containerStyles = {
            border: '1px solid black',
            height: '450px',
            maxWidth: '800px',
            backgroundColor: '#FFF',
            fontSize: '16px',
            lineHeight: '20px'
        };
        return (React.createElement("div", { className: mergeStyles(containerStyles) },
            React.createElement(NavigationController, { initialActiveViewKey: stepKeys.intro, showBackButton: true, onRenderHeader: this.onRenderHeader, backButtonAriaLabel: 'Back' }, function (NavigationControllerContext) { return [
                React.createElement(NavigationControllerView, { key: stepKeys.intro, viewKey: stepKeys.intro, headerText: "First step" },
                    React.createElement("div", { style: {
                            display: 'flex',
                            height: '100%',
                            flexDirection: 'column',
                            justifyContent: 'space-around'
                        } },
                        React.createElement("div", null,
                            React.createElement(Label, null, "Look at that fancy header."),
                            React.createElement(Label, null, "There is no footer for this step.")),
                        React.createElement(PrimaryButton, { onClick: function () { return NavigationControllerContext.goTo(stepKeys.end); }, text: "Next Step", ariaLabel: "Next Step" }))),
                React.createElement(NavigationControllerView, { key: stepKeys.end, viewKey: stepKeys.end, headerText: "Last step", footerButtons: [
                        {
                            primary: false,
                            onClick: function () { return NavigationControllerContext.reset(stepKeys.intro); },
                            text: 'Start again'
                        }
                    ] },
                    React.createElement("div", null, "This is some more text."))
            ]; })));
    };
    return NavigationControllerCustomExample;
}(React.Component));
export { NavigationControllerCustomExample };
