// @business-app/fabric/lib/components/NavigationController/examples/NavigationController.Form.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import { Label } from '@fluentui/react/lib/Label';
import { mergeStyles } from '@fluentui/react/lib/Styling';
import { TextField } from '@fluentui/react/lib/TextField';
import * as React from 'react';
import { NavigationController } from '..';
import { NavigationControllerView } from '../NavigationControllerView';
var stepsKeys = {
    basic: 'basic',
    moreInfo: 'moreInfo',
    summary: 'summary'
};
var NavigationControllerFormExample = /** @class */ (function (_super) {
    __extends(NavigationControllerFormExample, _super);
    function NavigationControllerFormExample(props) {
        var _this = _super.call(this, props) || this;
        _this.renderEmojiCell = function (value, isSelected, onSelect) {
            var emojiCellStyle = {
                width: '30px',
                textAlign: 'center',
                padding: '5px',
                margin: '5px',
                cursor: 'pointer',
                backgroundColor: isSelected ? 'grey' : 'white',
                selectors: {
                    ':hover': {
                        backgroundColor: 'lightgrey'
                    }
                }
            };
            return (React.createElement("div", { onClick: function () { return onSelect(value); }, className: mergeStyles(emojiCellStyle) }, value));
        };
        _this.renderEmojiPicker = function (value, onChange) {
            var emojis = "\uD83D\uDE00 \uD83D\uDE01 \uD83D\uDE02 \uD83E\uDD23 \uD83D\uDE03 \uD83D\uDE04 \uD83D\uDE05 \uD83D\uDE06 \uD83D\uDE09 \uD83D\uDE0A \uD83D\uDE0B \uD83D\uDE0E \uD83D\uDE0D \uD83D\uDE18 \uD83D\uDE17 \uD83D\uDE19 \uD83D\uDE1A \u263A\uFE0F \uD83D\uDE42 \uD83E\uDD17 \uD83E\uDD29 \uD83E\uDD14 \uD83E\uDD28 \uD83D\uDE10 \uD83D\uDE11 \uD83D\uDE36 \uD83D\uDE44 \uD83D\uDE0F \uD83D\uDE23 \uD83D\uDE25 \uD83D\uDE2E \uD83E\uDD10 \uD83D\uDE2F \uD83D\uDE2A \uD83D\uDE2B \uD83D\uDE34 \uD83D\uDE0C \uD83D\uDE1B \uD83D\uDE1C \uD83D\uDE1D \uD83E\uDD24 \uD83D\uDE12 \uD83D\uDE13 \uD83D\uDE14 \uD83D\uDE15 \uD83D\uDE43 \uD83E\uDD11 \uD83D\uDE32 \u2639\uFE0F \uD83D\uDE41 \uD83D\uDE16 \uD83D\uDE1E \uD83D\uDE1F \uD83D\uDE24 \uD83D\uDE22 \uD83D\uDE2D \uD83D\uDE26 \uD83D\uDE27 \uD83D\uDE28 \uD83D\uDE29 \uD83E\uDD2F \uD83D\uDE2C \uD83D\uDE30 \uD83D\uDE31 \uD83E\uDD75 \uD83E\uDD76 \uD83D\uDE33 \uD83E\uDD2A \uD83D\uDE35 \uD83D\uDE21 \uD83D\uDE20 \uD83E\uDD2C \uD83D\uDE37 \uD83E\uDD12 \uD83E\uDD15 \uD83E\uDD22 \uD83E\uDD2E \uD83E\uDD27 \uD83D\uDE07 \uD83E\uDD20 \uD83E\uDD21 \uD83E\uDD25 \uD83E\uDD2B \uD83E\uDD2D \uD83E\uDDD0 \uD83E\uDD13 \uD83D\uDE08 \uD83D\uDC7F \uD83D\uDC79 \uD83D\uDC7A \uD83D\uDC80 \uD83D\uDC7B \uD83D\uDC7D \uD83E\uDD16 \uD83D\uDCA9 \uD83D\uDE3A \uD83D\uDE38 \uD83D\uDE39 \uD83D\uDE3B \uD83D\uDE3C \uD83D\uDE3D \uD83D\uDE40 \uD83D\uDE3F \uD83D\uDE3E".split(' ');
            return (React.createElement("div", { style: { display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' } }, emojis.map(function (emoji) {
                return _this.renderEmojiCell(emoji, emoji === value, onChange);
            })));
        };
        _this.getNameErrorMessage = function (value) {
            return value.length > 3
                ? ''
                : "Input value length must be more than 3. Actual length is " + value.length + ".";
        };
        _this.getAgeErrorMessage = function (value) {
            return value < 21 ? '' : "Age must be less than 21.";
        };
        _this.NavigationControllerRef = React.createRef();
        _this.state = {
            name: '',
            age: NaN,
            favoriteEmoji: ''
        };
        return _this;
    }
    NavigationControllerFormExample.prototype.render = function () {
        var _this = this;
        var containerStyles = {
            border: '1px solid black',
            height: '500px',
            width: '300px',
            backgroundColor: '#FFF',
            fontSize: '16px',
            lineHeight: '20px'
        };
        return (React.createElement(React.Fragment, null,
            React.createElement("div", { className: mergeStyles(containerStyles) },
                React.createElement(NavigationController, { initialActiveViewKey: stepsKeys.basic, componentRef: this.NavigationControllerRef, showBackButton: true, backButtonAriaLabel: 'Back' }, function (context) { return [
                    React.createElement(NavigationControllerView, { viewKey: stepsKeys.basic, key: stepsKeys.basic, headerText: "Basic Details", footerButtons: [
                            {
                                primary: true,
                                onClick: function () { return context.goTo(stepsKeys.moreInfo); },
                                text: 'Next',
                                disabled: !!(_this.getNameErrorMessage(_this.state.name) || _this.getAgeErrorMessage(_this.state.age))
                            }
                        ] },
                        React.createElement(TextField, { key: "name", value: _this.state.name, onChange: function (event, value) { return _this.setState({ name: value }); }, onGetErrorMessage: _this.getNameErrorMessage, label: "Name" }),
                        React.createElement(TextField, { key: "age", value: _this.state.age ? _this.state.age.toString() : '', onChange: function (event, value) { return _this.setState({ age: Number(value) }); }, onGetErrorMessage: function (value) { return _this.getAgeErrorMessage(value === '' ? NaN : Number(value)); }, label: "Age" })),
                    React.createElement(NavigationControllerView, { key: stepsKeys.moreInfo, viewKey: stepsKeys.moreInfo, headerText: "Favorite Emoji", footerButtons: [
                            {
                                primary: true,
                                onClick: function () { return context.goTo(stepsKeys.summary); },
                                text: 'Done',
                                disabled: !_this.state.favoriteEmoji
                            }
                        ] }, _this.renderEmojiPicker(_this.state.favoriteEmoji, function (value) {
                        return _this.setState({ favoriteEmoji: value });
                    })),
                    React.createElement(NavigationControllerView, { key: stepsKeys.summary, viewKey: stepsKeys.summary, headerText: "Submit", footerButtons: [
                            {
                                primary: true,
                                onClick: function () { return alert('NavigationController is complete.'); },
                                text: 'Done'
                            },
                            {
                                primary: false,
                                onClick: function () {
                                    if (_this.NavigationControllerRef.current) {
                                        _this.NavigationControllerRef.current.reset(stepsKeys.basic);
                                        _this.setState({
                                            name: '',
                                            age: NaN,
                                            favoriteEmoji: ''
                                        });
                                    }
                                },
                                text: 'Start again'
                            }
                        ] },
                        React.createElement(Label, null, "Review your selections here:"),
                        React.createElement("div", null,
                            React.createElement(Label, null,
                                React.createElement("b", null, "Name: "),
                                _this.state.name),
                            React.createElement(Label, null,
                                React.createElement("b", null, "Age: "),
                                _this.state.age),
                            React.createElement(Label, null,
                                React.createElement("b", null, "Favorite Emoji: "),
                                _this.state.favoriteEmoji)))
                ]; }))));
    };
    return NavigationControllerFormExample;
}(React.Component));
export { NavigationControllerFormExample };
