// @business-app/fabric/lib/components/Card/NotificationCard/NotificationCard.base.js
import { __assign } from "tslib";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import * as React from 'react';
import { Icon } from '@fluentui/react/lib/Icon';
import { Text } from '@fluentui/react/lib/Text';
import { Link } from '@fluentui/react/lib/Link';
import { IconButton } from '@fluentui/react/lib/Button';
import { ProgressIndicator } from '@fluentui/react/lib/ProgressIndicator';
import { Spinner, SpinnerSize } from '@fluentui/react/lib/Spinner';
import { Persona, PersonaSize } from '@fluentui/react/lib/Persona';
import { classNamesFunction } from '@fluentui/react/lib/Utilities';
import { Card } from './../Card';
var getClassNames = classNamesFunction();
export var NotificationCardBase = function (props) {
    var className = props.className, theme = props.theme, styles = props.styles, isNew = props.isNew, notificationType = props.notificationType, message = props.message, timestamp = props.timestamp, heading = props.heading, subHeading = props.subHeading, actionLinks = props.actionLinks, hasCloseButton = props.hasCloseButton, hasSnoozeButton = props.hasSnoozeButton, iconProps = props.iconProps, persona = props.persona, spinner = props.spinner, progressStatus = props.progressStatus, percentComplete = props.percentComplete;
    var classNames = getClassNames(styles, {
        theme: theme,
        className: className,
        isNew: isNew,
        notificationType: notificationType
    });
    var cardRef = React.createRef();
    var onSnooze = React.useCallback(function () { var _a; return (_a = props.onSnooze) === null || _a === void 0 ? void 0 : _a.call(props); }, [props.onSnooze]);
    var onDismiss = React.useCallback(function () { var _a; return (_a = props.onDismiss) === null || _a === void 0 ? void 0 : _a.call(props); }, [props.onDismiss]);
    var onMouseEnter = React.useCallback(function (event) { var _a; return (_a = props.onMouseEnter) === null || _a === void 0 ? void 0 : _a.call(props, event); }, [props.onMouseEnter]);
    var onMouseLeave = React.useCallback(function (event) { var _a; return (_a = props.onMouseLeave) === null || _a === void 0 ? void 0 : _a.call(props, event); }, [props.onMouseLeave]);
    return (React.createElement(Card, { ref: cardRef, className: classNames.root, width: 308, onMouseEnter: onMouseEnter, onMouseLeave: onMouseLeave },
        React.createElement("div", { className: classNames.imageWrapper },
            (iconProps || notificationType === 'Meeting' || (notificationType === 'Progress' && !spinner)) && (React.createElement(Icon, __assign({ className: classNames.icon, iconName: (notificationType === 'Meeting' && 'Event') ||
                    (notificationType === 'Progress' && 'CloudUpload') }, iconProps))),
            spinner && React.createElement(Spinner, { className: classNames.icon, size: SpinnerSize.medium }),
            persona && React.createElement(Persona, __assign({ className: classNames.persona, size: PersonaSize.size32 }, persona))),
        React.createElement("div", { className: classNames.timeDismissWrapper },
            timestamp && React.createElement("div", { className: classNames.timestamp }, timestamp),
            hasCloseButton && (React.createElement(IconButton, { className: classNames.closeButton, title: 'Close', ariaLabel: 'Close', iconProps: { iconName: 'ChromeClose' }, onClick: onDismiss, styles: {
                    icon: {
                        fontSize: theme.fonts.small.fontSize
                    }
                } })),
            hasSnoozeButton && (React.createElement(IconButton, { className: classNames.closeButton, title: 'Snooze', ariaLabel: 'Snooze', iconProps: { iconName: 'AlarmClock' }, onClick: onSnooze }))),
        React.createElement("div", { className: classNames.contentWrapper },
            React.createElement("div", { className: classNames.cardHeadingWrapper },
                React.createElement("label", { className: classNames.cardHeading },
                    notificationType === 'Peer' && React.createElement(Icon, { className: classNames.peerIcon, iconName: "Accounts" }),
                    heading),
                subHeading && React.createElement("span", { className: classNames.cardSubHeading }, subHeading)),
            message && !spinner && (React.createElement(Text, { className: classNames.messageWrapper, variant: 'medium' }, message)),
            notificationType === 'Progress' && !spinner && (React.createElement(ProgressIndicator, { percentComplete: percentComplete })),
            progressStatus && !spinner && React.createElement("span", { className: classNames.cardSubHeading }, progressStatus),
            notificationType === 'Progress' && spinner && (React.createElement(Text, { className: classNames.messageWrapper, variant: 'medium' }, message)), actionLinks === null || actionLinks === void 0 ? void 0 :
            actionLinks.map(function (_a) {
                var link = _a.link, text = _a.text;
                return (React.createElement(Link, { key: text, className: classNames.actionButtons, href: link }, text));
            }))));
};
