// @business-app/fabric/lib-cjs/components/Wizard/Footer/WizardFooter.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = void 0;
var getStyles = function (props) {
    var _a;
    var palette = props.theme.palette;
    var smallScreenSelector = '@media (max-width: 767px)';
    return {
        root: [
            'ba-wizard-footer',
            props.className,
            {
                boxSizing: 'border-box',
                display: 'flex',
                flex: '0 0 72px',
                alignItems: 'center',
                padding: '0 24px',
                borderTop: "1px solid " + palette.neutralLight,
                background: palette.white,
                selectors: (_a = {},
                    _a[smallScreenSelector] = {
                        flex: '0 0 56px'
                    },
                    _a)
            }
        ],
        button: {
            margin: '0 16px 0 0'
        },
        summaryText: {
            color: palette.neutralDark,
            selectors: {
                '@media (max-width: 639px)': {
                    display: 'none' // hide summary on mobile
                }
            }
        }
    };
};
exports.getStyles = getStyles;
