// @business-app/fabric/lib/components/FilterSearchBox/examples/FilterSearchBox.SingleFilters.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { Label } from '@fluentui/react/lib/Label';
import { FilterSearchBox } from '../';
var filters = [{ key: 'only', text: 'Only Filter' }];
var FilterSearchBoxSingleFiltersExample = /** @class */ (function (_super) {
    __extends(FilterSearchBoxSingleFiltersExample, _super);
    function FilterSearchBoxSingleFiltersExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            searchText: '',
            selectedFilterKey: filters[0].key
        };
        _this._onSearchTextChange = function (event, text) {
            _this.setState({ searchText: text });
        };
        _this._onFilterChange = function (key) {
            _this.setState({ selectedFilterKey: key });
        };
        return _this;
    }
    FilterSearchBoxSingleFiltersExample.prototype.render = function () {
        var _a = this.state, searchText = _a.searchText, selectedFilterKey = _a.selectedFilterKey;
        return (React.createElement(React.Fragment, null,
            React.createElement(FilterSearchBox, { strings: {
                    searchBoxPlaceholder: 'Search',
                    filterButtonTooltipText: 'Tooltip text'
                }, searchText: searchText, onSearchTextChange: this._onSearchTextChange, filters: filters, selectedFilterKey: selectedFilterKey, onFilterChange: this._onFilterChange, styles: {
                    root: {
                        marginBottom: 16
                    }
                } }),
            React.createElement(Label, null, "Current search text: " + searchText),
            React.createElement(Label, null, "Current filter key: " + selectedFilterKey)));
    };
    return FilterSearchBoxSingleFiltersExample;
}(React.Component));
export { FilterSearchBoxSingleFiltersExample };
