// @business-app/fabric/lib/components/FilterSearchBox/examples/FilterSearchBox.LongFilterText.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { Label } from '@fluentui/react/lib/Label';
import { Toggle } from '@fluentui/react/lib/Toggle';
import { FilterSearchBox } from '../';
var filters = [
    {
        key: 'one',
        text: 'Filter 1Filter 1Filter 1Filter 1Filterter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1Filter 1'
    },
    { key: 'two', text: 'Filter 2' },
    { key: 'three', text: 'Filter 3' }
];
var FilterSearchBoxLongFilterTextExample = /** @class */ (function (_super) {
    __extends(FilterSearchBoxLongFilterTextExample, _super);
    function FilterSearchBoxLongFilterTextExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            searchText: '',
            selectedFilterKey: filters[0].key,
            filterCalloutMatchComponentWidth: false
        };
        _this._onSearchTextChange = function (event, text) {
            _this.setState({ searchText: text });
        };
        _this._onFilterChange = function (key) {
            _this.setState({ selectedFilterKey: key });
        };
        _this._onUseFullWidthChange = function (_, checked) {
            _this.setState({ filterCalloutMatchComponentWidth: checked });
        };
        return _this;
    }
    FilterSearchBoxLongFilterTextExample.prototype.render = function () {
        var _a = this.state, searchText = _a.searchText, selectedFilterKey = _a.selectedFilterKey, filterCalloutMatchComponentWidth = _a.filterCalloutMatchComponentWidth;
        return (React.createElement(React.Fragment, null,
            React.createElement(FilterSearchBox, { strings: {
                    searchBoxPlaceholder: 'Search'
                }, searchText: searchText, onSearchTextChange: this._onSearchTextChange, filters: filters, selectedFilterKey: selectedFilterKey, onFilterChange: this._onFilterChange, filterCalloutMatchComponentWidth: filterCalloutMatchComponentWidth, styles: {
                    root: {
                        marginBottom: 16
                    }
                } }),
            React.createElement(Label, null, "Current search text: " + searchText),
            React.createElement(Label, null, "Current filter key: " + selectedFilterKey),
            React.createElement(Toggle, { checked: filterCalloutMatchComponentWidth, onChange: this._onUseFullWidthChange, label: 'Filter callout should match component width' })));
    };
    return FilterSearchBoxLongFilterTextExample;
}(React.Component));
export { FilterSearchBoxLongFilterTextExample };
