// @business-app/fabric/src/components/Accordion/Accordion.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IButtonStyles } from '@fluentui/react/lib/Button';
import { ITheme } from '@fluentui/react/lib/Styling';
import { IAccordionStyleProps, IAccordionStyles } from './Accordion.types';

export const styles = ({ containerClassName, isContentVisible }: IAccordionStyleProps): IAccordionStyles => ({
  container: [
    'ba-Accordion',
    isContentVisible && 'ba-Accordion--contentVisible',
    containerClassName,
    {
      width: '100%'
    }
  ],
  header: [
    'ba-Accordion-header',
    {
      width: '100%',
      minWidth: 0,
      textAlign: 'left'
    }
  ],
  content: [
    'ba-Accordion-content',
    {
      selectors: {
        '> *': {
          width: '100%',
          display: 'block'
        }
      }
    }
  ]
});

export function getDefaultHeaderStyles(theme: ITheme): IButtonStyles {
  return {
    root: {
      borderColor: 'transparent',
      backgroundColor: theme.palette.neutralLighter
    },
    rootHovered: {
      backgroundColor: theme.palette.neutralQuaternaryAlt
    }
  };
}
