// @business-app/fabric/src/components/LoadingButton/examples/LoadingButton.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { CommandBar } from '@fluentui/react/lib/CommandBar';
import { Stack } from '@fluentui/react/lib/Stack';
import { Toggle } from '@fluentui/react/lib/Toggle';
import * as React from 'react';
import { LoadingCommandButton, LoadingPrimaryButton } from '../LoadingButton';

export class LoadingButtonExample extends React.Component<{}, { isLoading: boolean }> {
  public state = {
    isLoading: false
  };

  public render(): JSX.Element {
    return (
      <Stack maxWidth={400} gap={16}>
        <LoadingPrimaryButton
          iconProps={{ iconName: 'Save' }}
          text="Save data"
          loading={this.state.isLoading}
          styles={{ root: { width: 130 } }}
        />
        <CommandBar
          items={[
            {
              key: 'add',
              iconProps: { iconName: 'Add' },
              text: 'Add item',
              loading: this.state.isLoading
            },
            {
              key: 'edit',
              iconProps: { iconName: 'Edit' },
              text: 'Edit item',
              loading: this.state.isLoading
            },
            {
              key: 'remove',
              iconProps: { iconName: 'Delete' },
              text: 'Remove item',
              loading: this.state.isLoading
            }
          ]}
          buttonAs={LoadingCommandButton}
        />
        <Toggle
          label="Toggle loading"
          checked={!!this.state.isLoading}
          onClick={() => this.setState({ isLoading: !this.state.isLoading })}
        />
      </Stack>
    );
  }
}
