// @business-app/fabric/lib/components/Tree/examples/Tree.AlternatingRowColorExample.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import { mergeStyleSets } from '@fluentui/react/lib/Styling';
import * as React from 'react';
import { Tree } from '../Tree';
import { produceItemsWithIcons } from './Tree.Example.Data';
var TreeAlternatingRowColorExample = /** @class */ (function (_super) {
    __extends(TreeAlternatingRowColorExample, _super);
    function TreeAlternatingRowColorExample(props) {
        var _this = _super.call(this, props) || this;
        _this._getItemStyle = function (item, isSelected, isExpanded, index) {
            var styles = customizedStyles;
            if (isSelected) {
                return styles.oddRow;
            }
            else if (index % 2 === 1) {
                return styles.oddRow;
            }
            else {
                return styles.evenRow;
            }
        };
        _this._getCommands = function () {
            var selection = _this.state.selection;
            var keys = ['newItem', 'share', 'mobile'];
            var items = [
                {
                    key: keys[0],
                    text: 'News',
                    canCheck: true,
                    isChecked: selection[keys[0]],
                    onClick: _this._onToggleSelect,
                    icon: 'Add'
                },
                {
                    key: keys[1],
                    text: 'Share',
                    canCheck: true,
                    isChecked: selection[keys[1]],
                    onClick: _this._onToggleSelect,
                    icon: 'Mail'
                },
                {
                    key: keys[2],
                    text: 'Mobile',
                    canCheck: true,
                    isChecked: selection[keys[2]],
                    onClick: _this._onToggleSelect,
                    icon: 'Calendar'
                }
            ];
            return items;
        };
        _this._onToggleSelect = function (ev, item) {
            var selection = _this.state.selection;
            if (ev.preventDefault) {
                ev.preventDefault();
            }
            selection[item.key] = !selection[item.key];
            _this.setState({
                selection: selection
            });
        };
        _this.state = {
            selection: {}
        };
        return _this;
    }
    TreeAlternatingRowColorExample.prototype.render = function () {
        return (React.createElement("div", { className: exampleStyles.exampleContainer },
            React.createElement(Tree, { items: produceItemsWithIcons, strings: {
                    listAriaLabel: 'test',
                    moreCommandsAriaLabel: 'More Command'
                }, getItemCommands: this._getCommands, getItemStyle: this._getItemStyle })));
    };
    return TreeAlternatingRowColorExample;
}(React.Component));
export { TreeAlternatingRowColorExample };
var exampleStyles = mergeStyleSets({
    exampleContainer: {
        // 'display: flex' is important for the Tree to layout correctly.
        display: 'flex',
        flexDirection: 'column',
        width: 320,
        height: 320,
        userSelect: 'none'
    }
});
var customizedStyles = mergeStyleSets({
    oddRow: {},
    evenRow: {
        backgroundColor: '#FBFBFB'
    },
    selectedRow: {
        backgroundColor: '#DCDCDC'
    }
});
