// @business-app/fabric/lib/components/Tree/examples/Tree.Example.Data.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __assign } from "tslib";
export var produceItems = [
    {
        label: 'Fruits',
        key: 'fruits',
        isFolder: true,
        level: 0,
        setSize: 2,
        indexInParent: 0
    },
    {
        label: 'Apples',
        key: 'apples',
        isFolder: true,
        level: 1,
        setSize: 0,
        indexInParent: 0
    },
    {
        label: 'Granny Smith',
        key: 'granny smith',
        level: 2,
        setSize: 2,
        indexInParent: 0
    },
    {
        label: 'Red Delicious',
        key: 'red delicious',
        level: 2,
        setSize: 2,
        indexInParent: 1
    },
    {
        label: 'Vegetables',
        key: 'vegetables',
        isFolder: true,
        level: 0,
        setSize: 2,
        indexInParent: 1
    },
    {
        label: 'Carrots',
        key: 'carrots',
        level: 1,
        setSize: 3,
        indexInParent: 0
    },
    {
        label: 'Cucumbers',
        key: 'cucumbers',
        level: 1,
        setSize: 3,
        indexInParent: 1
    },
    {
        label: 'Lettuce',
        key: 'lettuce',
        level: 1,
        setSize: 3,
        indexInParent: 2
    }
];
export var produceItemsWithIcons = produceItems.map(function (item) {
    return __assign(__assign({}, item), { iconProps: { iconName: 'Info' } });
});
export function getFolderKeys(items) {
    var folderKeys = new Set();
    items.forEach(function (item) {
        if (item.isFolder) {
            folderKeys.add(item.key);
        }
    });
    return folderKeys;
}
export function cloneSet(set) {
    var newSet = new Set();
    set.forEach(function (key) { return newSet.add(key); });
    return newSet;
}
