// @business-app/fabric/src/components/FilterSearchBox/FilterSearchBox.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IStyle } from '@fluentui/react/lib/Styling';

import { IFilterSearchBoxStyleProps, IFilterSearchBoxStyles } from './FilterSearchBox.types';

const height = 36;
const separatorMargin = 8;

const sharedButtonRootStyles: IStyle = {
  backgroundColor: 'transparent',
  height: '100%',
  padding: '0 8px',
  borderColor: 'transparent'
};

export const getStyles = (props: IFilterSearchBoxStyleProps): IFilterSearchBoxStyles => {
  const { palette } = props.theme;
  return {
    root: {
      width: '100%',
      height,
      display: 'flex',
      flexDirection: 'row',
      border: `1px solid ${palette.neutralQuaternaryAlt}`
    },
    active: {
      borderColor: '#333'
    },
    searchBox: {
      flex: '1 1 auto',
      float: 'left',
      border: 'none',
      height: '100%'
    },
    separator: {
      flex: '0 0 1px',
      height: height - separatorMargin * 2,
      marginTop: separatorMargin,
      // margin right -1 will move it so when the filter button is hovered, the separator is covered and not seen.
      marginRight: -1,
      backgroundColor: palette.neutralQuaternaryAlt
    },
    tooltipHost: {
      height: '100%',
      maxWidth: '50%'
    },
    filterButtonRoot: {
      ...sharedButtonRootStyles,
      minWidth: 0,
      maxWidth: '50%'
    },
    filterButtonRootInTooltipHost: {
      ...sharedButtonRootStyles,
      width: '100%',
      maxWidth: '100%'
    },
    filterButtonLabel: {
      fontWeight: 400,
      textOverflow: 'ellipsis',
      overflow: 'hidden',
      lineHeight: 'normal'
    },
    filterButtonNoLabel: {
      margin: 0
    },
    filterButtonTextContainer: {
      overflow: 'hidden',
      whiteSpace: 'nowrap'
    }
  };
};
