// @business-app/fabric/lib/components/MultistepPanel/MultistepPanel.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __assign, __extends, __rest } from "tslib";
import { Panel } from '@business-app/fabric/lib/components/Panel';
import { BaseComponent } from '@fluentui/react/lib/Utilities';
import * as React from 'react';
import * as Styles from './Multistep.styles';
var MultistepPanel = /** @class */ (function (_super) {
    __extends(MultistepPanel, _super);
    function MultistepPanel(props) {
        return _super.call(this, props) || this;
    }
    MultistepPanel.prototype.componentDidUpdate = function (prevProps) {
        if (prevProps.panelOpenIndex !== this.props.panelOpenIndex) {
            if (this.props.panelOpenIndex < 0 || this.props.panelOpenIndex >= this.props.panelsListProps.length) {
                this.props.onDismiss();
            }
        }
    };
    MultistepPanel.prototype.render = function () {
        var _this = this;
        var _a = this.props, panelsListProps = _a.panelsListProps, panelOpenIndex = _a.panelOpenIndex, setPanelOpenIndex = _a.setPanelOpenIndex, componentRef = _a.componentRef, inheritedPanelProps = __rest(_a, ["panelsListProps", "panelOpenIndex", "setPanelOpenIndex", "componentRef"]);
        if (panelsListProps.length === 0) {
            return null;
        }
        if (panelOpenIndex < 0 || !panelOpenIndex) {
            panelOpenIndex = 0;
        }
        else if (panelOpenIndex >= panelsListProps.length) {
            panelOpenIndex = panelsListProps.length - 1;
        }
        var currentPanelListProps = panelsListProps[panelOpenIndex];
        var footerButtons = currentPanelListProps.footerButtons &&
            currentPanelListProps.footerButtons.map(function (footerProps) {
                var navigateableOnClick = footerProps.navigateableOnClick, restFooterProps = __rest(footerProps, ["navigateableOnClick"]);
                var dismissableOnClick = function (dismissPanel) {
                    if (navigateableOnClick) {
                        var nextIndex = navigateableOnClick(panelOpenIndex, dismissPanel);
                        if (nextIndex) {
                            setPanelOpenIndex(nextIndex);
                        }
                    }
                };
                return __assign(__assign({}, restFooterProps), { dismissableOnClick: dismissableOnClick });
            });
        var headerLeftButtonProps;
        if (!this.props.hideLeftHeaderBackButton && panelOpenIndex > 0) {
            headerLeftButtonProps = {
                ariaLabel: inheritedPanelProps.headerIconAriaLabel,
                iconProps: {
                    iconName: 'Back'
                },
                onClick: function () {
                    var newPanelOpenIndex = panelOpenIndex - 1;
                    if (_this.props.onLeftHeaderBackButtonClicked) {
                        newPanelOpenIndex = _this.props.onLeftHeaderBackButtonClicked(panelOpenIndex, _this.props.onDismiss);
                    }
                    setPanelOpenIndex(newPanelOpenIndex);
                }
            };
        }
        return (React.createElement(Panel, __assign({}, inheritedPanelProps, { headerLeftButtonProps: headerLeftButtonProps }, currentPanelListProps, { footerButtons: footerButtons, styles: __assign(__assign({}, Styles.multistepPanelStyles), currentPanelListProps.styles) })));
    };
    return MultistepPanel;
}(BaseComponent));
export { MultistepPanel };
