// @business-app/fabric/lib-cjs/components/Card/examples/RichCard.With.Sub.Components.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RichCardWithSubComponentsExample = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var ContextualMenu_1 = require("@fluentui/react/lib/ContextualMenu");
var Card_1 = require("@business-app/fabric/lib/components/Card");
var RichCardWithSubComponentsExample = /** @class */ (function (_super) {
    tslib_1.__extends(RichCardWithSubComponentsExample, _super);
    function RichCardWithSubComponentsExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RichCardWithSubComponentsExample.prototype.render = function () {
        return (React.createElement(Card_1.RichCard, { headerProps: {
                title: 'Card title',
                subTitle: 'Card subtitle',
                titleIconProps: {
                    iconName: 'Dictionary'
                },
                subTitleIconProps: {
                    iconName: 'Dictionary'
                },
                actionItemProps: [
                    {
                        type: Card_1.ActionItemType.Link,
                        props: {
                            href: 'https"//bing.com',
                            target: '_blank',
                            title: 'Hyperlink'
                        }
                    },
                    {
                        type: Card_1.ActionItemType.Button,
                        props: {
                            iconProps: {
                                iconName: 'Attach'
                            }
                        }
                    },
                    {
                        type: Card_1.ActionItemType.Button,
                        props: {
                            iconProps: {
                                iconName: 'Add'
                            }
                        }
                    },
                    {
                        type: Card_1.ActionItemType.ContextualMenu,
                        props: {
                            text: 'Content',
                            menuProps: {
                                items: [
                                    {
                                        key: 'emailMessage',
                                        text: 'Email message',
                                        iconProps: { iconName: 'Mail' }
                                    },
                                    {
                                        key: 'calendarEvent',
                                        text: 'Calendar event',
                                        iconProps: { iconName: 'Calendar' }
                                    }
                                ]
                            }
                        }
                    }
                ],
                contextualMenuProps: {
                    items: [
                        {
                            key: 'newItem',
                            text: 'New',
                            onClick: function () { return alert('New clicked'); }
                        },
                        {
                            key: 'divider_1',
                            itemType: ContextualMenu_1.ContextualMenuItemType.Divider
                        },
                        {
                            key: 'rename',
                            text: 'Rename',
                            onClick: function () { return alert('Rename clicked'); }
                        },
                        {
                            key: 'edit',
                            text: 'Edit',
                            onClick: function () { return alert('Edit clicked'); }
                        },
                        {
                            key: 'properties',
                            text: 'Properties',
                            onClick: function () { return alert('Properties clicked'); }
                        },
                        {
                            key: 'linkNoTarget',
                            text: 'Link same window',
                            href: 'http://bing.com'
                        }
                    ]
                }
            }, footerProps: {
                link: {
                    text: 'Hyperlink',
                    href: 'https://bing.com'
                },
                buttons: {
                    primaryButton: {
                        text: 'First Butjon'
                    },
                    secondaryButton: {
                        text: 'Second Butjon'
                    }
                }
            } },
            React.createElement("div", { style: { width: '100%', height: '100%', backgroundColor: '#EEE' } },
                React.createElement("img", { style: { width: '100%' }, src: 'https://placehold.it/600x400/EEE' }))));
    };
    return RichCardWithSubComponentsExample;
}(React.Component));
exports.RichCardWithSubComponentsExample = RichCardWithSubComponentsExample;
