// @business-app/fabric/lib-cjs/components/Card/examples/RichCard.With.Custom.Title.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RichCardWithCustomTitleExample = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var RichCard_1 = require("@business-app/fabric/lib/components/Card/RichCard");
var Icon_1 = require("@fluentui/react/lib/Icon");
var Styling_1 = require("@fluentui/react/lib/Styling");
var iconClass = Styling_1.mergeStyles({
    fontSize: 50,
    height: 50,
    width: 50
});
var RichCardWithCustomTitleExample = /** @class */ (function (_super) {
    tslib_1.__extends(RichCardWithCustomTitleExample, _super);
    function RichCardWithCustomTitleExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RichCardWithCustomTitleExample.prototype.render = function () {
        return (React.createElement(RichCard_1.RichCard, { height: 150, width: 300, headerProps: {
                onRenderTitle: function () {
                    return (React.createElement(React.Fragment, null,
                        React.createElement(Icon_1.FontIcon, { iconName: "FavoriteStarFill" }),
                        React.createElement(Icon_1.FontIcon, { iconName: "FavoriteStarFill" }),
                        React.createElement(Icon_1.FontIcon, { iconName: "FavoriteStarFill" }),
                        React.createElement(Icon_1.FontIcon, { iconName: "FavoriteStar" }),
                        React.createElement(Icon_1.FontIcon, { iconName: "FavoriteStar" })));
                }
            } }, "Title replaced with stars"));
    };
    return RichCardWithCustomTitleExample;
}(React.Component));
exports.RichCardWithCustomTitleExample = RichCardWithCustomTitleExample;
