// @business-app/fabric/lib/components/Resizer/Resizer.types.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
/**
 * Specifies the location of the {@link Resizer} with respect to the resizable element.
 */
export var ResizerLocation;
(function (ResizerLocation) {
    ResizerLocation[ResizerLocation["Top"] = 0] = "Top";
    ResizerLocation[ResizerLocation["Right"] = 1] = "Right";
    ResizerLocation[ResizerLocation["Bottom"] = 2] = "Bottom";
    ResizerLocation[ResizerLocation["Left"] = 3] = "Left";
})(ResizerLocation || (ResizerLocation = {}));
/**
 * Specifies whether the resize direction is horizontal or vertical.
 */
export var ResizeDirection;
(function (ResizeDirection) {
    ResizeDirection[ResizeDirection["Vertical"] = 0] = "Vertical";
    ResizeDirection[ResizeDirection["Horizontal"] = 1] = "Horizontal";
})(ResizeDirection || (ResizeDirection = {}));
/**
 * Specifies if the {@link Resizer} handler attaches inside or outside the edge of the element to be resized.
 */
export var ResizerAttachment;
(function (ResizerAttachment) {
    ResizerAttachment[ResizerAttachment["Inside"] = 0] = "Inside";
    ResizerAttachment[ResizerAttachment["Outside"] = 1] = "Outside";
})(ResizerAttachment || (ResizerAttachment = {}));
