// @business-app/fabric/src/components/Wizard/Footer/WizardFooter.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { WizardFooterStyleProps, WizardFooterStyles } from './WizardFooter.types';

export const getStyles = (props: WizardFooterStyleProps): WizardFooterStyles => {
  const { palette } = props.theme;
  const smallScreenSelector = '@media (max-width: 767px)';

  return {
    root: [
      'ba-wizard-footer',
      props.className,
      {
        boxSizing: 'border-box',
        display: 'flex',
        flex: '0 0 72px',
        alignItems: 'center',
        padding: '0 24px',
        borderTop: `1px solid ${palette.neutralLight}`,
        background: palette.white,
        selectors: {
          [smallScreenSelector]: {
            flex: '0 0 56px'
          }
        }
      }
    ],
    button: {
      margin: '0 16px 0 0'
    },
    summaryText: {
      color: palette.neutralDark,
      selectors: {
        '@media (max-width: 639px)': {
          display: 'none' // hide summary on mobile
        }
      }
    }
  };
};
