// @business-app/fabric/lib/components/SettingsExplorer/examples/TwoTierSettingsExplorer.Basic.Example.js
import { __extends } from "tslib";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { DefaultButton } from '@fluentui/react/lib/Button';
import { Dropdown } from '@fluentui/react/lib/Dropdown';
import { Modal } from '@fluentui/react/lib/Modal';
import { mergeStyles } from '@fluentui/react/lib/Styling';
import * as React from 'react';
import { SettingsExplorer } from '../SettingsExplorer';
import { SettingsItem } from '../SettingsItem';
import { SettingsPage } from '../SettingsPage';
import { TwoTierSettingsExplorer } from '../TwoTierSettingsExplorer';
import { dateFormatOptions, languageOptions, timeFormatOptions } from './ExampleInput';
var TwoTierSettingsExplorerBasicExample = /** @class */ (function (_super) {
    __extends(TwoTierSettingsExplorerBasicExample, _super);
    function TwoTierSettingsExplorerBasicExample(props) {
        var _this = _super.call(this, props) || this;
        _this._openSettingsExplorer = function (status) {
            _this.setState({ modalOpen: status });
        };
        _this.state = { modalOpen: false };
        return _this;
    }
    TwoTierSettingsExplorerBasicExample.prototype.render = function () {
        var _this = this;
        return (React.createElement("div", null,
            React.createElement(DefaultButton, { onClick: function () { return _this._openSettingsExplorer(true); }, text: "Two Tiered", className: mergeStyles({
                    width: 200
                }) }),
            React.createElement(Modal, { isOpen: this.state.modalOpen, onDismiss: function () { return _this._openSettingsExplorer(false); }, closeButtonAriaLabel: 'Cancel' },
                React.createElement(TwoTierSettingsExplorer, { name: 'General Settings', saveButtonProps: {
                        onClick: function () {
                            alert('save clicked');
                        },
                        text: 'Save'
                    }, cancelButtonProps: {
                        onClick: function () {
                            _this._openSettingsExplorer(false);
                        },
                        text: 'Cancel'
                    }, exitButtonProps: {
                        onClick: function () { return _this._openSettingsExplorer(false); },
                        ariaLabel: 'Clear'
                    } },
                    React.createElement(SettingsExplorer, { name: '[Page 1]', id: 'first' },
                        React.createElement(SettingsPage, { name: '[Submenu 1]', id: 'a' },
                            React.createElement(SettingsItem, null,
                                React.createElement(Dropdown, { label: 'Language', options: languageOptions, defaultSelectedKey: 'en-US' })),
                            React.createElement(SettingsItem, null,
                                React.createElement(Dropdown, { label: 'Date format', options: dateFormatOptions, defaultSelectedKey: 'M/D/YYYY' })),
                            React.createElement(SettingsItem, null,
                                React.createElement(Dropdown, { label: 'Time format', options: timeFormatOptions, defaultSelectedKey: '1:01 AM - 11:59 PM' }))),
                        React.createElement(SettingsPage, { name: '[Submenu 2]' }),
                        React.createElement(SettingsPage, { name: '[Submenu 3]' }),
                        React.createElement(SettingsPage, { name: '[Submenu 4]' }),
                        React.createElement(SettingsPage, { name: '[Submenu 5]' })),
                    React.createElement(SettingsExplorer, { name: '[Page 2]', id: 'second' },
                        React.createElement(SettingsPage, { name: '[Submenu 1]' }),
                        React.createElement(SettingsPage, { name: '[Submenu 2]' })),
                    React.createElement(SettingsExplorer, { name: '[Page 3]', id: 'third' },
                        React.createElement(SettingsPage, { name: '[Submenu 1]' }),
                        React.createElement(SettingsPage, { name: '[Submenu 2]' }))))));
    };
    return TwoTierSettingsExplorerBasicExample;
}(React.Component));
export { TwoTierSettingsExplorerBasicExample };
