// @business-app/fabric/lib-cjs/components/FilterableComboBox/examples/FilterableComboBox.Grouped.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterableComboBoxGroupedExample = void 0;
var tslib_1 = require("tslib");
var Label_1 = require("@fluentui/react/lib/Label");
var SelectableOption_1 = require("@fluentui/react/lib/SelectableOption");
var Stack_1 = require("@fluentui/react/lib/Stack");
var React = require("react");
var FilterableComboBox_1 = require("../FilterableComboBox");
var tags = [
    '_Colors',
    'Azure',
    'Black',
    'Cerulean',
    'Gray',
    'Green',
    'Lilac',
    'Rose',
    'Ruby',
    'Yellow',
    'オレンジ',
    '青い',
    '&div1',
    '_Flavors',
    'Berry',
    'Chocolate',
    'Pineapple',
    'Strawberry',
    'Vanilla',
    '&div2',
    '_Emotions',
    'Angry',
    'Bored',
    'Happy',
    'Lethargic',
    'Nonplussed',
    'Peaceful',
    'Restless',
    'Sad'
].map(function (name) { return ({
    key: name.toLowerCase(),
    text: name.indexOf('_') === 0 ? name.substr(1) : name,
    itemType: name.indexOf('_') === 0
        ? SelectableOption_1.SelectableOptionMenuItemType.Header
        : name.indexOf('&') === 0
            ? SelectableOption_1.SelectableOptionMenuItemType.Divider
            : SelectableOption_1.SelectableOptionMenuItemType.Normal
}); });
var FilterableComboBoxGroupedExample = /** @class */ (function (_super) {
    tslib_1.__extends(FilterableComboBoxGroupedExample, _super);
    function FilterableComboBoxGroupedExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            selectedKey: 'vanilla',
            selectedOption: { key: 'vanilla', text: 'Vanilla' }
        };
        return _this;
    }
    FilterableComboBoxGroupedExample.prototype.render = function () {
        var _this = this;
        return (React.createElement(Stack_1.Stack, { maxWidth: 400, gap: 16 },
            React.createElement(FilterableComboBox_1.FilterableComboBox, { label: 'Choose an option', options: tags, noResultsMessage: function (query) { return (React.createElement(Label_1.Label, { styles: { root: { padding: '8px 16px' } } }, "No results found matching \"" + query + "\".")); }, calloutProps: { calloutMaxHeight: 400 }, revertToLastValidOptionOnInvalidFilter: true, selectedKey: this.state.selectedKey, useComboBoxAsMenuWidth: true, onChange: function (evt, option) {
                    return option && _this.setState({ selectedKey: option.key, selectedOption: option });
                } }),
            React.createElement("p", null,
                React.createElement(Label_1.Label, null,
                    "Last selected option: ",
                    this.state.selectedOption && this.state.selectedOption.text))));
    };
    return FilterableComboBoxGroupedExample;
}(React.Component));
exports.FilterableComboBoxGroupedExample = FilterableComboBoxGroupedExample;
