// @business-app/fabric/src/components/NavigationController/NavigationController.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { FontWeights } from '@fluentui/react/lib/Styling';
import { ITheme } from '@fluentui/react/lib/Styling';

import { INavigationControllerStyles } from './NavigationController.types';

export const getStyles = (props: { theme: ITheme }): INavigationControllerStyles => {
  const { palette } = props.theme;
  return {
    root: {
      width: '100%',
      height: '100%',
      display: 'flex',
      flexDirection: 'column',
      boxSizing: 'border-box'
    },
    headerContainer: {
      flexGrow: 0,
      boxSizing: 'border-box',
      minHeight: '52px',
      paddingBottom: '8px',
      paddingTop: '16px',
      paddingLeft: '24px',
      paddingRight: '24px',
      textOverflow: 'ellipsis',
      overflow: 'hidden',
      whiteSpace: 'nowrap'
    },
    headerText: {
      fontSize: '20px',
      lineHeight: '28px',
      color: palette.neutralPrimary,
      fontWeight: FontWeights.semibold
    },
    headerBackButton: {
      marginRight: '8px',
      color: palette.neutralPrimary
    },
    footerContainer: {
      boxSizing: 'border-box',
      flexGrow: 0,
      height: '64px',
      padding: '16px 24px',
      borderTopWidth: '1px',
      borderTopColor: palette.neutralLight,
      borderTopStyle: 'solid'
    },
    bodyContainer: {
      flexGrow: 1,
      overflowY: 'auto',
      width: '100%',
      height: '100%',
      padding: '8px 24px',
      boxSizing: 'border-box'
    }
  };
};
