// @business-app/fabric/src/components/FilterSearchViews/examples/FilterSearchViewsExampleData.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IFilterItem } from '../../FilterSearchBox';
import { IItem } from '../index';

export interface IDataItem extends IItem {
  id: string;
  text: string;
  iconName: string;
}

export const arrows: IDataItem[] = [
  { id: '1', text: 'Up', iconName: 'Up' },
  { id: '2', text: 'Down', iconName: 'Down' },
  { id: '3', text: 'Right', iconName: 'Forward' },
  { id: '4', text: 'Left', iconName: 'Back' },
  { id: '5', text: 'Chevron down', iconName: 'ChevronDown' },
  { id: '6', text: 'Chevron up', iconName: 'ChevronUp' },
  { id: '7', text: 'Chevron left', iconName: 'ChevronLeft' },
  { id: '8', text: 'Chevron right', iconName: 'ChevronRight' }
];

export const charts: IDataItem[] = [
  { id: '9', text: 'Chart', iconName: 'Chart' },
  { id: '10', text: 'Area chart', iconName: 'AreaChart' },
  { id: '11', text: 'Pivot chart', iconName: 'PivotChart' },
  { id: '12', text: 'Line chart', iconName: 'LineChart' },
  { id: '13', text: 'Horizontal bar chart', iconName: 'BarChartHorizontal' },
  { id: '14', text: 'Vertical bar chart', iconName: 'BarChartVertical' },
  { id: '15', text: 'Chart series', iconName: 'ChartSeries' },
  { id: '16', text: 'Donut chart', iconName: 'DonutChart' }
];

export const groupedItems: IDataItem[] = [
  { id: 'arrowsId', text: 'Arrows', iconName: '', children: arrows },
  { id: 'chartsId', text: 'Charts', iconName: '', children: charts }
];

export const flatItems: IDataItem[] = arrows.concat(charts);

export const filters: IFilterItem[] = [
  { key: 'arrows', text: 'Arrows' },
  { key: 'charts', text: 'Charts' }
];

export const shouldFilterItem = (item: IDataItem, text: string, filter: IFilterItem): boolean => {
  // Groups should be visible only if any children match filter
  if (item.children) {
    return item.children.length > 0;
  }

  if (!!text && item.text.toLowerCase().indexOf(text.toLowerCase()) === -1) {
    return false;
  }

  if (!!filter) {
    const list = filter.key === 'arrows' ? arrows : charts;
    for (let i = 0; i < list.length; i++) {
      if (list[i].id === item.id) {
        return true;
      }
    }
    return false;
  }

  return true;
};
