// @business-app/fabric/lib-cjs/components/HighlightedText/HighlightedText.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccessibilityHighlightStyles = exports.getStyles = void 0;
var tslib_1 = require("tslib");
var getStyles = function (props) {
    return {
        highlights: {
            backgroundColor: '#f3c3ea',
            outline: '1px solid transparent'
        }
    };
};
exports.getStyles = getStyles;
// <mark> tags aren't handled well by screen-readers, we need pseudoelements to handle them
// see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/mark#Accessibility_concerns
// we aren't using <mark> any more because Windows High-Contrast Mode doesn't handle font color correctly,
// but the general principle applies
var pseudoStyles = {
    height: '0px',
    overflow: 'hidden',
    position: 'absolute',
    whiteSpace: 'nowrap',
    width: '0px'
};
/** Get styles to apply to the highlights portion of the IHighlightedTextStyles. */
var getAccessibilityHighlightStyles = function (labels) {
    return {
        selectors: {
            '::before': tslib_1.__assign(tslib_1.__assign({}, pseudoStyles), { content: "\" ;" + labels.beginHighlightedText + "; \"" }),
            '::after': tslib_1.__assign(tslib_1.__assign({}, pseudoStyles), { content: "\" ;" + labels.endHighlightedText + "; \"" })
        }
    };
};
exports.getAccessibilityHighlightStyles = getAccessibilityHighlightStyles;
