// @business-app/fabric/lib-cjs/components/NotificationsPanel/NotificationsPanel.test.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Enzyme = require("enzyme");
require("jest");
var List_1 = require("@fluentui/react/lib/List");
var Spinner_1 = require("@fluentui/react/lib/Spinner");
var React = require("react");
var EmptyState_1 = require("../EmptyState");
var NotificationsPanel_base_1 = require("./NotificationsPanel.base");
var NotificationsPanel_types_1 = require("./NotificationsPanel.types");
describe('NotificationsPanel', function () {
    it('renders one notification as expected', function () {
        var notification = {
            id: '1',
            title: 'title',
            body: 'App Checker Results',
            actionText: 'action text',
            headerPrimaryText: '1 hr ago',
            headerSecondaryText: '7/14/2018'
        };
        var wrapper = Enzyme.shallow(React.createElement(NotificationsPanel_base_1.NotificationsPanelBase, { onDismissAll: null, emptyStateMessage: 'No Notifications', dismissAllText: 'Dismiss all', notifications: [notification] }));
        var list = wrapper.find(List_1.List);
        expect(list.length).toBe(1);
        expect(list.prop('items').length).toBe(1);
        expect(list.prop('items')[0]).toEqual(notification);
    });
    it('Show spinner in loading state', function () {
        var props = {
            notifications: [],
            emptyStateMessage: 'No Notifications',
            dismissAllText: 'Dismiss all',
            onDismissAll: jest.fn(),
            panelState: NotificationsPanel_types_1.NotificationsPanelState.Loading,
            isOpen: true
        };
        var wrapper = Enzyme.shallow(React.createElement(NotificationsPanel_base_1.NotificationsPanelBase, tslib_1.__assign({}, props)));
        var panel = wrapper.find('Panel');
        expect(panel.length).toBe(1);
        var onRenderBodyWrapper = Enzyme.shallow(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        panel.props().onRenderBody());
        expect(onRenderBodyWrapper.find(Spinner_1.Spinner)).toHaveLength(1);
    });
    it('Show emptyState component in error state', function () {
        var props = {
            notifications: [],
            emptyStateMessage: 'No Notifications',
            dismissAllText: 'Dismiss all',
            onDismissAll: jest.fn(),
            panelState: NotificationsPanel_types_1.NotificationsPanelState.Error,
            isOpen: true,
            errorPrimaryMessage: 'error1',
            errorSecondaryMessage: 'error2'
        };
        var wrapper = Enzyme.shallow(React.createElement(NotificationsPanel_base_1.NotificationsPanelBase, tslib_1.__assign({}, props)));
        var panel = wrapper.find('Panel');
        expect(panel).toHaveLength(1);
        var onRenderBodyWrapper = Enzyme.shallow(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        panel.props().onRenderBody());
        var emptyState = onRenderBodyWrapper.find(EmptyState_1.EmptyState);
        expect(emptyState.length).toBe(1);
        expect(emptyState.props().primaryMessage).toBe('error1');
        expect(emptyState.props().secondaryMessage).toBe('error2');
    });
});
