// @business-app/fabric/lib/components/Sidebar/SidebarButton.styles.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { concatStyleSets } from '@fluentui/react/lib/Styling';
import { memoizeFunction } from '@fluentui/react/lib/Utilities';
import { getSidebarColors, sidebarFonts } from './Sidebar.styles';
import { SidebarStylingConstants } from './Sidebar.types';
export var getSidebarButtonStyles = memoizeFunction(function (theme, sidebarButtonStyles, customStyles) {
    var currentSidebarColors = getSidebarColors(theme);
    var buttonStyles = {
        root: {
            width: '100%',
            minWidth: SidebarStylingConstants.sidebarCollapsedWidth,
            padding: '0',
            minHeight: '40px',
            height: 'auto',
            lineHeight: '20px',
            borderColor: 'transparent',
            backgroundColor: currentSidebarColors.background
        },
        rootHovered: {
            backgroundColor: currentSidebarColors.backgroundHovered
        },
        rootFocused: {
            '@media screen and (-ms-high-contrast: active)': {
                border: '3px solid'
            }
        },
        flexContainer: {
            justifyContent: 'flex-start',
            minHeight: '40px',
            height: 'auto',
            selectors: {
                '.is-checked > &:before': {
                    content: '""',
                    width: '4px',
                    height: '24px',
                    backgroundColor: theme.palette.themePrimary,
                    borderRadius: '4px',
                    position: 'absolute',
                    left: '4px',
                    top: '50%',
                    transform: 'translateY(-50%)'
                }
            }
        },
        icon: {
            marginLeft: '16px',
            marginRight: '20px',
            fontSize: SidebarStylingConstants.sidebarIconSize,
            height: SidebarStylingConstants.sidebarIconSize,
            width: SidebarStylingConstants.sidebarIconSize
        },
        labelChecked: {
            fontWeight: 'bold',
            fontFamily: sidebarFonts.segoeUiSemibold
        },
        labelDisabled: {
            fontFamily: sidebarFonts.segoeUi
        },
        label: {
            verticalAlign: 'middle',
            textAlign: 'left',
            fontSize: '14px',
            whiteSpace: 'normal',
            display: 'inline-block',
            fontFamily: sidebarFonts.segoeUi,
            float: 'left',
            lineHeight: '20px',
            padding: '10px 0',
            fontWeight: 'normal'
        },
        menuIcon: {
            marginLeft: 'auto',
            marginRight: '10px',
            fontSize: SidebarStylingConstants.sidebarIconSize,
            height: SidebarStylingConstants.sidebarIconSize,
            width: SidebarStylingConstants.sidebarIconSize
        }
    };
    return concatStyleSets(buttonStyles, sidebarButtonStyles, customStyles);
});
export var getSidebarChildrenStyles = memoizeFunction(function (theme, customStyles) {
    var sidebarChildrenStyles = {
        flexContainer: {
            paddingLeft: '52px'
        },
        icon: {
            padding: '0',
            marginLeft: '0'
        }
    };
    return concatStyleSets(getSidebarButtonStyles(theme), sidebarChildrenStyles, customStyles);
});
