// @business-app/fabric/lib/components/Pane/Pane.styles.js
import { __assign } from "tslib";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { AnimationClassNames, getGlobalClassNames, HighContrastSelector, IconFontSizes, ScreenWidthMinXLarge } from '@fluentui/react/lib/Styling';
import { PaneSize } from './Pane.types';
var GlobalClassNames = {
    root: 'ba-Pane',
    main: 'ba-Pane-main',
    contentInner: 'ba-Pane-contentInner',
    scrollableContent: 'ba-Pane-scrollableContent',
    navigation: 'ba-Pane-navigation',
    commands: 'ba-Pane-commands',
    closeButton: 'ba-Pane-closeButton ba-PaneAction-close',
    header: 'ba-Pane-header',
    headerText: 'ba-Pane-headerText',
    content: 'ba-Pane-content',
    isOpen: 'is-open',
    hasCloseButton: 'ba-Pane--hasCloseButton'
};
var paneWidth = {
    full: '100%',
    sm: 340,
    md: 592,
    lg: 976
};
var paneMargin = {
    auto: 'auto',
    none: 0
};
var sharedPaddingStyles = {
    paddingLeft: '16px',
    paddingRight: '16px'
};
var smallPanelSelectors = {
    width: paneWidth.sm,
    minWidth: paneWidth.sm
};
var mediumPanelSelectors = {
    width: paneWidth.md,
    minWidth: paneWidth.md
};
var largePanelSelectors = {
    width: paneWidth.lg,
    minWidth: paneWidth.lg
};
var getPaneBreakpoints = function (type) {
    var selectors;
    switch (type) {
        case PaneSize.small:
            selectors = __assign({}, smallPanelSelectors);
            break;
        case PaneSize.medium:
            selectors = __assign({}, mediumPanelSelectors);
            break;
        case PaneSize.large:
            selectors = __assign({}, largePanelSelectors);
            break;
        default:
            break;
    }
    return selectors;
};
export var getStyles = function (props) {
    var _a, _b, _c;
    var className = props.className, hasCloseButton = props.hasCloseButton, headerClassName = props.headerClassName, isOpen = props.isOpen, theme = props.theme, _d = props.type, type = _d === void 0 ? PaneSize.small : _d;
    var palette = theme.palette, effects = theme.effects, fonts = theme.fonts;
    var classNames = getGlobalClassNames(GlobalClassNames, theme);
    return {
        root: [
            classNames.root,
            theme.fonts.medium,
            isOpen && classNames.isOpen,
            hasCloseButton && classNames.hasCloseButton,
            __assign(__assign({ pointerEvents: 'auto', position: 'relative', display: 'flex', flexDirection: 'column', overflowX: 'hidden', overflowY: 'auto', top: 0, left: 'auto', right: 0, bottom: 0 }, getPaneBreakpoints(type)), { height: '100%', backgroundColor: 'rgb(255, 255, 255)', borderLeft: "1px solid " + palette.neutralLight, borderRight: "1px solid " + palette.neutralLight, boxShadow: 'none', WebkitOverflowScrolling: 'touch', selectors: (_a = {},
                    _a[HighContrastSelector] = {
                        borderLeft: "3px solid " + palette.neutralLight,
                        borderRight: "3px solid " + palette.neutralLight
                    },
                    _a) }),
            isOpen && AnimationClassNames.slideRightIn40,
            isOpen && AnimationClassNames.slideLeftIn40,
            !isOpen && AnimationClassNames.slideLeftOut40,
            !isOpen && AnimationClassNames.slideRightOut40,
            className
        ],
        closeButton: [
            classNames.closeButton,
            {
                position: 'absolute',
                top: '0px',
                left: 'auto',
                right: '0px',
                marginTop: '14px',
                marginRight: '14px',
                marginBottom: '14px',
                marginLeft: '14px',
                flexShrink: '0',
                color: theme.palette.neutralSecondary,
                fontSize: IconFontSizes.large,
                selectors: (_b = {},
                    _b["@media (min-width: " + ScreenWidthMinXLarge + "px)"] = {
                        marginTop: '14px'
                    },
                    _b[':hover'] = { color: theme.palette.neutralPrimary },
                    _b)
            }
        ],
        contentInner: [
            classNames.contentInner,
            {
                display: 'flex',
                flexDirection: 'column',
                flexGrow: 1,
                overflowY: 'hidden'
            }
        ],
        header: [
            classNames.header,
            sharedPaddingStyles,
            {
                margin: '14px 0',
                flexShrink: 0,
                selectors: (_c = {},
                    _c["@media (min-width: " + ScreenWidthMinXLarge + "px)"] = {
                        marginTop: '14px'
                    },
                    _c)
            }
        ],
        headerText: [
            classNames.headerText,
            fonts.xLarge,
            {
                color: palette.neutralPrimary,
                lineHeight: '27px',
                margin: 0,
                overflowWrap: 'break-word',
                wordWrap: 'break-word',
                wordBreak: 'break-word',
                hyphens: 'auto'
            },
            headerClassName
        ],
        scrollableContent: [
            classNames.scrollableContent,
            sharedPaddingStyles,
            {
                overflowY: 'auto'
            }
        ],
        content: [
            classNames.content,
            sharedPaddingStyles,
            {
                marginBottom: 0,
                paddingBottom: 20
            }
        ]
    };
};
