// @business-app/fabric/lib/components/Panel/Panel.styles.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { FontSizes, getTheme, mergeStyleSets } from '@fluentui/react/lib/Styling';
export var panelStyleClasses = mergeStyleSets({
    navigationBackbar: [
        'ba-Panel-navigation-backBar',
        {
            padding: '8px 10px 0',
            height: '36px',
            display: 'flex',
            selectors: {
                i: { fontSize: FontSizes.medium }
            }
        }
    ],
    panelHeader: [
        'ba-Panel-header',
        {
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'space-between'
        }
    ],
    panelHeaderLeft: [
        'ba-Panel-headerLeft',
        {
            display: 'flex',
            alignItems: 'center',
            overflow: 'hidden'
        }
    ],
    panelHeaderText: [
        'ba-Panel-headerText',
        {
            margin: 0,
            paddingBottom: '4px',
            fontSize: '20px',
            fontWeight: '600',
            outline: 'none',
            whiteSpace: 'nowrap',
            textOverflow: 'ellipsis',
            overflow: 'hidden'
        }
    ],
    panelFooter: [
        'ba-Panel-footer',
        {
            display: 'flex',
            flexDirection: 'row',
            justifyContent: 'flex-start'
        }
    ],
    panelFooterButton: [
        'ba-Panel-footerButton',
        {
            marginRight: '8px'
        }
    ],
    emptyStateContainer: [
        'ba-Panel-emptyStateContainer',
        {
            flex: 1,
            display: 'flex',
            flexDirection: 'column',
            justifyContent: 'center'
        }
    ]
});
export var panelStyles = {
    root: [
        'ms-Panel',
        {
            selectors: {
                'ms-Panel-main': {
                    display: 'flex',
                    flexDirection: 'column'
                },
                "div[class^='ms-Panel-content']": {
                    paddingLeft: 0,
                    paddingRight: 0
                },
                '.ms-Overlay .ms-Spinner': {
                    position: 'relative',
                    top: '40%',
                    transform: 'translateY(-50%)'
                },
                '.ba-Panel-headerLeft .ms-Button--icon': {
                    paddingLeft: 0
                },
                '.ba-Panel-headerLeft .ms-Button-flexContainer .ms-Button-icon': { marginLeft: 0 },
                '.ba-Panel-headerLeft .ms-Button-flexContainer': { justifyContent: 'flex-start' }
            }
        }
    ],
    contentInner: [
        'ms-Panel-contentInner',
        {
            position: 'unset',
            flex: 1,
            selectors: {
                "div[class^='ba-Panel-headerContainer']": {
                    paddingLeft: '16px',
                    paddingRight: '16px',
                    paddingTop: '14px',
                    paddingBottom: '4px'
                }
            }
        }
    ],
    scrollableContent: {
        paddingLeft: '16px',
        paddingRight: '16px'
    },
    footer: {
        borderTop: '1px solid',
        borderTopColor: getTheme().palette.neutralLight,
        padding: '0px 16px 20px 16px'
    },
    footerInner: [
        'ms-Panel-footerInner',
        {
            paddingBottom: '0px !important',
            paddingRight: '0px !important',
            paddingLeft: '0px !important'
        }
    ]
};
