// @business-app/fabric/src/components/ColorInput/examples/ColorInput.Basic.Example.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import * as React from 'react';
import { ColorInput } from '..';
import { isValidColor } from '../ColorInput.base';

export class ColorInputBasicExample extends React.Component {
  private getErrorMessage = (value: string): string => {
    return !isValidColor(value) ? `Please enter a valid color value.` : '';
  };

  public render(): JSX.Element {
    return (
      <ColorInput
        label="Color Input"
        tooltipIconProps={{
          iconProps: { iconName: 'Info', styles: { root: { color: '#666666' } } },
          ariaLabel: 'New Color Input'
        }}
        tooltipProps={{ content: 'New Color Input' }}
        colorBoxProps={{
          onClick: () => {
            return;
          },
          ariaLabel: 'Show color picker',
          title: 'Show color picker',
          disabled: false
        }}
        colorTextFieldProps={{
          onBlur: () => {
            return;
          },
          ariaLabel: 'Enter color code',
          disabled: false,
          onGetErrorMessage: (value: string) => {
            return this.getErrorMessage(value);
          }
        }}
        colorPickerProps={{
          color: 'Red'
        }}
      />
    );
  }
}
