// @business-app/fabric/lib-cjs/components/Page/examples/Page.Basic.Example.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageBasicExample = void 0;
var tslib_1 = require("tslib");
var Breadcrumb_1 = require("@fluentui/react/lib/Breadcrumb");
var ChoiceGroup_1 = require("@fluentui/react/lib/ChoiceGroup");
var Label_1 = require("@fluentui/react/lib/Label");
var MessageBar_1 = require("@fluentui/react/lib/MessageBar");
var Pivot_1 = require("@fluentui/react/lib/Pivot");
var Toggle_1 = require("@fluentui/react/lib/Toggle");
var React = require("react");
var DetailsList_1 = require("@fluentui/react/lib/DetailsList");
var Pivot_2 = require("@fluentui/react/lib/Pivot");
var __1 = require("../");
var contextualMenuItems_1 = require("../../../utilities/data/contextualMenuItems");
var CommandBarWithSearchBox_1 = require("../../CommandBarWithSearchBox");
var Pane_1 = require("../../Pane");
var categories = ['Cat', 'House', 'Table'];
var PageBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(PageBasicExample, _super);
    function PageBasicExample(props) {
        var _this = _super.call(this, props) || this;
        _this.onChangePaneType = function (ev, option) {
            var paneType;
            switch (option.key) {
                case 'A':
                    paneType = Pane_1.PaneSize.small;
                    break;
                case 'B':
                    paneType = Pane_1.PaneSize.medium;
                    break;
                default:
                    paneType = Pane_1.PaneSize.small;
                    break;
            }
            _this.setState({ paneType: paneType, isOpenPane: false });
        };
        _this._onRenderPaneBody = function () {
            return React.createElement("div", null, "Content goes here...");
        };
        _this._onDismissPane = function () {
            _this.setState({ isOpenPane: false });
        };
        _this._onTogglePane = function (isOpenPane) {
            _this.setState({ isOpenPane: isOpenPane });
        };
        _this._renderBody = function () {
            return (React.createElement("div", null,
                React.createElement(DetailsList_1.DetailsList, { items: _this.detailsListItems, columns: _this.columns, setKey: "set", layoutMode: DetailsList_1.DetailsListLayoutMode.justified, selectionPreservedOnEmptyClick: true, ariaLabelForSelectionColumn: "Toggle selection", ariaLabelForSelectAllCheckbox: "Toggle selection for all items", checkButtonAriaLabel: "Row checkbox" }),
                React.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."),
                React.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."),
                React.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."),
                React.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."),
                React.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."),
                React.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.")));
        };
        _this.state = {
            isLoading: false,
            isOpenPane: false,
            paneType: Pane_1.PaneSize.small
        };
        _this.detailsListItems = [];
        for (var i = 0; i < 75; i++) {
            _this.detailsListItems.push({
                key: i,
                name: 'Item ' + i,
                value: i
            });
        }
        _this.columns = [
            { key: 'column1', name: 'Name', fieldName: 'name', minWidth: 100, maxWidth: 200, isResizable: true },
            { key: 'column2', name: 'Value', fieldName: 'value', minWidth: 100, maxWidth: 200, isResizable: true }
        ];
        _this._onLoadingSet = _this._onLoadingSet.bind(_this);
        return _this;
    }
    PageBasicExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(Toggle_1.Toggle, { label: "Show loading overlay", checked: this.state.isLoading, onChanged: this._onLoadingSet, onText: "Loading", offText: "Loaded" }),
            React.createElement(ChoiceGroup_1.ChoiceGroup, { className: "defaultChoiceGroup", defaultSelectedKey: "A", options: [
                    {
                        key: 'A',
                        text: 'Small pane'
                    },
                    {
                        key: 'B',
                        text: 'Medium pane'
                    }
                ], onChange: this.onChangePaneType, label: "Pane type", required: true }),
            React.createElement(Toggle_1.Toggle, { label: "Show Pane", checked: this.state.isOpenPane, onChanged: this._onTogglePane }),
            React.createElement(Label_1.Label, null, "The page is surrounded by a black border so you can tell where the page ends."),
            React.createElement("div", { style: { border: '2px solid black' } },
                React.createElement(__1.Page, { style: { height: '500px', overflowX: 'scroll' }, paneProps: {
                        isOpen: this.state.isOpenPane,
                        headerText: 'Pane Example',
                        onDismiss: this._onDismissPane,
                        hasCloseButton: true,
                        onRenderBody: this._onRenderPaneBody,
                        type: this.state.paneType
                    }, spinnerProps: {
                        ariaLabel: 'Loading'
                    }, showOverlay: this.state.isLoading, renderHeader: function () { return React.createElement(CommandBarWithSearchBox_1.CommandBarWithSearchBox, { items: contextualMenuItems_1.items }); }, renderMessageBar: function () { return (React.createElement(MessageBar_1.MessageBar, { messageBarType: MessageBar_1.MessageBarType.error }, "Lets display an error")); }, renderFooter: function () { return React.createElement(MessageBar_1.MessageBar, null, "This is a page footer"); }, renderTitle: function () { return (React.createElement(Breadcrumb_1.Breadcrumb, { items: [
                            { text: 'Test', key: 'Test' },
                            { text: 'Header', key: 'Header' }
                        ] })); }, renderPivot: function () { return (React.createElement(Pivot_2.Pivot, null,
                        React.createElement(Pivot_1.PivotItem, { linkText: 'Test' }),
                        React.createElement(Pivot_1.PivotItem, { linkText: 'Another Test' }),
                        React.createElement(Pivot_1.PivotItem, { linkText: 'One More' }),
                        React.createElement(Pivot_1.PivotItem, { linkText: 'Hi' }))); }, renderBody: this._renderBody }))));
    };
    PageBasicExample.prototype._onLoadingSet = function (isLoading) {
        this.setState({ isLoading: isLoading });
    };
    return PageBasicExample;
}(React.Component));
exports.PageBasicExample = PageBasicExample;
