// @business-app/fabric/src/components/SettingsExplorer/SettingsExplorer.styles.tsx
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IStyle, mergeStyles } from '@fluentui/react/lib/Styling';
import { IStyleFunction } from '@fluentui/react/lib/Utilities';
import {
  ISettingsExplorerStyleProps,
  ISettingsExplorerStyles,
  ITwoTierSettingsExplorerStyles
} from './SettingsExplorer.types';

const rootStyles: IStyle = {
  display: 'flex',
  flexFlow: 'row nowrap',
  justifyContent: 'flex-start',
  height: 632,
  maxHeight: '80vh',
  borderSpacing: '0',
  borderCollapse: 'collapse',
  padding: 0,
  overflow: 'hidden'
};

const pageContainer: IStyle = {
  width: 776,
  display: 'flex',
  flexGrow: 1
};

export const settingsStyles: IStyleFunction<ISettingsExplorerStyleProps, ISettingsExplorerStyles> = ({
  theme: { palette }
}) => ({
  root: [
    rootStyles,
    {
      backgroundColor: palette.white
    }
  ],
  pageContainer
});

export const twoTierStyles: IStyleFunction<ISettingsExplorerStyleProps, ITwoTierSettingsExplorerStyles> = ({
  theme: { palette }
}) => ({
  root: [rootStyles],
  pageContainer
});

export const itemClassName = (propStyles: IStyle) => mergeStyles({ paddingBottom: 8 }, propStyles);
