// @business-app/fabric/lib-cjs/components/Wizard/Progress/WizardProgress.base.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WizardProgressBase = void 0;
var tslib_1 = require("tslib");
var Icon_1 = require("@fluentui/react/lib/Icon");
var Styling_1 = require("@fluentui/react/lib/Styling");
var Utilities_1 = require("@fluentui/react/lib/Utilities");
var Utilities_2 = require("@fluentui/react/lib/Utilities");
var React = require("react");
var StepStatus_1 = require("../StepStatus");
var getClassNames = Utilities_1.classNamesFunction();
var WizardProgressBase = /** @class */ (function (_super) {
    tslib_1.__extends(WizardProgressBase, _super);
    function WizardProgressBase() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            isWizardProgressCollapsedOnSmallScreens: true
        };
        _this.onToggleClicked = function () {
            _this.setState({
                isWizardProgressCollapsedOnSmallScreens: !_this.state.isWizardProgressCollapsedOnSmallScreens
            });
        };
        return _this;
    }
    WizardProgressBase.prototype.render = function () {
        var _this = this;
        var _a = this.props, styles = _a.styles, currentIndex = _a.currentIndex, stepStatuses = _a.stepStatuses, theme = _a.theme, className = _a.className, expandButtonLabel = _a.expandButtonLabel, collapseButtonLabel = _a.collapseButtonLabel, onProgressStepClick = _a.onProgressStepClick;
        var classNames = getClassNames(styles, {
            theme: theme,
            className: className
        });
        return (React.createElement("nav", { className: Utilities_2.css(classNames.root, this.state.isWizardProgressCollapsedOnSmallScreens && 'is-collapsed') },
            React.createElement("ol", { className: classNames.list }, this.props.stepNames.map(function (stepName, stepIndex) { return (React.createElement("li", { key: stepName, className: classNames.itemWrapper },
                React.createElement("div", { className: _this.getStepClassName(stepIndex, currentIndex, classNames), "aria-label": stepName, "aria-describedby": "WizardProgress-step" + stepIndex + "-description WizardProgress-step" + stepIndex + "-summary-container", 
                    // eslint-disable-next-line react/jsx-no-bind
                    onClick: onProgressStepClick ? onProgressStepClick.bind(_this, stepIndex) : undefined, style: { cursor: onProgressStepClick ? 'pointer' : undefined }, role: onProgressStepClick ? 'button' : undefined, tabIndex: onProgressStepClick ? 0 : undefined },
                    React.createElement(Icon_1.Icon, { iconName: getIconForStepStatus(stepStatuses[stepIndex]), className: _this.getIconClassName(stepIndex, classNames) }),
                    React.createElement("span", { className: classNames.label }, stepName),
                    React.createElement("div", { id: "WizardProgress-step" + stepIndex + "-summary-container" }, _this.props.stepSummaries[stepIndex].map(function (summary, i) { return (React.createElement("div", { key: i, className: classNames.summary }, summary)); }))),
                React.createElement("div", { className: classNames.visuallyHidden, id: "WizardProgress-step" + stepIndex + "-description" }, _this.props.computeAriaDescription(stepIndex)))); })),
            React.createElement("button", { className: classNames.toggle, onClick: this.onToggleClicked, "aria-expanded": !this.state.isWizardProgressCollapsedOnSmallScreens, "aria-label": this.state.isWizardProgressCollapsedOnSmallScreens ? expandButtonLabel : collapseButtonLabel, title: this.state.isWizardProgressCollapsedOnSmallScreens ? expandButtonLabel : collapseButtonLabel },
                React.createElement(Icon_1.Icon, { iconName: "Cancel" }))));
    };
    WizardProgressBase.prototype.getIconClassName = function (stepIndex, classNames) {
        var _a = this.props, stepStatuses = _a.stepStatuses, currentIndex = _a.currentIndex;
        var status = stepStatuses[stepIndex];
        switch (status) {
            case StepStatus_1.StepStatus.Valid:
                return classNames.validIcon;
            case StepStatus_1.StepStatus.Warning:
                return classNames.warningIcon;
            case StepStatus_1.StepStatus.Error:
                return classNames.errorIcon;
            case StepStatus_1.StepStatus.Empty:
                if (stepIndex === currentIndex) {
                    return classNames.currentEmptyIcon;
                }
                else {
                    return classNames.emptyIcon;
                }
            default:
                break;
        }
        return '';
    };
    WizardProgressBase.prototype.getStepClassName = function (stepIndex, currentIndex, classNames) {
        var className = '';
        if (stepIndex < currentIndex) {
            className = classNames.passedStep;
        }
        else if (stepIndex === currentIndex) {
            className = classNames.currentStep;
        }
        else if (stepIndex > currentIndex) {
            className = classNames.futureStep;
        }
        if (stepIndex === this.props.stepNames.length - 1) {
            className = Styling_1.mergeStyles(className, classNames.lastStep);
        }
        return className;
    };
    return WizardProgressBase;
}(React.Component));
exports.WizardProgressBase = WizardProgressBase;
function getIconForStepStatus(stepStatus) {
    switch (stepStatus) {
        case StepStatus_1.StepStatus.Valid:
            return 'CheckMark';
        case StepStatus_1.StepStatus.Warning:
            return 'WarningSolid';
        case StepStatus_1.StepStatus.Error:
            return 'StatusErrorFull';
        case StepStatus_1.StepStatus.Empty:
            return '';
        default:
            throw new Error('Not implemented');
    }
}
