// @business-app/fabric/lib/components/SettingsExplorer/ExplorerPanel.styles.js
// If the width of the panel is less than this, the panel will collapse to the icon-sized version
export var minWidth = 172;
export function getStyles(props) {
    var theme = props.theme, isCollapsed = props.isCollapsed;
    var palette = theme.palette;
    var commonButtonStyle = {
        fontSize: 14,
        width: '100%',
        height: isCollapsed ? 'unset' : 40,
        minHeight: isCollapsed ? 40 : undefined,
        margin: 0,
        paddingLeft: 20,
        textAlign: 'left',
        borderColor: 'transparent',
        padding: isCollapsed ? '10px 0px' : undefined
    };
    return {
        root: {
            margin: 0,
            paddingInlineStart: 0,
            overflowY: 'auto',
            overflowX: 'hidden',
            borderRight: '1px solid',
            borderWidth: 1,
            borderColor: palette.neutralLight,
            backgroundColor: palette.neutralLighter,
            paddingTop: 16,
            width: isCollapsed ? 120 : 200,
            minWidth: 76
        },
        header: {
            flexGrow: 1,
            fontSize: 20,
            fontWeight: 600,
            alignContent: 'flex-start',
            paddingLeft: 24,
            marginBottom: 20
        },
        notSelectedButton: [
            commonButtonStyle,
            {
                backgroundColor: palette.neutralLighter,
                selectors: {
                    ':hover': {
                        backgroundColor: palette.neutralQuaternaryAlt
                    }
                },
                '.ms-Button-label': {
                    fontWeight: 400
                }
            }
        ],
        selectedButton: [
            commonButtonStyle,
            {
                backgroundColor: palette.white,
                selectors: {
                    ':hover': {
                        backgroundColor: palette.neutralQuaternaryAlt
                    }
                }
            }
        ],
        subComponentStyles: {
            buttonStyles: isCollapsed && {
                root: { minWidth: 0 },
                flexContainer: { justifyContent: 'start' },
                textContainer: { display: 'grid' },
                label: { lineHeight: 17, fontSize: 13, textOverflow: 'ellipsis', overflow: 'hidden' }
            }
        }
    };
}
