// @business-app/fabric/src/components/EmptyState/EmptyState.styles.ts
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */

import { IStyleFunction } from '@fluentui/react/lib/Utilities';

import { IEmptyStateStyleProps, IEmptyStateStyles } from './EmptyState.types';

export const styles: IStyleFunction<IEmptyStateStyleProps, IEmptyStateStyles> = ({
  theme,
  className,
  buttonClassName
}) => {
  const { neutralSecondary, neutralTertiary } = theme.palette;

  return {
    root: [
      'ba-EmptyState',
      className,
      {
        color: neutralSecondary,
        margin: '0 auto',
        marginBottom: 20,
        display: 'flex',
        alignItems: 'center',
        flexDirection: 'column',
        selectors: {
          '> i': {
            fontSize: 48,
            border: `2.5px solid ${neutralTertiary}`,
            borderRadius: 999999,
            padding: 0,
            width: 86,
            height: 86,
            textAlign: 'center',
            lineHeight: 86,
            color: neutralTertiary,
            marginBottom: 20
          }
        }
      }
    ],
    primaryMessage: [
      'ba-EmptyState-PrimaryMessage',
      {
        fontSize: 14,
        fontWeight: 'bold',
        textAlign: 'center',
        margin: 0,
        marginBottom: 16,
        width: '100%'
      }
    ],
    secondaryMessage: [
      'ba-EmptyState-SecondaryMessage',
      {
        fontSize: 12,
        lineHeight: 16,
        textAlign: 'center',
        margin: 0,
        marginBottom: 16,
        width: '100%'
      }
    ],
    button: [
      'ba-EmptyState-Button',
      buttonClassName,
      {
        marginTop: 8
      }
    ]
  };
};
