// @business-app/fabric/lib-cjs/components/Page/Page.styles.js
"use strict";
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = void 0;
var getStyles = function (_a) {
    var className = _a.className, paneProps = _a.paneProps;
    return ({
        root: [
            'ba-Page',
            className,
            {
                display: 'flex',
                flexDirection: 'column',
                position: 'relative',
                height: '100%'
            }
        ],
        container: [
            'ba-Page-Container',
            {
                display: 'flex',
                flexDirection: 'row',
                position: 'relative',
                width: '100%',
                height: '100%'
            }
        ],
        content: [
            'ba-Page-Content',
            {
                width: '100%'
            },
            paneProps && paneProps.isOpen && { overflow: 'auto' }
        ],
        header: [
            'ba-Page-Header',
            {
                position: 'sticky',
                top: 0,
                zIndex: 1,
                selectors: {
                    '.ms-CommandBar': {
                        paddingLeft: 32
                    }
                }
            }
        ],
        body: [
            'ba-Page-Body',
            {
                flex: '1 1 auto',
                position: 'relative',
                padding: '0 32px'
            }
        ],
        title: [
            'ba-Page-Title',
            {
                flex: '0 0 auto',
                paddingBottom: 10
            }
        ],
        pivot: [
            'ba-Page-Pivot',
            {
                paddingBottom: 20
            }
        ],
        footer: [
            'ba-Page-Footer',
            {
                flex: '0 0 auto'
            }
        ],
        loadingOverlay: [
            'ba-Page-LoadingOverlay',
            {
                display: 'flex',
                position: 'absolute',
                justifyContent: 'center',
                alignItems: 'center',
                zIndex: 10000000 // Have to fight office fabric Layers
            }
        ]
    });
};
exports.getStyles = getStyles;
