// @business-app/fabric/lib/components/Wizard/examples/Wizard.Basic.Example.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __extends } from "tslib";
import * as React from 'react';
import { StepStatus } from '../StepStatus';
import { WizardTipsAndHelp } from '../TipsAndHelp/WizardTipsAndHelp';
import { TipsAndHelpCard } from '../TipsAndHelpCard/TipsAndHelpCard';
import { Wizard } from '../Wizard';
var steps = [
    {
        name: 'Step 1',
        summary: ['Summary 1'],
        ariaDescription: 'Summary 1'
    },
    {
        name: 'Step 2',
        summary: ['Summary 2'],
        ariaDescription: 'Summary 2'
    },
    {
        name: 'Step 3',
        hasWarning: true,
        summary: ['Summary 3'],
        ariaDescription: 'Summary 3'
    },
    {
        name: 'Step 4',
        hasError: true,
        summary: ['Summary 4'],
        ariaDescription: 'Summary 4'
    },
    {
        name: 'Step 5',
        summary: ['Summary 5'],
        ariaDescription: 'Summary 5'
    }
];
var WizardBasicExample = /** @class */ (function (_super) {
    __extends(WizardBasicExample, _super);
    function WizardBasicExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            currentIndex: 0
        };
        _this.getStatuses = function () {
            return steps.map(function (s, index) {
                if (index >= _this.state.currentIndex) {
                    return StepStatus.Empty;
                }
                if (s.hasError) {
                    return StepStatus.Error;
                }
                if (s.hasWarning) {
                    return StepStatus.Warning;
                }
                return StepStatus.Valid;
            });
        };
        return _this;
    }
    WizardBasicExample.prototype.render = function () {
        var _this = this;
        return (React.createElement(Wizard, { className: "my-wizard-example", progressProps: {
                currentIndex: this.state.currentIndex,
                stepNames: steps.map(function (s) { return s.name; }),
                stepStatuses: this.getStatuses(),
                stepSummaries: steps.map(function (s) { return s.summary; }),
                computeAriaDescription: function (index) { return steps[index].ariaDescription; },
                // this callback is optional in IWizardProgressProps, in an instance maybe use wizardStore.goToNthStep
                onProgressStepClick: this.setIndex,
                styles: {
                    root: {
                        flex: '0 0 auto',
                        width: '19vw',
                        minWidth: '183px',
                        maxWidth: '280px'
                    }
                },
                expandButtonLabel: 'Expand timeline',
                collapseButtonLabel: 'Collapse timeline'
            }, footerProps: {
                backButtonText: 'Back',
                canGoToNextStep: this.state.currentIndex < steps.length - 1,
                canGoToPreviousStep: this.state.currentIndex > 0,
                continueButtonText: 'Next',
                summaryText: 'Summary text',
                shouldDisplayBackButton: true,
                goToPreviousStep: function () {
                    _this.setState({
                        currentIndex: _this.state.currentIndex - 1
                    });
                },
                goToNextStep: function () {
                    _this.setState({
                        currentIndex: _this.state.currentIndex + 1
                    });
                }
            }, onRenderHeaderItems: function () { return 'Title'; }, onRenderHeaderFarItems: function () { return React.createElement("button", null, "Save"); }, onRenderStep: function () { return React.createElement("span", null,
                "This is step ",
                _this.state.currentIndex + 1); }, onRenderRightPane: function () { return (React.createElement(WizardTipsAndHelp, { title: "Tips", closeButtonText: "Close", styles: {
                    root: {
                        flex: '0 0 auto',
                        width: '21vw',
                        minWidth: '180px',
                        maxWidth: '272px'
                    }
                } },
                React.createElement(TipsAndHelpCard, { card: {
                        title: 'Get help or send feedback',
                        content: 'Get the answers you need, or tell us about your experiences.',
                        action: {
                            iconName: 'Feedback',
                            name: 'Get help',
                            onClick: function () {
                                // eslint-disable-next-line no-console
                                console.log('help');
                            }
                        }
                    } }))); }, styles: {
                root: {
                    minHeight: '500px'
                }
            } }));
    };
    WizardBasicExample.prototype.setIndex = function (index) {
        this.setState({ currentIndex: index });
    };
    return WizardBasicExample;
}(React.Component));
export { WizardBasicExample };
