// @business-app/fabric/lib/components/HighlightedText/HighlightedText.styles.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { __assign } from "tslib";
export var getStyles = function (props) {
    return {
        highlights: {
            backgroundColor: '#f3c3ea',
            outline: '1px solid transparent'
        }
    };
};
// <mark> tags aren't handled well by screen-readers, we need pseudoelements to handle them
// see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/mark#Accessibility_concerns
// we aren't using <mark> any more because Windows High-Contrast Mode doesn't handle font color correctly,
// but the general principle applies
var pseudoStyles = {
    height: '0px',
    overflow: 'hidden',
    position: 'absolute',
    whiteSpace: 'nowrap',
    width: '0px'
};
/** Get styles to apply to the highlights portion of the IHighlightedTextStyles. */
export var getAccessibilityHighlightStyles = function (labels) {
    return {
        selectors: {
            '::before': __assign(__assign({}, pseudoStyles), { content: "\" ;" + labels.beginHighlightedText + "; \"" }),
            '::after': __assign(__assign({}, pseudoStyles), { content: "\" ;" + labels.endHighlightedText + "; \"" })
        }
    };
};
